{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths( c( .libPaths(), \"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\") )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(stringr, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(data.table, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(ggbeeswarm, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(glue, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(ggforce, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(pec, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(patchwork, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(survival, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(survminer, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(IRdisplay, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(cowplot, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "# library(igraph)\n",
    "library(ggpubr, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\") # load before survminer\n",
    "library(ggrepel, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(kableExtra, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "source('/work/isabl/home/gaot/facets-ch/R/facets-ch-utils.R')\n",
    "source('/work/isabl/home/gaot/ch_cnv/ch-cnv/notebooks/plot_forest.R')\n",
    "setwd('/work/isabl/home/gaot/ch_cnv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"ggpubr\", lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "library(dplyr, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(stringr, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(data.table, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(ggbeeswarm, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(glue, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(ggforce, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(pec, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(patchwork, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(survival, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(survminer, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(IRdisplay, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(cowplot, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "# library(igraph)\n",
    "library(ggpubr, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\") # load before survminer\n",
    "library(ggrepel, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "library(kableExtra, lib.loc=\"/work/isabl/home/sirenkom/R/R-3.6.1/R-3.6.1/library\")\n",
    "source('/work/isabl/home/gaot/facets-ch/R/facets-ch-utils.R')\n",
    "source('/work/isabl/home/gaot/ch_cnv/ch-cnv/notebooks/plot_forest.R')\n",
    "setwd('/work/isabl/home/gaot/ch_cnv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide = fread('/work/isabl/home/sirenkom/teng_mCA/data/M_wide_preprocessed.tsv',  sep = '\\t')\n",
    "\n",
    "M_long = fread('/work/isabl/home/sirenkom/teng_mCA/data/M_long_preprocessed.tsv', sep = '\\t')\n",
    "\n",
    "segs_filtered = fread('/work/isabl/home/sirenkom/teng_mCA/data/segs_filtered_preprocessed.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# M_wide = fread('./data/sm_wide_jul13.tsv')\n",
    "# # segs = fread('./data/segs_aberrant_may6.tsv', sep = '\\t')\n",
    "\n",
    "# segs = rbind(\n",
    "#     fread('./data/segs_aberrant_may18.tsv', sep = '\\t'),\n",
    "#     fread('./data/segs_aberrant_nonparta.tsv', sep = '\\t'),\n",
    "#     fill = TRUE\n",
    "# )\n",
    "\n",
    "# segs_filtered_old = fread('./data/segs_aberrant_apr19.tsv', sep = '\\t')\n",
    "gene_bed = fread('./data/impact_genes.tsv') %>% \n",
    "    mutate(chrom = ifelse(chrom == 'X', 23, chrom)) %>%\n",
    "    mutate(chrom = ifelse(chrom == 'Y', 24, chrom)) %>%\n",
    "    mutate(chrom = as.integer(chrom))\n",
    "chrom_lens = fread('/home/gaot/ref/chrom_lens.tsv') %>% filter(chrom != 24)\n",
    "\n",
    "acen = fread('/home/gaot/ref/acen.tsv')\n",
    "colnames(acen) = c('chrom', 'start', 'end', 'arm', 'type')\n",
    "chrom_lens = fread('/home/gaot/ref/chrom_lens.tsv') %>% filter(chrom != 24)\n",
    "chrom_levels = c(as.character(1:22), 'X')\n",
    "\n",
    "acen = acen %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "    filter(chrom != 'Y') %>%\n",
    "    mutate(\n",
    "        chrom = ifelse(chrom == 'X', 23, chrom),\n",
    "        chrom = as.integer(chrom)\n",
    "    ) %>%\n",
    "    arrange(chrom) %>%\n",
    "    group_by(chrom) %>%\n",
    "    summarise(cen_start = min(start), cen_end = max(end)) %>%\n",
    "    ungroup()\n",
    "\n",
    "# # mCA positive patient EMR text search\n",
    "# emr_review = fread('./data/clinicalreview/cnv_patient_asm_review.tsv') %>% filter(exclude == 1)\n",
    "\n",
    "# M_wide = M_wide %>% \n",
    "#     mutate(\n",
    "#         DateofProcedure = as.Date(DateofProcedure, format = '%Y-%m-%d'),\n",
    "#         date_birth = as.Date(date_birth, format = '%Y-%m-%d'),\n",
    "#         date_death = as.Date(date_death, format = '%Y-%m-%d')\n",
    "#     ) %>%\n",
    "#     mutate(age = as.numeric((DateofProcedure - date_birth)/365)) %>%\n",
    "#     filter(!active_heme) %>%\n",
    "#     filter(QC %in% c('Passed', '#N/A')) %>%\n",
    "#     filter(age > 20) %>%\n",
    "#     filter(!dmp_patient_id %in% emr_review$dmp_patient_id)\n",
    "\n",
    "# # clean up heme diagnosis\n",
    "# M_wide = M_wide %>%\n",
    "#     mutate(heme_cat_billing = heme_cat) %>%\n",
    "#     mutate(heme_cat = ifelse(is.na(heme_cat_review), heme_cat, heme_cat_review)) %>%\n",
    "#     mutate(\n",
    "#         manual_review = as.integer(!is.na(valid_diagnosis)),\n",
    "#         valid_diagnosis = ifelse(is.na(valid_diagnosis), 0, valid_diagnosis)\n",
    "#     ) %>%\n",
    "#     mutate(\n",
    "#         heme_cat = ifelse(valid_diagnosis == 0 & manual_review, 'Invalid', heme_cat),\n",
    "#         post_heme = ifelse(valid_diagnosis == 1 & manual_review, secondary_heme, 0)\n",
    "#     ) %>%\n",
    "#     mutate(\n",
    "#         post_cll = as.integer(heme_cat == 'CLL' & post_heme),\n",
    "#         post_mpn = as.integer(heme_cat %in% c('PV', 'ET', 'MPN') & post_heme),\n",
    "#         post_tmn = as.integer(heme_cat %in% c('AML', 'MDS') & post_heme),\n",
    "#         post_cml = as.integer(heme_cat %in% c('CML') & post_heme)\n",
    "#     ) %>%\n",
    "#     mutate(\n",
    "#         post_anyblood = post_heme,\n",
    "#         post_mn = as.integer(post_mpn | post_tmn),\n",
    "#         post_leukemia = as.integer(post_cll | post_mpn | post_tmn | post_cml)\n",
    "#     ) %>%\n",
    "#     mutate(heme_cat = ifelse(heme_cat %in% c('ET', 'PV'), 'MPN', heme_cat))\n",
    "\n",
    "# # recode hemecat\n",
    "M_wide = M_wide %>% \n",
    "#     mutate(heme_cat2 = heme_cat) %>%\n",
    "#     mutate(\n",
    "#         heme_cat2 = ifelse(heme_cat2 == 'LCH', NA, heme_cat2),\n",
    "#         heme_cat2 = ifelse(heme_cat2 %in% c('ET', 'PV'), 'MPN', heme_cat2),\n",
    "#         heme_cat2 = ifelse(heme_cat2 %in% c('ALCL', 'CTCL'), 'T-NHL', heme_cat2),\n",
    "#         heme_cat2 = ifelse(heme_cat2 %in% c('DLBCL', 'FL', 'MZL', 'PCNSL', 'WM'), 'B-NHL', heme_cat2)\n",
    "#     ) %>%\n",
    "    mutate(heme_cat2 = factor(heme_cat2, c('MDS', 'MPN', 'AML', 'CML', 'CLL', 'B-NHL', 'T-NHL', 'MM')))\n",
    "\n",
    "# M_wide = M_wide %>% mutate(post_heme = ifelse(is.na(heme_cat2), 0, post_heme))\n",
    "\n",
    "# display(paste0(sum(M_wide$post_tmn), 'tMN, ', sum(M_wide$post_cll), 'CLL, ', sum(M_wide$post_mpn), 'MPN'))\n",
    "\n",
    "# # clean up survival data\n",
    "# M_wide = M_wide %>%\n",
    "#     mutate(race_b = ifelse(Race == \"WHITE\",1,0)) %>%\n",
    "#     rowwise() %>%\n",
    "#     mutate(\n",
    "#         date_lastfu = as.Date(date_lastfu, format = '%Y-%m-%d'),\n",
    "#         # sometimes last contact date is not accurate\n",
    "#         time_lastfu = max(as.integer(date_lastfu - DateofProcedure), 0), \n",
    "#         time_tmn = ifelse(post_tmn == 1, seq_to_diag, time_lastfu),\n",
    "#         time_mn = ifelse(post_mn == 1, seq_to_diag, time_lastfu),\n",
    "#         time_cll = ifelse(post_cll == 1, seq_to_diag, time_lastfu),\n",
    "#         time_anyblood = ifelse(post_anyblood == 1, seq_to_diag, time_lastfu),\n",
    "#         time_leukemia = ifelse(post_leukemia == 1, seq_to_diag, time_lastfu)\n",
    "#     ) %>%\n",
    "#     ungroup() %>%\n",
    "#     mutate(\n",
    "#         ch_sm = ch_all\n",
    "#     ) %>%\n",
    "#     mutate(tumor_type = GeneralTumorType)\n",
    "\n",
    "# # clean up smoke\n",
    "M_wide = M_wide %>%\n",
    "#     mutate(smoke = ifelse(is.na(smoke_bin), 'missing', c('1' = 'yes', '0' = 'no')[as.character(smoke_bin)])) %>%\n",
    "    mutate(smoke = factor(smoke, c('no', 'yes', 'missing')))\n",
    "\n",
    "# therapy_detailed = fread('./data/therapy_chemoxrt.tsv')\n",
    "# therapy_detailed = therapy_detailed %>% filter(complete.cases(.))\n",
    "\n",
    "# M_wide = M_wide %>% \n",
    "#     left_join(\n",
    "#             therapy_detailed,\n",
    "#             by = \"dmp_patient_id\"\n",
    "#         ) %>%\n",
    "#         mutate(therapy_detailed = !is.na(XRT))\n",
    "\n",
    "# ### merge with CNV calls\n",
    "# failed_samples = segs %>% filter(is.na(seg)) %>% pull(dmp_sample_id)\n",
    "\n",
    "# M_wide = M_wide %>% filter(!dmp_sample_id %in% failed_samples)\n",
    "\n",
    "# M_wide %>% count(therapy_detailed)\n",
    "\n",
    "# M_long = fread('./data/sm_long_jul13.tsv') %>%\n",
    "#     filter(dmp_patient_id %in% M_wide$dmp_patient_id)\n",
    "\n",
    "M_long = M_long %>% \n",
    "#     rename(chrom = Chrom, start = Start) %>%\n",
    "    mutate(chrom = as.integer(ifelse(chrom == 'X', 23, chrom)))\n",
    "\n",
    "# M_long = M_long %>% distinct(dmp_sample_id, start, Ref, Alt, `.keep_all` = T)\n",
    "\n",
    "# M_long = M_long %>% mutate(\n",
    "#         VariantClass2 = case_when(\n",
    "#             VariantClass %in% c('Translation_Start_Site', \"5'Flank\", \"3'Flank\", \"3'UTR\", \"5'UTR\") ~ 'Regulatory',\n",
    "#             VariantClass %in% c('Missense_Mutation', 'In_Frame_Del', 'In_Frame_Ins') ~ 'Missense',\n",
    "#             VariantClass %in% c('Nonsense_Mutation', 'Nonstop_Mutation', 'Splice_Site', 'Splice_Region', 'Frame_Shift_Del', 'Frame_Shift_Ins') ~ 'Truncating',\n",
    "#             T ~ VariantClass)\n",
    "#     )\n",
    "\n",
    "# segs = segs %>% \n",
    "#     filter(!is.na(seg)) %>%\n",
    "#     inner_join(M_wide) %>%\n",
    "#     mutate(active_heme = ifelse(is.na(active_heme), T, active_heme))\n",
    "\n",
    "# segs = segs %>% mutate(\n",
    "#     z_x = ifelse(is.na(z_x), t_cnlr, z_x),\n",
    "#     z_y = ifelse(is.na(z_y), t_valor, z_y),\n",
    "#     p_x = ifelse(is.na(p_x), p_cnlr, p_x),\n",
    "#     p_y = ifelse(is.na(p_y), p_valor, p_y)\n",
    "# )\n",
    "\n",
    "# segs_filtered = segs %>%\n",
    "#     ungroup() %>%\n",
    "#     filter((p_chisq < 1e-10 & p_y < 1e-3)) %>%\n",
    "#     filter(type != 'err' & err < 0.06) %>%\n",
    "#     # germline amplifications\n",
    "#     filter(!(phi >= 0.8 & type %in% c('amp'))) %>%\n",
    "#     filter(!(type == 'amp' & size < 24e6)) %>% \n",
    "#     filter(cnlr <= 0.5) %>%\n",
    "#     filter(!(type == 'loh' & size < 8e6)) %>%\n",
    "#     filter(!(chrom == 23 & Gender == 'Male' & type == 'loh')) %>%\n",
    "#     filter(!(chrom == 23 & type == 'amp')) %>%\n",
    "#     # recurrent focal artifacts\n",
    "#     filter(!(chrom == 2 & type == 'del' & start > 11e7 & end < 13e7)) %>%\n",
    "#     filter(!(chrom == 11 & type == 'del' & start > 60e6 & end < 70e6)) %>%\n",
    "#     filter(!(chrom == 7 & start > 40e6 & end < 70e6))\n",
    "\n",
    "# # circulating tumor DNA\n",
    "# segs_filtered = segs_filtered %>% filter(dmp_patient_id != 'P-0027364')\n",
    "\n",
    "segs_filtered = segs_filtered %>% \n",
    "#     left_join(chrom_lens %>% rename(chrom_length = length), by = 'chrom') %>%\n",
    "#     mutate(prop_chrom = size / chrom_length) %>%\n",
    "#     mutate(focality = ifelse(prop_chrom < 0.15, 'focal', 'broad')) %>%\n",
    "#     mutate(cnv_label = paste0(chrom, c('amp' = '+', 'del' = '-', 'loh' = '=')[type])) %>%\n",
    "    mutate(type2 = factor(c('del' = 'DEL', 'loh' = 'CNLOH', 'amp' = 'AMP')[type], c('DEL', 'CNLOH', 'AMP')))\n",
    "\n",
    "# # annodate locality\n",
    "# segs_filtered = segs_filtered %>% left_join(acen, by = \"chrom\")\n",
    "\n",
    "# segs_filtered = segs_filtered %>% \n",
    "#     rowwise() %>%\n",
    "#     mutate(\n",
    "#         p_length = cen_start,\n",
    "#         q_length = chrom_length - cen_end,\n",
    "#         p_size = max(min(end, cen_start) - max(start, 0), 0),\n",
    "#         q_size = max(min(end, chrom_length) - max(start, cen_end), 0),\n",
    "#         p_frac = p_size/p_length,\n",
    "#         q_frac = q_size/q_length,\n",
    "#         arm = case_when(\n",
    "#             p_frac > 0.5 & q_frac > 0.5 ~ 'p,q',\n",
    "#             p_frac > q_frac ~ 'p',\n",
    "#             T ~ 'q'\n",
    "#         ),\n",
    "#         focality = ifelse(max(p_frac, q_frac) < 0.1, 'focal', 'broad')\n",
    "#     ) %>%\n",
    "#     ungroup()\n",
    "\n",
    "# n_samples = segs_filtered %>% group_by(dmp_patient_id, dmp_sample_id) %>% summarise(n_samples = length(unique(dmp_sample_id))) %>% pull(n_samples) %>% max\n",
    "\n",
    "# display(n_samples)\n",
    "\n",
    "# display(nrow(segs_filtered))\n",
    "\n",
    "# # gemrline calls\n",
    "# M_germline = fread('./data/germline_muts.tsv')\n",
    "# M_germline = M_germline %>% filter(dmp_patient_id %in% M_wide$dmp_patient_id)\n",
    "\n",
    "# M_germline_wide = M_germline %>% \n",
    "#     mutate(ind_germline = 1) %>%\n",
    "#     reshape2::dcast(\n",
    "#         dmp_patient_id ~ Gene,\n",
    "#         value.var = 'ind_germline',\n",
    "#         fun.aggregate = max,\n",
    "#         fill = 0\n",
    "#     ) %>%\n",
    "#     setNames(c('dmp_patient_id', paste0(colnames(.)[-1], '_g'))) %>%\n",
    "#     mutate(any_germline = 1)\n",
    "\n",
    "# M_wide = M_wide %>% left_join(\n",
    "#         M_germline_wide,\n",
    "#         by = 'dmp_patient_id'\n",
    "#     ) %>%\n",
    "#     mutate_at(colnames(M_germline_wide), function(x){ifelse(is.na(x), 0, x)})\n",
    "\n",
    "# segs_filtered = segs_filtered %>% \n",
    "#     left_join(M_germline_wide, by = \"dmp_patient_id\")\n",
    "\n",
    "# merge with GM data\n",
    "CNV = segs_filtered %>% \n",
    "    mutate(ch_cnv = 1) %>%\n",
    "    group_by(dmp_patient_id, chrom) %>%\n",
    "    summarise(ch_cnv = max(ch_cnv)) %>%\n",
    "    arrange(chrom) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(chrom = paste0('chr', chrom)) %>%\n",
    "    mutate(chrom = factor(chrom, unique(chrom))) %>%\n",
    "    reshape2::dcast(\n",
    "        dmp_patient_id ~ chrom,\n",
    "        value.var = 'ch_cnv',\n",
    "        fill = 0\n",
    "    ) %>%\n",
    "    mutate(ch_cnv = 1)\n",
    "\n",
    "CNV_detailed = segs_filtered %>% \n",
    "    mutate(ch_cnv = 1) %>%\n",
    "    group_by(dmp_patient_id, chrom, type) %>%\n",
    "    summarise(ch_cnv = max(ch_cnv)) %>%\n",
    "    arrange(chrom) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(chrom = paste0('chr', chrom)) %>%\n",
    "    mutate(chrom = factor(chrom, unique(chrom))) %>%\n",
    "    reshape2::dcast(\n",
    "        dmp_patient_id ~ chrom + type,\n",
    "        value.var = 'ch_cnv',\n",
    "        fill = 0\n",
    "    )\n",
    "\n",
    "CNV_detailed_arm = segs_filtered %>% \n",
    "    mutate(ch_cnv = 1) %>%\n",
    "    group_by(dmp_patient_id, chrom, type, arm) %>%\n",
    "    summarise(ch_cnv = max(ch_cnv)) %>%\n",
    "    arrange(chrom) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(chrom = paste0('chr', chrom)) %>%\n",
    "    mutate(chrom = factor(chrom, unique(chrom))) %>%\n",
    "    mutate(arm = str_remove(arm, ',')) %>%\n",
    "    reshape2::dcast(\n",
    "        dmp_patient_id ~ chrom + arm + type,\n",
    "        value.var = 'ch_cnv',\n",
    "        fill = 0\n",
    "    )\n",
    "\n",
    "CNV_types = segs_filtered %>% \n",
    "    mutate(ch_cnv = 1) %>%\n",
    "    mutate(type = paste0('ch_', type)) %>%\n",
    "    group_by(dmp_patient_id, type) %>%\n",
    "    summarise(ch_cnv = max(ch_cnv)) %>%\n",
    "    ungroup() %>%\n",
    "    reshape2::dcast(\n",
    "        dmp_patient_id ~ type,\n",
    "        value.var = 'ch_cnv',\n",
    "        fill = 0\n",
    "    )\n",
    "\n",
    "segs_filtered_auto = segs_filtered %>% filter(chrom != 23)\n",
    "\n",
    "# window = 5e6\n",
    "\n",
    "# max0 = function(x) {\n",
    "#   if (length(x) == 0) {\n",
    "#     return(0)\n",
    "#   } else {\n",
    "#     return(max(x))\n",
    "#   }\n",
    "# }\n",
    "\n",
    "# MAXVAF_trans = M_long %>%\n",
    "#     filter(ch_nonsilent == 1) %>%\n",
    "#     rowwise() %>%\n",
    "#     mutate(\n",
    "#         locality = ifelse(\n",
    "#             any(chrom == segs_filtered_auto$chrom & dmp_patient_id == segs_filtered_auto$dmp_patient_id &\n",
    "#                 start >= segs_filtered_auto$start - window & start <= segs_filtered_auto$end + window\n",
    "#             ),\n",
    "#             'cis',\n",
    "#             'trans'\n",
    "#         )\n",
    "#     ) %>%\n",
    "#     ungroup() %>%\n",
    "#     filter(locality == 'trans') %>%\n",
    "#     group_by(dmp_patient_id) %>%\n",
    "#     summarise(\n",
    "#         VAF_trans = max(VAF_N),\n",
    "#         mutnum_trans = sum(ch_nonsilent),\n",
    "#         mutnum_pd_trans = sum(ch_nonsilent & ch_pd),\n",
    "#         VAF_pd_trans = max0(VAF_N[ch_pd == 1])\n",
    "#     )\n",
    "\n",
    "# chroms = colnames(CNV)[-1]\n",
    "# chroms = chroms[chroms != 'ch_cnv']\n",
    "# genes = c('DNMT3A', 'TET2', 'ASXL1', 'JAK2', 'TP53', 'PPM1D', 'ATM', 'EZH2')\n",
    "\n",
    "# M_wide = M_wide %>%\n",
    "#     left_join(CNV, by = \"dmp_patient_id\") %>%\n",
    "#     left_join(CNV_detailed, by = \"dmp_patient_id\") %>%\n",
    "#     left_join(CNV_types, by = \"dmp_patient_id\") %>%\n",
    "#     left_join(CNV_detailed_arm, by = \"dmp_patient_id\") %>%\n",
    "#     left_join(MAXVAF_trans, by = 'dmp_patient_id') %>%\n",
    "#     left_join(\n",
    "#         segs_filtered_auto %>%\n",
    "#             group_by(dmp_patient_id) %>%\n",
    "#             summarise(\n",
    "#                 phi_cnv = max(phi),\n",
    "#                 phi_cnv_auto = max(phi[chrom != 23]),\n",
    "#                 n_cnv = n(),\n",
    "#                 n_cnv_chrom = length(unique(chrom)),\n",
    "#                 ch_cnv_auto = 1,\n",
    "#                 .groups = 'drop'),\n",
    "#         by = 'dmp_patient_id'\n",
    "#     ) %>%\n",
    "#     mutate_at(colnames(CNV), function(x){ifelse(is.na(x), 0, x)}) %>%\n",
    "#     mutate_at(colnames(CNV_detailed), function(x){ifelse(is.na(x), 0, x)}) %>%\n",
    "#     mutate_at(colnames(CNV_detailed_arm), function(x){ifelse(is.na(x), 0, x)}) %>%\n",
    "#     mutate_at(colnames(CNV_types), function(x){ifelse(is.na(x), 0, x)}) %>%\n",
    "#     mutate_at(\n",
    "#         c('phi_cnv', 'VAF_trans', 'VAF_pd_trans', 'n_cnv', 'n_cnv_chrom', 'ch_cnv_auto', 'mutnum_trans', 'mutnum_pd_trans', 'phi_cnv_auto', 'phi_cnv'),\n",
    "#         function(x){ifelse(is.na(x), 0, x)}\n",
    "#     )\n",
    "\n",
    "M_long = M_long %>%\n",
    "#     left_join(CNV, by = \"dmp_patient_id\") %>%\n",
    "#     left_join(CNV_detailed, by = \"dmp_patient_id\") %>%\n",
    "#     left_join(CNV_types, by = \"dmp_patient_id\") %>%\n",
    "#     left_join(CNV_detailed_arm, by = \"dmp_patient_id\") %>%\n",
    "#     mutate_at(colnames(CNV), function(x){ifelse(is.na(x), 0, x)}) %>%\n",
    "#     mutate_at(colnames(CNV_detailed), function(x){ifelse(is.na(x), 0, x)}) %>%\n",
    "#     mutate_at(colnames(CNV_types), function(x){ifelse(is.na(x), 0, x)}) %>%\n",
    "    mutate(ch_cnv_auto = as.integer(ch_cnv & (!chr23)))\n",
    "\n",
    "# M_wide = M_wide %>%\n",
    "#     mutate(age_bin = cut(age, c(seq(20, 80, 10), Inf))) %>%\n",
    "#     mutate(\n",
    "#         age10 = age/10,\n",
    "#         age10_sq = age10^2\n",
    "#     )\n",
    "\n",
    "# M_wide = M_wide %>%\n",
    "#     mutate(\n",
    "#         Race = case_when(\n",
    "#             Race == 'ASIAN-FAR EAST/INDIAN SUBCONT' ~ 'asian',\n",
    "#             Race == 'BLACK OR AFRICAN AMERICAN' ~ 'black',\n",
    "#             Race == 'WHITE' ~ 'white',\n",
    "#             T ~ 'other'\n",
    "#         )\n",
    "#     )\n",
    "\n",
    "# # loss of Y\n",
    "# segs_y = fread('./data/y_segs.tsv', sep = '\\t')\n",
    "# LOY = segs_y %>% filter(p_x < 1e-17 & cnlr < 0)\n",
    "# M_wide = M_wide %>%\n",
    "#     mutate(loy = ifelse(Gender == 'Male', dmp_sample_id %in% LOY$dmp_sample_id, NA)) %>% \n",
    "#     left_join(LOY %>% mutate(loy_logr = cnlr)) %>%\n",
    "#     mutate(loy_dose = ifelse(loy == 1, -exp(loy_logr), 0)) %>%\n",
    "#     mutate(loy_bin = as.character(ifelse(loy == 1, cut(loy_dose, 3), 0)))\n",
    "\n",
    "# cbcs\n",
    "cbcs = c(\"anc\", \"alc\", \"amc\", \"hgb\", \"mcv\", \"rdw\", \"plt\", \"wbc\", \"rbc\")\n",
    "# M_wide = M_wide %>% mutate(rdw = rdw * mcv)\n",
    "\n",
    "# M_wide = M_wide %>%\n",
    "#     mutate_at(cbcs, .funs = list(n = function(x){(x - mean(na.omit(x))) / sd(na.omit(x))}))\n",
    "\n",
    "# nrow(M_wide) %>% paste(' patients valid for analysis!')\n",
    "\n",
    "cnv_pal = c('loh' = 'darkgreen', 'amp' = 'darkblue', 'del' = 'darkred', 'err' = 'gray')\n",
    "cnv_pal2 = c('CNLOH' = 'darkgreen', 'AMP' = 'darkblue', 'DEL' = 'darkred')\n",
    "sm_pal = c('Regulatory' = 'darkturquoise', 'Missense' = 'salmon', 'Truncating' = 'purple')\n",
    "\n",
    "gene_structures = fread('./data/gene_structures.tsv', sep = '\\t') %>% filter(class == 'protein_coding')\n",
    "\n",
    "sm_col = '#FDCDAC'\n",
    "cnv_col = '#B3E2CD'\n",
    "\n",
    "cbc_labels = c('anc' = 'Neutrophil', 'alc' = 'Lymphocyte', 'amc' = 'Monocyte',\n",
    "  'hgb' = 'Hemoglobin', 'mcv' = 'Mean corpuscular volume', \n",
    "  'rdw' = 'Red cell distribution width', 'plt' = 'Platelets',\n",
    "  'wbc' = 'White blood cell', 'rbc' = 'Red blood cell')\n",
    "\n",
    "do_plot = function(p, f, w, h) {\n",
    "    ggsave(filename = paste0('/work/isabl/home/sirenkom/teng_mCA/figures/', f), plot = p, width = w, height = h, device = 'pdf')\n",
    "    options(repr.plot.width = w, repr.plot.height = h, repr.plot.res = 300)\n",
    "    print(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for getting gene structures\n",
    "# gtf = fread('/work/isabl/ref/homo_sapiens/GRCh37d5/ensembl/75/Homo_sapiens.GRCh37.75.gtf', skip = 'gene_id')\n",
    "# colnames(gtf) = c('chrom', 'class', 'structure_type', 'start', 'end', 'x', 'y', 'z', 'description')\n",
    "# gene_structures = gtf %>% filter(str_detect(description, paste(paste0('\"', gene_bed$Gene, '\"'), collapse = '|')) & structure_type == 'exon')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cohort characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source('./ch-cnv/notebooks/table_one.R')\n",
    "\n",
    "\n",
    "\n",
    "options(warn = -1)\n",
    "table1 = tab_response(M_wide, xs = c('total', 'age', 'Gender', 'Race', 'smoke', 'months_followup', cbcs), y = 'ch_cnv')\n",
    "options(warn = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kb = table1 %>% \n",
    "    mutate(overall = ifelse(variable %in% c('Gender', 'Race', 'smoke'), ' ', overall)) %>%\n",
    "    rename(Overall = overall) %>%\n",
    "    mutate(variable = ifelse(variable == 'smoke', 'Smoking history', variable)) %>%\n",
    "    mutate(variable = ifelse(variable %in% names(cbc_labels), cbc_labels[variable], variable)) %>%\n",
    "    mutate(variable = str_replace(variable, '-', ' ')) %>%\n",
    "    mutate(levels = ifelse(levels == 'total', ' ', levels)) %>%\n",
    "    mutate(variable = ifelse(variable == 'total', 'total subjects', variable)) %>%\n",
    "    mutate(levels = Hmisc::capitalize(levels)) %>%\n",
    "    mutate(variable = Hmisc::capitalize(variable)) %>%\n",
    "    rename(\n",
    "        'mCA-' = `0`, 'mCA+' = `1`,\n",
    "        Characteristics = variable,\n",
    "        ' ' = levels\n",
    "    ) %>%\n",
    "    kable(format = \"html\", booktabs = T) %>%\n",
    "    collapse_rows(columns = 1:2, row_group_label_position = 'stack') %>%\n",
    "    row_spec(0:nrow(table1), background = 'white', align = 'center', extra_css = \"border-bottom: 1px solid\") %>%\n",
    "    row_spec(0, background = 'gray', extra_css = \"border-bottom: 2px solid\") %>%\n",
    "    kable_styling(position = \"center\", full_width = F)\n",
    "\n",
    "kb %>%\n",
    "toString() %>%\n",
    "display_html()\n",
    "\n",
    "save_kable(kb, './figures/cohort.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": false
   },
   "source": [
    "# collect all segs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# segs = P$dmp_sample_id %>%\n",
    "#     lapply(function(sid){\n",
    "#         outdir = '/work/isabl/home/gaot/ch_cnv/results'\n",
    "#         segfile = glue('{outdir}/{sid}_seg.tsv')\n",
    "#         if (file.exists(segfile)) {\n",
    "#             seg = fread(segfile)\n",
    "#             noisy = nrow(seg) > 35\n",
    "#             if (noisy) {\n",
    "#                 data.frame()\n",
    "#             } else {\n",
    "#                 seg %>% mutate(sample = sid) %>% filter(aberrant)\n",
    "#             }\n",
    "#         } else {\n",
    "#             data.frame()\n",
    "#         }\n",
    "#     }) %>% Reduce(rbind, .)\n",
    "\n",
    "# segs = segs %>% left_join(P, by = c('sample' = 'dmp_sample_id'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# outdir = '/work/isabl/home/gaot/ch_cnv/results_jul2'\n",
    "\n",
    "# segs = M_wide$dmp_sample_id %>%\n",
    "#     lapply(function(sid){\n",
    "#         segfile = glue('{outdir}/{sid}_seg.tsv')\n",
    "#         if (file.exists(segfile)) {\n",
    "#             seg = fread(segfile)\n",
    "#             seg %>% mutate(dmp_sample_id = sid) %>% filter(aberrant)\n",
    "#         } else {\n",
    "#             data.frame(dmp_sample_id = c(sid)) %>% mutate(dmp_sample_id = as.character(dmp_sample_id))\n",
    "#         }\n",
    "#     }) %>% Reduce(bind_rows, .)\n",
    "\n",
    "# print('done!')\n",
    "\n",
    "# segs %>% fwrite('./data/segs_aberrant_jul2.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# outdir = '/work/isabl/public/facets-ch/dmp_processed'\n",
    "\n",
    "# segs = M_wide %>%\n",
    "#     filter(nonparta == 1) %>%\n",
    "#     pull(dmp_sample_id) %>%\n",
    "#     lapply(function(sid){\n",
    "#         segfile = glue('{outdir}/{sid}_seg.tsv')\n",
    "#         if (file.exists(segfile)) {\n",
    "#             seg = fread(segfile)\n",
    "#             seg %>% mutate(dmp_sample_id = sid) %>% filter(aberrant)\n",
    "#         } else {\n",
    "#             data.frame(dmp_sample_id = c(sid)) %>% mutate(dmp_sample_id = as.character(dmp_sample_id))\n",
    "#         }\n",
    "#     }) %>% Reduce(bind_rows, .)\n",
    "\n",
    "# print('done!')\n",
    "\n",
    "# # correct for phi and variant type\n",
    "# segs[c('type', 'phi', 'err')] = t(mapply(variant_type, segs$cnlr_adj, segs$valor, segs$aberrant))\n",
    "\n",
    "# segs %>% fwrite('./data/segs_aberrant_nonparta.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# General CNV characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segs_filtered %>% nrow %>% paste('unique events')\n",
    "segs_filtered %>% count(auto = chrom != 23)\n",
    "segs_filtered %>% pull(dmp_patient_id) %>% unique %>% length %>% paste('unique individuals with CNV')\n",
    "\n",
    "segs_filtered %>% \n",
    "filter(chrom != 23) %>%\n",
    "count(type) %>% mutate(prop = signif(n/sum(n), 2))\n",
    "\n",
    "segs_filtered %>%\n",
    "filter(chrom != 23) %>%\n",
    "count(dmp_patient_id) %>% count(n) %>% mutate(prop = signif(nn/sum(nn), 2))\n",
    "\n",
    "cat('Cell fraction ranges:', segs_filtered$phi %>% range %>% signif(2) %>% paste(collapse = '-') %>% paste('\\n'))\n",
    "cat('median cell fraction:', segs_filtered$phi %>% median %>% signif(2) %>% paste('\\n'))\n",
    "cat('median cell fraction:', segs_filtered$phi %>% median %>% signif(2) %>% paste('\\n'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cancer types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "signif.num <- function(x, ns = FALSE, pch = TRUE) {\n",
    "    if (ns) {\n",
    "        symbols = c(\"***\", \"**\", \"*\", \".\", \"ns\")\n",
    "    } else {\n",
    "        symbols = c(\"***\", \"**\", \"*\", \".\", \"\")\n",
    "    } \n",
    "    \n",
    "    if (pch) {\n",
    "        symbols = c(8, 8, 3, 20)\n",
    "    }\n",
    "    \n",
    "    symnum(unlist(x), corr = FALSE, na = FALSE, legend = FALSE,\n",
    "           cutpoints = c(0, 0.001, 0.01, 0.05, 0.1, 1), \n",
    "           symbols = symbols)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit = glm(data = M_wide, formula = ch_cnv_auto ~ age + Gender + race_b, family = 'binomial')\n",
    "\n",
    "D = M_wide %>%\n",
    "    group_by(tumor_type) %>%\n",
    "    summarise(\n",
    "        cnv_freq = sum(ch_cnv_auto)/n(),\n",
    "        cnv_count = sum(ch_cnv_auto),\n",
    "        sm_freq = sum(ch_nonsilent)/n(),\n",
    "        n_male = sum(Gender == 'Male'),\n",
    "        loy_freq = sum(loy[Gender == 'Male'])/n_male,\n",
    "        n = n(),\n",
    "        age_median = median(age),\n",
    "        cnv_freq_pred = sum(predict(fit, data.frame(age = age, Gender = Gender, race_b = race_b), type=\"response\"))/n()\n",
    "    ) %>%\n",
    "    mutate(\n",
    "        cnv_freq_lower = qbinom(p = 0.05, size = n, prob = cnv_freq)/n,\n",
    "        cnv_freq_upper = qbinom(p = 0.95, size = n, prob = cnv_freq)/n,\n",
    "        sm_freq_lower = qbinom(p = 0.05, size = n, prob = sm_freq)/n,\n",
    "        sm_freq_upper = qbinom(p = 0.95, size = n, prob = sm_freq)/n,\n",
    "        loy_freq_lower = qbinom(p = 0.05, size = n, prob = loy_freq)/n_male,\n",
    "        loy_freq_upper = qbinom(p = 0.95, size = n, prob = loy_freq)/n_male\n",
    "    ) %>%\n",
    "    filter(n > 300) %>%\n",
    "    arrange(cnv_freq_pred) %>%\n",
    "#   mutate(tumor_type = paste0(tumor_type, '(', n, ')')) %>%\n",
    "    mutate(tumor_type = factor(tumor_type, unique(tumor_type)))\n",
    "    \n",
    "D = D %>% \n",
    "    rowwise() %>%\n",
    "    mutate(p = prop.test(x = cnv_count, n = n, p = cnv_freq_pred)$p.value) %>%\n",
    "    mutate(\n",
    "        q = p.adjust(p, method = 'BH'),\n",
    "        q.stars = gtools::stars.pval(q)\n",
    "    ) %>%\n",
    "    ungroup()\n",
    "\n",
    "coef = mean(D$sm_freq)/mean(D$cnv_freq)\n",
    "\n",
    "p1 = D %>% \n",
    "    ggplot(\n",
    "        aes()\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    geom_point(\n",
    "        inherit.aes = F,\n",
    "        aes(x = tumor_type,\n",
    "            y = cnv_freq_pred\n",
    "        ),\n",
    "        color = 'royalblue',\n",
    "        pch = 21,\n",
    "        size = 3\n",
    "    ) +\n",
    "    geom_pointrange(\n",
    "        inherit.aes = F,\n",
    "        aes(x = tumor_type,\n",
    "            y = cnv_freq, \n",
    "            ymin = cnv_freq_lower, \n",
    "            ymax = cnv_freq_upper,\n",
    "#             color = 'CNV-CH'\n",
    "        ),\n",
    "        stat = 'identity',\n",
    "        pch = 5\n",
    "    ) +\n",
    "#     geom_text(\n",
    "#         aes(label = q.stars, x = tumor_type, y = 0.03),\n",
    "# #         hjust = 1.5,\n",
    "# #         vjust = -0.2,\n",
    "#         size = 5\n",
    "#     ) +\n",
    "    theme(\n",
    "        axis.text.x = element_blank(),\n",
    "        plot.margin = unit(c(0,0,0,0), 'mm')\n",
    "#         axis.title.x = element_blank()\n",
    "    ) +\n",
    "    xlab('') +\n",
    "    ylab('Proportion with mCA')\n",
    "\n",
    "give.n <- function(x){\n",
    "  return(c(y = 100, label = length(x))) \n",
    "}\n",
    "\n",
    "p2 = M_wide %>% filter(tumor_type %in% D$tumor_type) %>%\n",
    "    mutate(tumor_type = factor(tumor_type, levels(D$tumor_type))) %>%\n",
    "    ggplot(aes(x = tumor_type, y = age)) +\n",
    "    theme_classic() +\n",
    "    geom_violin(draw_quantiles = c(0.5)) +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 30, hjust = 1, size = 7),\n",
    "        plot.margin = unit(c(0,0,0,0), 'mm')\n",
    "    ) +\n",
    "    xlab('') +\n",
    "    stat_summary(fun.data = give.n, geom = \"text\", vjust = 0, size = 2) +\n",
    "    scale_y_continuous(expand = expansion(add = 10)) +\n",
    "    ylab('Age')\n",
    "\n",
    "panel = p1/p2 + plot_layout(heights = c(2,1))\n",
    "\n",
    "do_plot(panel, 'cancer_types.pdf', w = 5, h = 4)\n",
    "\n",
    "\n",
    "D %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/Fig_1F_top.tsv\", sep= \"\\t\", quote = FALSE)\n",
    "M_wide %>% filter(tumor_type %in% D$tumor_type) %>%\n",
    "    mutate(tumor_type = factor(tumor_type, levels(D$tumor_type))) %>% select(dmp_patient_id, age, tumor_type) %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/Fig_1F_bottom.tsv\", sep= \"\\t\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Co-mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table(M_wide$DNMT3A, M_wide$ch_cnv_auto)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table(M_wide$DNMT3A, M_wide$ch_cnv_auto) %>% fisher.test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 4, repr.plot.height = 2, repr.plot.res = 400)\n",
    "\n",
    "sm_classes = M_long %>% pull(VariantClass2) %>% unique\n",
    "\n",
    "colors = c(cnv_pal2, sm_pal) \n",
    "\n",
    "heme_colors = c(' ' = 'white', 'CLL' = 'royalblue', 'MDS' = 'tomato3', 'MPN' = 'purple', 'Lymphoma' = 'yellowgreen')\n",
    "\n",
    "colors = c(colors, heme_colors)\n",
    "\n",
    "gene_set = c(fread('./data/gene_set.tsv')$Gene, 'PPM1D', 'SRSF2')\n",
    "\n",
    "# basically the same\n",
    "gene_set = M_long %>% \n",
    "    filter(ch_nonsilent == 1) %>%\n",
    "    filter(Gene %in% gene_set) %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    count(Gene) %>% filter(n >= 2) %>% pull(Gene)\n",
    "\n",
    "# gene_set = c('DNMT3A', 'TET2', 'JAK2', 'ASXL1', 'TP53', 'ATM', 'PPM1D', 'CHEK2',\n",
    "#              'SRSF2', 'SF3B1', 'U2AF1', 'EZH2', 'MPL', 'NOTCH1', 'TERT', 'SUZ12', 'CBL')\n",
    "\n",
    "chroms = segs_filtered %>% count(chrom) %>%\n",
    "#     filter(n >= 4) %>% \n",
    "    pull(chrom)\n",
    "# max_rows = 15\n",
    "max_rows = 22\n",
    "\n",
    "plots = list()\n",
    "\n",
    "for (this_chrom in chroms) {\n",
    "    \n",
    "    chrom_label = ifelse(this_chrom == 23, 'X', this_chrom)\n",
    "\n",
    "    segs_auto = segs_filtered %>% filter(chrom == this_chrom)\n",
    "\n",
    "    M_co = rbind(\n",
    "            M_long %>% \n",
    "            filter(Gene %in% gene_set) %>%\n",
    "            filter(ch_nonsilent == 1) %>%\n",
    "            filter(dmp_sample_id %in% segs_auto$dmp_sample_id) %>% \n",
    "            mutate(\n",
    "                mutation = Gene,\n",
    "                type = VariantClass2,\n",
    "                arm = 'na',\n",
    "                focality = 'na'\n",
    "            ) %>%\n",
    "            select(mutation, type, arm, focality, dmp_patient_id),\n",
    "            segs_auto %>% \n",
    "            mutate(\n",
    "                mutation = paste0(arm, type, focality)\n",
    "            ) %>%\n",
    "            select(mutation, type = type2, arm, focality, dmp_patient_id)\n",
    "        ) %>%\n",
    "        group_by(mutation) %>%\n",
    "        mutate(mutation_n = n()) %>%\n",
    "        ungroup()\n",
    "\n",
    "    D = M_co %>%\n",
    "        distinct(mutation, dmp_patient_id, `.keep_all` = T) %>%\n",
    "        group_by(mutation) %>% mutate(mutation_n = n()) %>% ungroup() %>%\n",
    "        arrange(type %in% sm_classes, desc(mutation_n)) %>%\n",
    "        mutate(mutation = factor(mutation, rev(unique(mutation)))) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(mut_weight = 2^which(mutation == levels(mutation))) %>%\n",
    "        ungroup() %>%\n",
    "        group_by(dmp_patient_id) %>% \n",
    "        mutate(patient_weight = sum(mut_weight)) %>%\n",
    "        ungroup() %>%\n",
    "        arrange(desc(patient_weight), type) %>%\n",
    "        mutate(pid = as.integer(factor(dmp_patient_id, unique(dmp_patient_id))))\n",
    "    \n",
    "    D = D %>% arrange(desc(mutation)) %>%\n",
    "        mutate(mutation = as.character(mutation)) %>%\n",
    "        mutate(mutation = str_remove_all(mutation, 'amp|del|loh|broad|focal')) %>%\n",
    "        mutate(mutation = factor(mutation, rev(unique(mutation))))\n",
    "    \n",
    "    n_rows = length(levels(D$mutation))\n",
    "    n_cols = length(unique(D$pid))\n",
    "    \n",
    "    D = D %>% mutate(mutation = factor(mutation, c(as.character(1:(max_rows - n_rows)), levels(mutation))))\n",
    "    nrows_sm = D %>% filter(type %in% sm_classes) %>% pull(mutation) %>% unique %>% length\n",
    "\n",
    "    p = ggplot(\n",
    "        D,\n",
    "        aes(x = pid, y = mutation, fill = type)\n",
    "    ) +\n",
    "    geom_rect(\n",
    "        xmin = 0, xmax = n_cols + 0.5, ymin = max_rows - n_rows + nrows_sm + 0.5, ymax = max_rows + 0.5,\n",
    "        fill = 'whitesmoke',\n",
    "    ) +\n",
    "    geom_segment(\n",
    "        aes(x = pid + 0.5, xend = pid + 0.5),\n",
    "        y = max_rows - n_rows + 0.5,\n",
    "        yend = max_rows + 0.5,\n",
    "        color = 'gray90',\n",
    "        size = 0.3\n",
    "    ) +\n",
    "    geom_hline(yintercept = max_rows - n_rows + nrows_sm + 0.5, size = 0.3, color = 'gray90') +\n",
    "#     geom_hline(yintercept = max_rows + 0.5, size = 0.5, color = 'black') +\n",
    "#     geom_hline(yintercept = max_rows - n_rows + 0.5, size = 0.5, color = 'black') +\n",
    "    geom_segment(x = 0.5, xend = 0.5, y = max_rows - n_rows + 0.5, yend = max_rows + 0.5) +\n",
    "    geom_rect(\n",
    "        aes(\n",
    "            xmin = pid - 0.4,\n",
    "            xmax = pid + 0.4, \n",
    "            ymin = case_when(\n",
    "                focality == 'focal' ~ as.numeric(mutation) - 0.1,\n",
    "                arm == 'p' ~ as.numeric(mutation),\n",
    "                type %in% sm_classes ~ as.numeric(mutation) - 0.45,\n",
    "                T ~ as.numeric(mutation) - 0.45\n",
    "            ),\n",
    "            ymax = case_when(\n",
    "                focality == 'focal' ~ as.numeric(mutation) + 0.1,\n",
    "                arm == 'p' ~ as.numeric(mutation) + 0.45,\n",
    "                arm == 'q' ~ as.numeric(mutation),\n",
    "                arm == 'p,q' ~ as.numeric(mutation) + 0.45,\n",
    "                type %in% sm_classes ~ as.numeric(mutation) + 0.45\n",
    "            )\n",
    "        ),\n",
    "        size = 0, colour = \"white\"\n",
    "    ) +\n",
    "    # centromeres\n",
    "    geom_point(\n",
    "        data = D %>% filter((!type %in% sm_classes) & focality != 'focal'),\n",
    "        aes(color = type),\n",
    "        size = 1,\n",
    "        show.legend = FALSE\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        axis.text.x = element_blank(),\n",
    "        plot.margin = margin(t = 0, r = 1, b = 0, l = 0, unit = \"mm\"),\n",
    "#         panel.spacing = unit(1,'cm'),\n",
    "        axis.line = element_blank(),\n",
    "        axis.ticks = element_blank(),\n",
    "        legend.box = \"horizontal\",\n",
    "        plot.title = element_text(size = 10)\n",
    "    ) +\n",
    "    guides(\n",
    "        fill = guide_legend(ncol = 2, title = 'Variant Types'),\n",
    "        color = FALSE\n",
    "    ) +\n",
    "    scale_fill_manual(\n",
    "        values = colors\n",
    "    ) +\n",
    "    scale_color_manual(\n",
    "        values = colors\n",
    "    ) +\n",
    "    scale_x_discrete(\n",
    "        expand = expansion(add = 0.2)\n",
    "    ) +\n",
    "    scale_y_discrete(\n",
    "        expand = expansion(add = 0.2),\n",
    "        labels = str_remove_all(levels(D$mutation), 'broad|focal|del|del|loh|amp|^\\\\d+$'),\n",
    "        drop = F\n",
    "    ) +\n",
    "    ggtitle(chrom_label) +\n",
    "    ylab('') +\n",
    "    xlab('')\n",
    "    \n",
    "    if (this_chrom != 14) {\n",
    "        p = p + theme(legend.position = 'none')\n",
    "    } else {\n",
    "        p = p + theme(legend.position = 'right')\n",
    "    }\n",
    "    \n",
    "    # patient heme status\n",
    "#     D_heme = D %>% distinct(dmp_patient_id, pid) %>% \n",
    "#         left_join(\n",
    "#             M_wide %>% select(dmp_patient_id, heme_cat),\n",
    "#             by = 'dmp_patient_id'\n",
    "#         ) %>% \n",
    "#         mutate(\n",
    "#             heme_cat = ifelse(heme_cat %in% c('DLBCL', 'FL', 'MZL'), 'Lymphoma', heme_cat)\n",
    "#         ) %>% filter(heme_cat != '')\n",
    "        \n",
    "#     if (nrow(D_heme) != 0) {\n",
    "#         p = p + geom_point(\n",
    "#             inherit.aes = F,\n",
    "#             data = D_heme,\n",
    "#             mapping = aes(x = pid, fill = heme_cat),\n",
    "#             y = max_rows - n_rows,\n",
    "#             pch = 24, size = 2,\n",
    "#             colour = \"white\"\n",
    "#         )\n",
    "        \n",
    "#     }\n",
    "\n",
    "    plots[[this_chrom]] = p\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plots = plots[lengths(plots) != 0]\n",
    "\n",
    "plot_widths = segs_filtered %>% \n",
    "    distinct(chrom, dmp_patient_id) %>%\n",
    "    filter(chrom %in% chroms) %>% count(chrom) %>% pull(n)\n",
    "\n",
    "panel = wrap_plots(plotlist = plots, widths = plot_widths, nrow = 1, guides = 'collect')\n",
    "\n",
    "do_plot(panel, 'oncoplot.pdf', w = 40, h = 3.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% count(smoke)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## trans effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes = c('DNMT3A', 'TET2', 'ASXL1', 'EZH2', 'TP53', 'ATM', 'CHEK2', 'PPM1D',\n",
    "#         'SRSF2', 'SF3B1', 'U2AF1', 'MPL', 'JAK2', 'CBL', 'RUNX1', 'NOTCH1', 'MYD88', 'TERT', 'SUZ12', 'BRCA2', 'KIT', 'IKZF1')\n",
    "\n",
    "genes = c('DNMT3A', 'TET2', 'ASXL1', 'EZH2', 'TP53', 'ATM', 'CHEK2', 'PPM1D',\n",
    "        'SRSF2', 'SF3B1', 'U2AF1', 'MPL', 'JAK2', 'TERT')\n",
    "\n",
    "res = data.frame()\n",
    "\n",
    "# don't test for all if chrom has predominately one kind of change\n",
    "single_type_chroms = segs_filtered %>% \n",
    "    count(chrom, type) %>% \n",
    "    group_by(chrom) %>%\n",
    "    mutate(total = sum(n)) %>%\n",
    "    mutate(prop = n/total) %>% \n",
    "    filter(prop > 0.7) %>% pull(chrom) %>%\n",
    "    paste0('chr', .)\n",
    "\n",
    "chroms_detailed = colnames(CNV_detailed)[-1][(CNV_detailed[,-1] %>% colSums) > 3]\n",
    "chroms_broad = colnames(CNV)[-1][(CNV[,-1] %>% colSums) > 3]\n",
    "chroms_broad = chroms_broad[chroms_broad != 'ch_cnv']\n",
    "chroms_broad = chroms_broad[!(chroms_broad %in% single_type_chroms)]\n",
    "chroms = c(chroms_broad, chroms_detailed)\n",
    "# chroms = chroms[str_extract(chroms, '\\\\d+') != 23]\n",
    "# chroms = c(chroms, 'loy')\n",
    "\n",
    "ntest = 0\n",
    "\n",
    "for (gene in genes) {\n",
    "    \n",
    "    display(gene)\n",
    "    gene_chrom = gene_bed %>% filter(Gene == gene) %>% pull(chrom)\n",
    "    \n",
    "    for (chrom in chroms) {\n",
    "        \n",
    "#         chroms[str_extract(chroms, '\\\\d+') != gene_chrom]\n",
    "        \n",
    "        ntest = ntest + 1\n",
    "        \n",
    "#         D = M_wide %>% filter(!cytopenia)\n",
    "        D = M_wide\n",
    "        \n",
    "        counts = table(D[[gene]], D[[chrom]])\n",
    "        \n",
    "        if (ncol(counts) != 1) {\n",
    "            \n",
    "            if (chrom == 'chr23_del') {\n",
    "                counts = table(D[D$Gender == 'Female',][[gene]], D[D$Gender == 'Female',][[chrom]])\n",
    "            }\n",
    "\n",
    "            if (chrom == 'loy') {\n",
    "                counts = table(D[D$Gender == 'Male',][[gene]], D[D$Gender == 'Male',][[chrom]])\n",
    "            }\n",
    "            \n",
    "            co = counts[2,2]\n",
    "\n",
    "            if (co >= 2) {\n",
    "\n",
    "                test = fisher.test(counts)\n",
    "                res = rbind(res,\n",
    "                    data.frame(\n",
    "                        gene = gene,\n",
    "                        chrom = chrom,\n",
    "                        p = test$p.value,\n",
    "                        OR = test$estimate,\n",
    "                        co = co\n",
    "                    ))\n",
    "            }\n",
    "            \n",
    "        }\n",
    "        \n",
    "    }\n",
    "}\n",
    "\n",
    "ntest_broad = length(chroms_broad) * length(genes)\n",
    "ntest_specific = (length(chroms) - length(chroms_broad)) * length(genes)\n",
    "\n",
    "res = res %>% \n",
    "    mutate(category = ifelse(str_detect(chrom, '_'), 'broad', 'specific')) %>%\n",
    "    mutate(q = ifelse(\n",
    "        category == 'broad', \n",
    "        p.adjust(p, method = 'BH', n = ntest_broad),\n",
    "        p.adjust(p, method = 'BH', n = ntest_specific)\n",
    "    )) %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "#     mutate(q = p.adjust(p, method = 'BH', n = ntest)) %>%\n",
    "    mutate(logOR = log2(OR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ntest_broad\n",
    "ntest_specific"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "limor = 10\n",
    "width = 0.5\n",
    "star_dodge = 0.3\n",
    "key_width = 0.7\n",
    "\n",
    "ddr = c('TP53', 'PPM1D', 'CHEK2', 'ATM')\n",
    "splicing = c('SRSF2', 'SF3B1', 'U2AF1')\n",
    "dna_methylation = c('TET2', 'DNMT3A')\n",
    "chromatin = c('ASXL1', 'EZH2')\n",
    "\n",
    "genes = unique(res$gene)\n",
    "\n",
    "gene_locations = gene_bed %>% filter(Gene %in% genes)\n",
    "\n",
    "res = res %>% \n",
    "    mutate(signif_cat = case_when(\n",
    "        q < 0.01 ~ 'q<0.01',\n",
    "        q < 0.05 ~ 'q<0.05',\n",
    "        q < 0.1 ~ 'q<0.1',\n",
    "        T ~ 'ns'\n",
    "    ))\n",
    "\n",
    "D = res %>% \n",
    "    rowwise() %>%\n",
    "    mutate(\n",
    "        type = str_split(chrom, '_')[[1]][2],\n",
    "        chrom = str_split(chrom, '_')[[1]][1]\n",
    "    ) %>%\n",
    "    mutate(type = ifelse(is.na(type), 'any', type)) %>%\n",
    "    mutate(type = ifelse(chrom == 'loy', 'del', type)) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(chrom = ifelse(chrom == 23, 'X', as.character(chrom))) %>%\n",
    "    mutate(\n",
    "        chrom = factor(chrom, gtools::mixedsort(unique(chrom))),\n",
    "        gene = factor(gene, rev(genes))\n",
    "    ) %>%\n",
    "    mutate(\n",
    "        chrom_i = as.integer(chrom),\n",
    "        gene_i = as.integer(gene)\n",
    "    )\n",
    "\n",
    "D = D %>% mutate(gene_function = case_when(\n",
    "    gene %in% dna_methylation ~ 'DNA methylation',\n",
    "    gene %in% splicing ~ 'Splicing',\n",
    "    gene %in% chromatin ~ 'Chromatin',\n",
    "    gene %in% ddr ~ 'DDR',\n",
    "    T ~ 'Other'\n",
    "))\n",
    "\n",
    "p_main = D %>%\n",
    "    rbind(\n",
    "        mutate(., \n",
    "            chrom_i = case_when(\n",
    "                type == 'amp' ~ chrom_i - width,\n",
    "                type == 'loh' ~ chrom_i - width,\n",
    "                type == 'del' ~ chrom_i + width,\n",
    "                type == 'any' ~ chrom_i - width,\n",
    "            ),\n",
    "            gene_i = case_when(\n",
    "                type == 'amp' ~ gene_i + width,\n",
    "                type == 'loh' ~ gene_i + width,\n",
    "                type == 'del' ~ gene_i + width,\n",
    "                type == 'any' ~ gene_i - width,\n",
    "            )\n",
    "        ),\n",
    "        mutate(.,\n",
    "            chrom_i = case_when(\n",
    "                type == 'amp' ~ chrom_i - width,\n",
    "                type == 'loh' ~ chrom_i + width,\n",
    "                type == 'del' ~ chrom_i + width,\n",
    "                type == 'any' ~ chrom_i + width,\n",
    "            ), \n",
    "            gene_i = case_when(\n",
    "                type == 'amp' ~ gene_i - width,\n",
    "                type == 'loh' ~ gene_i + width,\n",
    "                type == 'del' ~ gene_i - width,\n",
    "                type == 'any' ~ gene_i - width,\n",
    "            )\n",
    "        )\n",
    "    ) %>%\n",
    "    rowwise() %>% mutate(OR = min(OR, limor), logOR = log2(OR)) %>% ungroup() %>%\n",
    "    ggplot(\n",
    "        aes()\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    geom_polygon(\n",
    "        aes(x = chrom_i, y = gene_i, group = paste0(gene, chrom, type), fill = OR),\n",
    "        color = 'white',\n",
    "        size = 0.5,\n",
    "#         alpha = 1\n",
    "    ) +\n",
    "    geom_vline(data = D, aes(xintercept = chrom_i + 0.5), color = 'gray95') +\n",
    "    geom_hline(data = D, aes(yintercept = gene_i + 0.5), color = 'gray95') +\n",
    "    geom_point(\n",
    "        aes(x = case_when(\n",
    "                type == 'loh' ~ as.numeric(chrom_i),\n",
    "                type == 'del' ~ as.numeric(chrom_i) + star_dodge,\n",
    "                type == 'amp' ~ as.numeric(chrom_i) - star_dodge,\n",
    "                type == 'any' ~ as.numeric(chrom_i)\n",
    "            ),\n",
    "            y = case_when(\n",
    "                type == 'loh' ~ gene_i + star_dodge,\n",
    "                type == 'del' ~ as.numeric(gene_i),\n",
    "                type == 'amp' ~ as.numeric(gene_i),\n",
    "                type == 'any' ~ as.numeric(gene_i) - star_dodge\n",
    "            ),\n",
    "            pch = signif_cat\n",
    "        ),\n",
    "        data = D %>% filter(q < 0.1),\n",
    "        color = 'black',\n",
    "        size = 1.3\n",
    "    ) +\n",
    "    geom_tile(\n",
    "        data = gene_locations %>% \n",
    "            mutate(\n",
    "                chrom_i = as.integer(factor(chrom, levels(D$chrom))),\n",
    "                gene_i = as.integer(factor(Gene, levels(D$gene)))\n",
    "            ),\n",
    "        aes(x = chrom_i, y = gene_i, linetype = ' '),\n",
    "        alpha = 0,\n",
    "        color = 'black',\n",
    "        size = 0.5\n",
    "    ) +\n",
    "    scale_fill_gradient2(low = 'blue', high = 'indianred2', mid = 'white', na.value = 'white', midpoint = 1, limits = c(0, limor)) +\n",
    "#     scale_fill_manual(values = c(cnv_pal, 'any' = 'darkorange')) +\n",
    "    ylab('') +\n",
    "    xlab('') +\n",
    "    scale_x_continuous(\n",
    "        expand = expansion(add = 0.05),\n",
    "        labels = levels(D$chrom),\n",
    "        breaks = 1:length(levels(D$chrom)),\n",
    "        position = \"top\"\n",
    "    ) +\n",
    "    scale_y_continuous(\n",
    "        expand = expansion(add = 0.05), labels = levels(D$gene), breaks = 1:length(levels(D$gene))\n",
    "    ) +\n",
    "    theme(\n",
    "        plot.margin = unit(c(0,0,0,0), 'mm'),\n",
    "        axis.text.y = element_text(\n",
    "            face = 'bold.italic',\n",
    "            color = case_when(\n",
    "                as.character(levels(D$gene)) %in% dna_methylation ~ 'royalblue',\n",
    "                as.character(levels(D$gene)) %in% splicing ~ 'firebrick',\n",
    "                as.character(levels(D$gene)) %in% chromatin ~ 'darkolivegreen3',\n",
    "                as.character(levels(D$gene)) %in% ddr ~ 'goldenrod2',\n",
    "                T ~ 'black'\n",
    "            )\n",
    "        ),\n",
    "        axis.text.x = element_text(\n",
    "            face = 'bold'\n",
    "        ),\n",
    "        legend.key.size = unit(0.3, 'cm'),\n",
    "        legend.position = 'right',\n",
    "        panel.border = element_rect(color = 'gray', fill = NA),\n",
    "        axis.line = element_line(color = 'gray'),\n",
    "        legend.text = element_text(size = 8),\n",
    "        legend.title = element_text(size = 8),\n",
    "        legend.spacing = unit(2, 'mm')\n",
    "    ) +\n",
    "    scale_shape_manual(values = c('q<0.01' = 8, 'q<0.05' = 3, 'q<0.1' = 20), name = 'FDR') +\n",
    "    geom_text(\n",
    "        data = D %>% distinct(gene, gene_function) %>% mutate(gene_function = relevel(factor(gene_function), 'Other')),\n",
    "        aes(x = 1, \n",
    "            y = 1,\n",
    "            color = gene_function,\n",
    "            label = gene),\n",
    "        size = 0,\n",
    "        key_glyph = draw_key_rect\n",
    "    ) +\n",
    "#     scale_alpha_continuous(range = c(0, 0.5)) +\n",
    "    scale_color_manual(\n",
    "        values = c(\n",
    "            'DNA methylation' = 'royalblue', \n",
    "            'Splicing' = 'firebrick', \n",
    "            'Chromatin' = 'darkolivegreen3',\n",
    "            'DDR' = 'goldenrod2',\n",
    "            'Other' = 'black'\n",
    "        )\n",
    "    ) +\n",
    "    guides(\n",
    "        color = guide_legend(keyheight = key_width, keywidth = key_width, title = 'Gene function', override.aes = list(size = 3)),\n",
    "#         alpha = guide_legend(keyheight = key_width, keywidth = key_width),\n",
    "#         fill = guide_legend(keyheight = key_width, keywidth = key_width, override.aes = list(alpha = 0.5)),\n",
    "        linetype = guide_legend(title = 'Gene location', keyheight = key_width, keywidth = key_width)\n",
    "    )\n",
    "\n",
    "do_plot(p_main, 'trans_heatmap.pdf', w = 6.5, h = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segs_filtered_auto = segs_filtered_auto %>% mutate(\n",
    "        trans_effect = \n",
    "            (chrom == 3 & type == 'amp' & TP53) |\n",
    "            (chrom == 4 & type == 'del' & (!TET2) & (SRSF2 | ATM | CHEK2 | ASXL1)) |\n",
    "            (chrom == 5 & type == 'del' & TP53) |\n",
    "            (chrom == 7 & type == 'del' & (TP53 | PPM1D)) |\n",
    "            (chrom == 8 & type == 'amp' & TET2) |\n",
    "            (chrom == 12 & type == 'amp' & (NOTCH1 | MYD88 | FBXW7)) |\n",
    "            (chrom == 13 & type == 'del' & ATM) |\n",
    "            (chrom == 20 & type == 'del' & (U2AF1 | TERT))\n",
    "    )\n",
    "\n",
    "trans_total = sum(segs_filtered_auto$trans_effect)\n",
    "trans_del = sum(segs_filtered_auto %>% filter(type == 'del') %>% pull(trans_effect))\n",
    "trans_amp = sum(segs_filtered_auto %>% filter(type == 'amp') %>% pull(trans_effect))\n",
    "\n",
    "trans_total %>% paste0(' total trans events')\n",
    "trans_del %>% paste0(' total trans deletions')\n",
    "trans_amp %>% paste0(' total trans amps')\n",
    "(trans_total/nrow(segs_filtered_auto)) %>% signif(3) %>% {.*100} %>% paste0('% of total events')\n",
    "(trans_del/sum(segs_filtered_auto$type == 'del')) %>% signif(3) %>% {.*100} %>% paste0('% of total deletions')\n",
    "(trans_amp/sum(segs_filtered_auto$type == 'amp')) %>% signif(3) %>% {.*100} %>% paste0('% of total amps')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genomic distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acen = fread('/home/gaot/ref/acen.tsv')\n",
    "colnames(acen) = c('chrom', 'start', 'end', 'arm', 'type')\n",
    "chrom_lens = fread('/home/gaot/ref/chrom_lens.tsv') %>% filter(chrom != 24)\n",
    "chrom_levels = c(as.character(1:22), 'X')\n",
    "\n",
    "acen = acen %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "    filter(chrom != 'Y') %>%\n",
    "    mutate(\n",
    "        chrom = ifelse(chrom == 'X', 23, chrom),\n",
    "        chrom = as.integer(chrom)\n",
    "    ) %>%\n",
    "    arrange(chrom) %>%\n",
    "    group_by(chrom) %>%\n",
    "    summarise(cen_start = min(start), cen_end = max(end)) %>%\n",
    "    ungroup()\n",
    "\n",
    "cytoband = fread('./cytoBand.txt') %>% setNames(c('chrom', 'start', 'end', 'name', 'stain'))\n",
    "\n",
    "cytoband = cytoband %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "    mutate(chrom = ifelse(chrom == 'X', 23, chrom)) %>%\n",
    "    filter(chrom != 'Y') %>%\n",
    "    mutate(chrom = as.integer(chrom))\n",
    "\n",
    "cyto_colors = c(\n",
    "    'gpos100'= rgb(0/255.0,0/255.0,0/255.0),\n",
    "    'gpos'   = rgb(0/255.0,0/255.0,0/255.0),\n",
    "    'gpos75' = rgb(130/255.0,130/255.0,130/255.0),\n",
    "    'gpos66' = rgb(160/255.0,160/255.0,160/255.0),\n",
    "    'gpos50' = rgb(200/255.0,200/255.0,200/255.0),\n",
    "    'gpos33' = rgb(210/255.0,210/255.0,210/255.0),\n",
    "    'gpos25' = rgb(200/255.0,200/255.0,200/255.0),\n",
    "    'gvar'   = rgb(220/255.0,220/255.0,220/255.0),\n",
    "    'gneg'  = rgb(255/255.0,255/255.0,255/255.0),\n",
    "    'acen'  = rgb(217/255.0,47/255.0,39/255.0),\n",
    "    'stalk' = rgb(100/255.0,127/255.0,164/255.0)\n",
    ")\n",
    "\n",
    "window = 5e6\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "            start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        T ~ 'coding'\n",
    "    ))\n",
    "\n",
    "p_segs = segs_filtered %>%\n",
    "    mutate(chrom = factor(ifelse(chrom == 23, 'X', as.character(chrom)), chrom_levels)) %>%\n",
    "    group_by(dmp_sample_id, chrom) %>%\n",
    "    mutate(total_length = sum(size)) %>%\n",
    "    mutate(type2 = factor(type2, c('AMP', 'CNLOH', 'DEL'))) %>%\n",
    "    arrange(type2, round(start/1e7), -total_length) %>%\n",
    "    group_by(chrom) %>%\n",
    "    mutate(index = as.integer(factor(dmp_sample_id, unique(dmp_sample_id)))) %>% \n",
    "#     left_join(\n",
    "#         M_overlap %>% select(dmp_patient_id, chrom, sm_pos = start, Gene, VariantClass, VAF_N),\n",
    "#         by = c('chrom', 'dmp_patient_id')\n",
    "#     ) %>%\n",
    "    ungroup() %>%\n",
    "    ggplot(\n",
    "        aes(x = start, xend = end, y = index, yend = index, color = type2)\n",
    "    ) +\n",
    "    geom_rect(\n",
    "        inherit.aes = F,\n",
    "        data = chrom_lens %>% mutate(chrom = factor(ifelse(chrom == 23, 'X', as.character(chrom)), chrom_levels)),\n",
    "        aes(xmin = 0, xmax = length, ymin = -2, ymax = -0.5),\n",
    "        color = 'black',\n",
    "        fill = 'white',\n",
    "        size = 1\n",
    "    ) +\n",
    "    geom_rect(\n",
    "        data = cytoband %>% mutate(chrom = factor(ifelse(chrom == 23, 'X', as.character(chrom)), chrom_levels)),\n",
    "        inherit.aes = F,\n",
    "        aes(xmin = start, xmax = end, ymin = -2, ymax = -0.5, fill = stain),\n",
    "        size = 1,\n",
    "        show.legend = FALSE\n",
    "    ) +\n",
    "    geom_segment(\n",
    "        lineend = 'round',\n",
    "        size = 1\n",
    "    ) +\n",
    "    scale_fill_manual(values = cyto_colors) +\n",
    "    theme_void() +\n",
    "    theme(\n",
    "        axis.text = element_blank(),\n",
    "        axis.ticks = element_blank(),\n",
    "        axis.title = element_blank(),\n",
    "        legend.position = 'right',\n",
    "        panel.spacing.y = unit(0, \"lines\"),\n",
    "        plot.margin = unit(c(0,0,0,0), \"lines\"),\n",
    "        strip.text.y = element_text(size = 10, hjust = 0.5, angle = 0),\n",
    "        panel.border = element_blank(),\n",
    "        strip.background = element_blank()\n",
    "    ) +\n",
    "    scale_y_discrete(expand = expansion(add = 1)) +\n",
    "    scale_x_discrete(expand = c(0.02,0)) +\n",
    "    facet_grid(.~chrom, space = 'free', scales = 'free', switch=\"both\") +\n",
    "    xlab('') +\n",
    "    ylab('') +\n",
    "    scale_color_manual(values = cnv_pal2) +\n",
    "    labs(color = '')\n",
    "\n",
    "do_plot(p_segs, 'genome_wide.pdf', w = 20, h = 2.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv_pal2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "impact_bed = fread('/work/isabl/data/techniques/51/bed_files/GRCh37/dna-td-impact-468-grch37.targets.bed.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(impact_bed) = c('chrom', 'start', 'end', 'strand', 'target')\n",
    "impact_bed = impact_bed %>% filter(chrom != 'Y')\n",
    "impact_bed = impact_bed %>% mutate(chrom = as.integer(ifelse(chrom == 'X', 23, chrom)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CDRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "focal_dels = function(gene) {\n",
    "    \n",
    "    all_colors = c(\n",
    "        'amp' = 'darkblue', 'loh' = 'darkgreen', 'del' = 'darkred',\n",
    "        'Regulatory' = 'darkturquoise', 'Missense' = 'salmon', 'Truncating' = 'purple') \n",
    "    \n",
    "    gene_structure = gene_structures %>% filter(str_detect(description, paste0('\"', gene, '\"')))\n",
    "\n",
    "    gene_loc = gene_bed %>% \n",
    "        filter(Gene == gene) %>%\n",
    "        as.data.frame(stringsAsFactors = F) %>%\n",
    "        mutate(loc = as.integer((start + end) / 2), chrom = as.integer(chrom))\n",
    "\n",
    "    this_chrom_lens = chrom_lens %>% filter(chrom == gene_loc$chrom)\n",
    "\n",
    "    acen_chrom = acen %>% filter(chrom == gene_loc$chrom)\n",
    "\n",
    "    D = segs_filtered %>%\n",
    "        filter(type == 'del' & start < gene_loc$start + 5e6 & end > gene_loc$end - 5e6) %>%\n",
    "        filter(chrom == gene_loc$chrom) %>%\n",
    "        mutate(chrom = factor(chrom)) %>%\n",
    "        group_by(dmp_sample_id, chrom) %>%\n",
    "        mutate(total_length = sum(size)) %>%\n",
    "        mutate(type = factor(type, c('amp', 'loh', 'del'))) %>%\n",
    "        arrange(desc(type), round(start/1e7), -total_length) %>%\n",
    "        group_by(chrom) %>%\n",
    "        mutate(index = as.integer(factor(dmp_sample_id, unique(dmp_sample_id)))) %>%\n",
    "        ungroup()\n",
    "\n",
    "    p_segs = ggplot(\n",
    "            D,\n",
    "            aes(x = start, xend = end, y = index, yend = index, color = type)\n",
    "        ) +\n",
    "        geom_rect(\n",
    "            inherit.aes = F,\n",
    "            data = acen_chrom,\n",
    "            aes(xmin = cen_start, xmax = cen_end, ymin = -Inf, ymax = Inf),\n",
    "            color = 'gray',\n",
    "            size = 0,\n",
    "            alpha = 0.3\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            size = 1,\n",
    "            alpha = 0.8,\n",
    "            lineend = 'round'\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = this_chrom_lens,\n",
    "            aes(x = 0, xend = length, y = 0, yend = 0),\n",
    "            alpha = 0,\n",
    "            size = 0,\n",
    "            color = 'white'\n",
    "        ) +\n",
    "        scale_alpha(\n",
    "            range = c(0, 1)\n",
    "        ) +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            axis.text = element_blank(),\n",
    "            axis.ticks = element_blank(),\n",
    "            axis.title = element_blank(),\n",
    "            legend.position = 'none',\n",
    "            panel.spacing.y = unit(1, \"mm\"),\n",
    "            plot.margin = unit(c(0.5,0.2,0.2,0.2), \"lines\"),\n",
    "            strip.text.y = element_blank(),\n",
    "            panel.border = element_rect(size = 0.2, fill = NA, color = 'gray'),\n",
    "            strip.background = element_rect(size = 0.5, fill = 'seashell3', color = 'seashell3'),\n",
    "            plot.title = element_text(size = 7)\n",
    "        ) +\n",
    "        scale_y_discrete(expand = expansion(add = 1)) +\n",
    "        xlab('') +\n",
    "        ylab('') +\n",
    "        scale_color_manual(values = all_colors) +\n",
    "        ggtitle(paste0(gene, ', ', 'chr', gene_loc$chrom)) +\n",
    "        # exon structures\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = gene_structure %>% mutate(zoom = TRUE),\n",
    "            aes(x = min(start), xend = max(end), y = -1, yend = -1),\n",
    "            size = 0.1,\n",
    "            color = 'black'\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = gene_structure %>% mutate(zoom = TRUE),\n",
    "            aes(x = start, xend = end, y = -1, yend = -1),\n",
    "            size = 2,\n",
    "            color = 'black'\n",
    "        ) +\n",
    "        facet_zoom(\n",
    "            zoom.size = 1,\n",
    "            zoom.data = zoom,\n",
    "            xlim = c(gene_loc$start, gene_loc$end)\n",
    "        )\n",
    "\n",
    "    p_segs\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "focal_genes = c('DNMT3A', 'TET2', 'ATM', 'ETV6', 'NF1', 'CHEK2')\n",
    "\n",
    "plots = list()\n",
    "\n",
    "for (gene in focal_genes) {\n",
    "    plots[[gene]] = focal_dels(gene)\n",
    "}\n",
    "\n",
    "p_focal = plot_grid(plotlist = plots, nrow = 2)\n",
    "\n",
    "do_plot(p_focal, 'focal_del.pdf', w = 4, h = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Co-localization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 5e6\n",
    "\n",
    "segs_filtered_auto = segs_filtered %>% filter(chrom != 23)\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered_auto$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    mutate(\n",
    "        locality = ifelse(\n",
    "            any(chrom == segs_filtered_auto$chrom & dmp_patient_id == segs_filtered_auto$dmp_patient_id &\n",
    "                start >= segs_filtered_auto$start - window & start <= segs_filtered_auto$end + window\n",
    "            ),\n",
    "            'cis',\n",
    "            'trans'\n",
    "        )\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(variant_type = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        T ~ 'coding'\n",
    "    ))\n",
    "\n",
    "perm = fread('./data/permutation.tsv', sep = '\\t')\n",
    "\n",
    "# perm = data.frame()\n",
    "\n",
    "# for (i in 1:2000) {\n",
    "    \n",
    "#     if (i %% 500 == 0) {\n",
    "#         display(i)\n",
    "#     }\n",
    "    \n",
    "#     res = M_overlap %>%\n",
    "#         mutate(dmp_patient_id = sample(dmp_patient_id)) %>%\n",
    "#         rowwise() %>%\n",
    "#         mutate(\n",
    "#             locality = ifelse(\n",
    "#                 any(chrom == segs_filtered_auto$chrom & dmp_patient_id == segs_filtered_auto$dmp_patient_id &\n",
    "#                     start >= segs_filtered_auto$start - window & start <= segs_filtered_auto$end + window\n",
    "#                 ),\n",
    "#                 'cis',\n",
    "#                 'trans'\n",
    "#             )\n",
    "#         ) %>%\n",
    "#         ungroup() %>%\n",
    "#         group_by(variant_type) %>%\n",
    "#         summarise(\n",
    "#             prop_cis = sum(locality == 'cis')/n(),\n",
    "#             .groups = 'drop'\n",
    "#         ) %>%\n",
    "#         mutate(i = i)\n",
    "    \n",
    "#     perm = perm %>% rbind(res)\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prop_cis_coding = M_overlap %>% filter(variant_type == 'coding') %>% pull(locality) %>% {sum(. == 'cis')/length(.)}\n",
    "prop_cis_noncoding = M_overlap %>% filter(variant_type == 'noncoding') %>% pull(locality) %>% {sum(. == 'cis')/length(.)}\n",
    "prop_coding_nonrecurrent = M_overlap %>%\n",
    "    filter(variant_type == 'coding') %>%\n",
    "    filter(!Gene %in% c('MPL', 'DNMT3A', 'TET2', 'EZH2', 'JAK2', 'ATM', 'TP53')) %>% \n",
    "    pull(locality) %>% {sum(. == 'cis')/length(.)}\n",
    "    \n",
    "prop_cis_coding %>% paste('cis coding')\n",
    "prop_coding_nonrecurrent %>% paste('cis coding non recurrent')\n",
    "prop_cis_noncoding %>% paste('cis noncoding')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_coding = perm %>% filter(variant_type == 'coding') %>% pull(prop_cis) %>% {sum(. > prop_cis_coding)/length(.)}\n",
    "p_noncoding = perm %>% filter(variant_type == 'noncoding') %>% pull(prop_cis) %>% {sum(. > prop_cis_noncoding)/length(.)}\n",
    "\n",
    "D = M_overlap %>% \n",
    "    group_by(variant_type) %>%\n",
    "    summarise(\n",
    "        prop_cis_median = sum(locality == 'cis')/n()\n",
    "    ) %>%\n",
    "    mutate(\n",
    "        prop_cis_lower = prop_cis_median, \n",
    "        prop_cis_upper = prop_cis_median,\n",
    "        estimate = 'observed'\n",
    "    ) %>%\n",
    "    rbind(\n",
    "        perm %>% \n",
    "        group_by(variant_type) %>%\n",
    "        summarise(\n",
    "            prop_cis_lower = quantile(prop_cis, c(.05))[1],\n",
    "            prop_cis_upper = quantile(prop_cis, c(.95))[1],\n",
    "            prop_cis_median = median(prop_cis)\n",
    "        ) %>%\n",
    "        mutate(\n",
    "            estimate = 'expected'\n",
    "        )\n",
    "    ) %>%\n",
    "    mutate(\n",
    "        variant_type = Hmisc::capitalize(variant_type)\n",
    "    )\n",
    "\n",
    "\n",
    "p_perm = ggplot(\n",
    "        D,\n",
    "        aes(x = variant_type, y = prop_cis_median, ymin = prop_cis_lower, ymax = prop_cis_upper, pch = estimate)\n",
    "    ) +\n",
    "    geom_pointrange() +\n",
    "    theme_classic() +\n",
    "    scale_shape_manual(values=c(16, 5)) +\n",
    "    theme(legend.position = 'right') +\n",
    "    ylab('Proportion of cis locality') +\n",
    "    ylim(0, 0.18) +\n",
    "    xlab('') +\n",
    "    # stat_compare_means(\n",
    "    #     comparisons = list(c('coding', 'noncoding')),\n",
    "    #     label.y = 0.14,\n",
    "    #     size = 0\n",
    "    # )\n",
    "    annotate('text', x = 1, y = 0.18, label = glue('p<0.001'), size = 3) +\n",
    "    annotate('text', x = 2, y = 0.18, label = glue('p={signif(p_noncoding, 2)}'), size = 3) +\n",
    "    theme(legend.title = element_blank())\n",
    "\n",
    "do_plot(p_perm, 'permutation.pdf', w = 3, h = 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## general characterstics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 4, repr.plot.height = 2.5, repr.plot.res = 300)\n",
    "\n",
    "p_dev = ggplot(\n",
    "  segs_filtered,\n",
    "  aes(x = valor, y = cnlr_adj, color = type)\n",
    ") +\n",
    "scale_color_manual(values = cnv_pal) +\n",
    "# geom_hline(yintercept = 0, linetype = 'dashed', alpha = 0.8, color = 'darkgreen') +\n",
    "geom_point(pch = 21, stroke = 0.5) +\n",
    "scale_x_continuous(limits = c(0, 2.5), expand = c(0,0)) +\n",
    "scale_y_continuous(limits = c(-1, 0.6), expand = c(0,0)) +\n",
    "theme_classic() +\n",
    "theme(\n",
    "    legend.position = 'none',\n",
    "    axis.title = element_text(size = 8)\n",
    ") +\n",
    "# geom_text(size = 2) +\n",
    "geom_line(\n",
    "  inherit.aes = F,\n",
    "  data = del_line,\n",
    "  aes(x = valor, y = cnlr),\n",
    "  linetype = 'dashed', alpha = 0.8, color = 'darkred'\n",
    ") +\n",
    "geom_line(\n",
    "  inherit.aes = F,\n",
    "  data = amp_line,\n",
    "  aes(x = valor, y = cnlr),\n",
    "  linetype = 'dashed', alpha = 0.8, color = 'darkblue'\n",
    ") +\n",
    "geom_line(\n",
    "  inherit.aes = F,\n",
    "  data = loh_line,\n",
    "  aes(x = valor, y = cnlr),\n",
    "  linetype = 'dashed', alpha = 0.8, color = 'darkgreen'\n",
    ") +\n",
    "xlab('logOR') +\n",
    "ylab('logR') + \n",
    "annotate('text', x = 0.9, y = 0.5, label = 'AMP', color = cnv_pal['amp'], size = 3) +\n",
    "annotate('text', x = 2.1, y = 0.2, label = 'CNLOH', color = cnv_pal['loh'], size = 3) +\n",
    "annotate('text', x = 2, y = -0.5, label = 'DEL', color = cnv_pal['del'], size = 3)\n",
    "# annotate('text', x = 0.2, y = -0.8, label = paste0('n=', nrow(segs_filtered)), color = 'gray30', size = 3)\n",
    "\n",
    "p_dev"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 2, repr.plot.res = 300)\n",
    "\n",
    "p_age = ggplot(\n",
    "    M_wide %>% \n",
    "        mutate(age_bin = cut(age, c(0, seq(40, 80, 10), 100))) %>%\n",
    "        group_by(age_bin) %>%\n",
    "        summarise(\n",
    "            prop_ch_cnv = sum(ch_cnv)/n(),\n",
    "            prop_ch_cnv_auto = sum(ch_cnv_auto)/n(),\n",
    "            prop_both_ch = sum(ch_cnv & ch_nonsilent)/n(),\n",
    "            n = n()\n",
    "        ),\n",
    "    aes(x = age_bin, y = prop_ch_cnv, group = '1')\n",
    ") +\n",
    "geom_line() +\n",
    "theme_classic() +\n",
    "geom_point() +\n",
    "ylim(0, 0.04) +\n",
    "theme(\n",
    "    axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1, size = 8),\n",
    "    axis.title = element_text(size = 8)\n",
    ") +\n",
    "ylab('Proportion with mCA')\n",
    "\n",
    "p_age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 3, repr.plot.height = 3, repr.plot.res = 300)\n",
    "\n",
    "give.min <- function(x){\n",
    "  return(c(y = min(x), label = min(x))) \n",
    "}\n",
    "\n",
    "p_phi = ggplot(\n",
    "        segs_filtered,\n",
    "        aes(x = type2, y = phi, fill = type2, color = type2)\n",
    "    ) +\n",
    "#     geom_hline(yintercept = 0.1, linetype = 'dashed', color = 'gray', size = 0.5, alpha = 0.8) +\n",
    "    geom_violin(size = 0.3, position = position_dodge(width = 0.75), alpha = 0.2, scale = 'width') +\n",
    "    theme_classic() +\n",
    "    geom_quasirandom(width = 0.1, method = 'smiley', size = 0.5) +\n",
    "    scale_fill_manual(values = cnv_pal2) +\n",
    "    scale_color_manual(values = cnv_pal2) +\n",
    "    ylim(0,NA) +\n",
    "    theme(\n",
    "        legend.position = 'none',\n",
    "        axis.title = element_text(size = 8),\n",
    "        axis.text.x = element_text(size = 8, angle = 30, hjust = 1, vjust = 1)\n",
    "    ) +\n",
    "    ylab('Aberrant cell fraction') +\n",
    "    xlab('Event type') +\n",
    "    stat_summary(fun.data = give.min, geom = \"text\", vjust = 1.5, size = 2.5)\n",
    "\n",
    "p_phi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 1.5, repr.plot.height = 2, repr.plot.res = 300)\n",
    "\n",
    "p_n = ggplot(\n",
    "        segs_filtered %>% group_by(dmp_sample_id) %>% mutate(n_sample = n()) %>%\n",
    "        ungroup %>% count(n_sample, active_heme, type) %>% mutate(n = n/n_sample) %>%\n",
    "            mutate(n_sample = case_when(\n",
    "                n_sample < 3 ~ as.character(n_sample), T ~ '3+')),\n",
    "        aes(x = n_sample, y = n, fill = type)\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    geom_bar(stat = 'identity', position = 'stack') +\n",
    "    scale_fill_manual(values = cnv_pal) +\n",
    "    theme(\n",
    "        legend.position = 'none',\n",
    "        axis.title = element_text(size = 8)\n",
    "    ) +\n",
    "    xlab('Events per sample')\n",
    "\n",
    "p_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "panel = p_age + p_n + p_dev + p_phi + plot_layout(widths = c(1.8,1))\n",
    "\n",
    "do_plot(panel, 'ai_general.pdf', w = 3.8, h = 3.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 2, repr.plot.res = 300)\n",
    "\n",
    "pval = 0.01\n",
    "label_size = 8\n",
    "\n",
    "p_mutnum = ggplot(\n",
    "    M_wide %>% \n",
    "        mutate(mutnum_silent = mutnum_silent + mutnum_intronic) %>%\n",
    "        mutate(ch_cnv = ifelse(ch_cnv == 1, 'mCA+', 'mCA-')) %>%\n",
    "        reshape2::melt(\n",
    "            measure.vars = c('mutnum_all', 'mutnum_nonsilent', 'mutnum_silent'),\n",
    "            variable.name = 'mutation_type',\n",
    "            value.name = 'mutnum'\n",
    "        ) %>%\n",
    "        filter(mutation_type == 'mutnum_all') %>%\n",
    "        mutate(mutnum = ifelse(mutnum >= 3, '3+', as.character(mutnum))) %>%\n",
    "        count(ch_cnv, mutnum, mutation_type) %>%\n",
    "        group_by(ch_cnv) %>%\n",
    "        mutate(prop = n/sum(n)) %>%\n",
    "        ungroup() %>%\n",
    "        filter(mutnum > 0),\n",
    "    aes(x = ch_cnv, y = prop, fill = mutnum)\n",
    ") +\n",
    "geom_col(color = 'black', size = 0.2) +\n",
    "theme_classic() +\n",
    "scale_fill_manual(values = scales::seq_gradient_pal(\"white\", sm_col)(c(0.3, 0.6, 1))) +\n",
    "scale_y_continuous(expand = expansion(add = 0.05), limits = c(0,0.85)) +\n",
    "stat_compare_means(\n",
    "    comparisons = list(c('mCA+', 'mCA-')),\n",
    "    label.y = 0.8,\n",
    "    size = 0,\n",
    "    method = 't.test'\n",
    ") +\n",
    "annotate('text', x = 1.5, y = 0.8, label = '****', size = 3) +\n",
    "xlab('') +\n",
    "theme(\n",
    "    legend.key.size = unit(3, 'mm'),\n",
    "    legend.title = element_text(size = 7),\n",
    "    legend.position = 'top',\n",
    "    legend.spacing = unit(0,'mm'),\n",
    "    axis.text.x = element_text(size = 8, angle = 30, hjust = 1)\n",
    ") +\n",
    "guides(fill = guide_legend(title.position = \"top\"))\n",
    "\n",
    "p_vaf = ggplot(\n",
    "        M_wide %>% \n",
    "            rowwise() %>% mutate(VAF_silent = max(VAF_silent, VAF_intronic)) %>% ungroup() %>%\n",
    "            mutate(ch_cnv = ifelse(ch_cnv == 1, 'mCA+', 'mCA-')) %>%\n",
    "            reshape2::melt(\n",
    "                measure.vars = c('VAF_all', 'VAF_nonsilent', 'VAF_silent'),\n",
    "                variable.name = 'mutation_type',\n",
    "                value.name = 'VAF'\n",
    "            ) %>% \n",
    "            filter(VAF > 0 & VAF < 0.5 & mutation_type == 'VAF_all'),\n",
    "        aes(x = ch_cnv, y = VAF, fill = ch_cnv)\n",
    "    ) +\n",
    "    geom_violin(draw_quantiles = c(0.5)) +\n",
    "    theme_classic() +\n",
    "    stat_compare_means(\n",
    "        size = 2.5,\n",
    "        method = 't.test', \n",
    "        label = 'p.signif', \n",
    "        comparisons = list(c('mCA+', 'mCA-'))) +\n",
    "    ylim(0,0.55) +\n",
    "    theme(\n",
    "        axis.text.x = element_text(size = 8, angle = 30, hjust = 1),\n",
    "        legend.position = 'none'\n",
    "    ) +\n",
    "    xlab('') +\n",
    "    scale_fill_manual(values = c(cnv_col, 'gainsboro'))\n",
    "\n",
    "D = M_wide %>%\n",
    "    mutate(ch_status = case_when(\n",
    "        ch_cnv & ch_nonsilent ~ 'Both',\n",
    "        ch_cnv == 1 ~ 'mCA only',\n",
    "        ch_nonsilent == 1 ~ 'GM only',\n",
    "        T ~ 'none'\n",
    "    )) %>%\n",
    "    group_by(ch_status) %>%\n",
    "    mutate(n = n()) %>%\n",
    "    ungroup() %>%\n",
    "    filter(ch_status != 'none')\n",
    "\n",
    "\n",
    "give.n <- function(x){\n",
    "  return(c(y = min(x), label = length(x))) \n",
    "}\n",
    "\n",
    "p_age = ggplot(\n",
    "        D,\n",
    "        aes(x = ch_status, y = age, fill = ch_status)\n",
    "    ) +\n",
    "#     geom_violin(draw_quantiles = c(0.5)) +\n",
    "    geom_boxplot(outlier.alpha = 0) +\n",
    "    theme_classic() +\n",
    "    stat_compare_means(\n",
    "        comparisons = list(\n",
    "            c('mCA only', 'GM only'),\n",
    "            c('Both', 'GM only'),\n",
    "            c('mCA only', 'Both')\n",
    "        ),\n",
    "        size = 2.2,\n",
    "        method = 't.test', label = 'p.value',\n",
    "        vjust = 0,\n",
    "        step.increase = 0.15\n",
    "    ) +\n",
    "#     ylim(39, 130) +\n",
    "    theme(\n",
    "#         plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = \"pt\"),\n",
    "#         axis.title.x = element_blank()\n",
    "        axis.text.x = element_text(size = 8, angle = 30, hjust = 1),\n",
    "        legend.position = 'none'\n",
    "    ) +\n",
    "    xlab('') +\n",
    "    scale_fill_manual(values = c('mCA only' = cnv_col, 'Both' = 'darkorange', 'GM only' = sm_col)) +\n",
    "    ylab('Age') + coord_cartesian(ylim=c(39, 130))  + scale_y_continuous(breaks = seq(40,110,20), labels=c(\"40\" = \"40\", \"60\" = \"60\",\n",
    "                              \"80\" = \"80\", \"100\" = \"100\"))\n",
    "\n",
    "# p_phi = ggplot(\n",
    "#         segs_filtered %>% \n",
    "#             mutate(ch_nonsilent = ifelse(ch_nonsilent == 1, 'GM+', 'GM-')),\n",
    "#         aes(x = ch_nonsilent, y = phi, fill = ch_nonsilent)\n",
    "#     ) +\n",
    "#     geom_violin(\n",
    "#         draw_quantiles = c(0.5),\n",
    "#         scale = 'width',\n",
    "#         trim = T\n",
    "#     ) +\n",
    "#     theme_classic() +\n",
    "#     stat_compare_means(\n",
    "#         size = 2.5,\n",
    "#         method = 't.test',\n",
    "#         label = 'p.value',\n",
    "#         label.x = 1.5,\n",
    "#         comparisons = list(c('GM+', 'GM-'))\n",
    "#     ) +\n",
    "#     ylim(0,1) +\n",
    "#     xlab('') +\n",
    "#     ylab('Cell fraction') +\n",
    "#     theme(\n",
    "#         axis.text.x = element_text(size = 8, angle = 30, hjust = 1),\n",
    "#         legend.position = 'none'\n",
    "#     ) +\n",
    "#     xlab('') +\n",
    "#     scale_fill_manual(values = c(sm_col, 'gainsboro'))\n",
    "\n",
    "do_plot(p_age, 'age_pvalue.pdf', w = 1.6, h = 1.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 2, repr.plot.res = 300)\n",
    "\n",
    "ggplot(\n",
    "    segs_filtered %>% \n",
    "        mutate(age_bin = cut(age, c(20, seq(40, 80, 10), Inf))) %>%\n",
    "        group_by(age_bin, type) %>%\n",
    "        summarise(\n",
    "            ci_lower = mean(phi) - sd(phi),\n",
    "            ci_upper = mean(phi) + sd(phi),\n",
    "            phi = mean(phi),\n",
    "            n = n()\n",
    "        ),\n",
    "    aes(x = age_bin, y = phi, ymin = ci_lower, ymax = ci_upper, group = type, color = type, fill = type)\n",
    ") +\n",
    "geom_line() +\n",
    "theme_classic() +\n",
    "geom_point() +\n",
    "# geom_errorbar(size = 0.5, width = 0.1) +\n",
    "geom_ribbon(alpha = 0.2, size = 0) +\n",
    "scale_color_manual(values = c('loh' = 'darkgreen', 'amp' = 'darkblue', 'del' = 'darkred', 'err' = 'gray')) +\n",
    "scale_fill_manual(values = c('loh' = 'darkgreen', 'amp' = 'darkblue', 'del' = 'darkred', 'err' = 'gray')) +\n",
    "facet_wrap(~type) +\n",
    "theme(\n",
    "    axis.text.x = element_text(angle = 90, vjust = 0.5),\n",
    "    strip.background = element_blank()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comparison with other studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segs_laurie = fread('./data/external/segs_laurie.txt')\n",
    "\n",
    "segs_laurie = segs_laurie %>% \n",
    "    mutate(\n",
    "        size = length * 1e6,\n",
    "        type = c('aupd' = 'loh', 'gain' = 'amp', 'loss' = 'del')[mosaic.type],\n",
    "        start = `Start (GRCh37)`,\n",
    "        end = `End (GRCh37)`,\n",
    "        chrom = chromosome,\n",
    "        patient_id = as.character(subject.id)\n",
    "    ) %>%\n",
    "    mutate(n_total = 50222)\n",
    "\n",
    "segs_loh = fread('./data/external/segs_loh.txt') %>%\n",
    "    filter(COPY_CHANGE != 'undetermined') %>%\n",
    "    mutate(\n",
    "        chrom = as.integer(ifelse(CHR == 'X', 23, CHR)),\n",
    "        size = SIZE_MB * 1e6,\n",
    "        start = `Start (GRCh37)`,\n",
    "        end = `End (GRCh37)`,\n",
    "        type = c('neutral' = 'loh', 'loss' = 'del', 'gain' = 'amp')[COPY_CHANGE],\n",
    "        phi = as.numeric(CELL_FRAC),\n",
    "        patient_id = as.character(ID)\n",
    "    ) %>%\n",
    "    mutate(n_total = 151202)\n",
    "\n",
    "segs_jacobs = fread('./data/external/segs_jacobs.txt') %>%\n",
    "    mutate(\n",
    "        chrom = CHROM,\n",
    "        start = `Start (GRCh37)`,\n",
    "        end = `End (GRCh37)`,\n",
    "        size = SEG_SIZE,\n",
    "        type = c('NEUTRAL' = 'loh', 'LOSS' = 'del', 'GAIN' = 'amp')[STATE],\n",
    "        phi = PROPORTION_ABNORMAL,\n",
    "        patient_id = as.character(ID)\n",
    "    ) %>%\n",
    "    mutate(n_total = 31717 + 26136)\n",
    "\n",
    "segs_machiela = fread('./data/external/segs_machiela.txt') %>%\n",
    "    mutate(\n",
    "        chrom = as.integer(CHR),\n",
    "        start = `Start (GRCh37)`,\n",
    "        end = `End (GRCh37)`,\n",
    "        size = SEG_SIZE,\n",
    "        type = c('Neutral' = 'loh', 'Loss' = 'del', 'Gain' = 'amp')[STATE],\n",
    "        phi = PERCENT,\n",
    "        age = as.numeric(AGE),\n",
    "        patient_id = as.character(ID)\n",
    "    ) %>%\n",
    "    select(-AGE, -CHR) %>%\n",
    "    filter(!is.na(type)) %>%\n",
    "    mutate(n_total = 24849)\n",
    "    \n",
    "segs_all = bind_rows(\n",
    "        segs_filtered %>% mutate(study = 'This study') %>%\n",
    "            mutate(n_total = 32444) %>%\n",
    "            mutate(patient_id = dmp_patient_id),\n",
    "        segs_loh %>% mutate(study = 'Loh'),\n",
    "        segs_laurie %>% mutate(study = 'Laurie'),\n",
    "        segs_jacobs %>% mutate(study = 'Jacobs'),\n",
    "        segs_machiela %>% mutate(study = 'Machiela')\n",
    "    ) %>% \n",
    "    mutate(study_label = paste0(study, '\\n(n=', n_total, ')')) %>%\n",
    "    mutate(study_label = factor(study_label, unique(study_label))) %>%\n",
    "    mutate(study = factor(study, unique(study))) %>%\n",
    "    mutate(type2 = factor(c('del' = 'DEL', 'loh' = 'CNLOH', 'amp' = 'AMP')[type], c('DEL', 'CNLOH', 'AMP')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "study_info = data.frame(\n",
    "    'Study' = c('This study', 'Loh et al.', 'Laurie et al.', 'Jacobs et al.', 'Machiela et al. (TSGII)', 'Jaiswal et al.', 'Genovese et al.', 'Zink et al.', 'Terao et al.', 'Loh et al.'),\n",
    "    'Year' = c(2020, 2018, 2012, 2012, 2015, 2014, 2014, 2017, 2020, 2020),\n",
    "    'Genotyping Platform' = c('Targeted Sequencing', 'SNP-array', 'SNP-array', 'SNP-array', 'SNP-array', 'WES', 'WES', 'WGS', 'SNP-array', 'SNP-array/WES(partial)'),\n",
    "    'Mutation Type' = c('mCA/GM(468 genes)', 'mCA', 'mCA', 'mCA', 'mCA', 'GM', 'GM', 'GM', 'mCA', 'mCA/GM(3 genes)'),\n",
    "    'Sample Size' = c('32,442', '151,202', '50,222', '57,853', '24,849', '17,182', '12,380', '11,262', '179,417', '482,789/49,960'),\n",
    "    'Cohort Type' = c('Cancer', 'Normal', 'Normal', 'Cancer/Normal', 'Cancer/Normal', 'Normal', 'Normal', 'Normal', 'Normal', 'Normal'),\n",
    "    'Haplotype Info' = c('No', 'Yes', 'No', 'No', 'No', '-', '-', '-', 'Yes', 'Yes'),\n",
    "    check.names = FALSE\n",
    ") %>%\n",
    "arrange(Year, Study == 'This study')\n",
    "\n",
    "kb = study_info %>%\n",
    "    kable(format = \"html\", booktabs = T) %>%\n",
    "    row_spec(0:nrow(study_info), background = 'white', align = 'center', extra_css = \"border-bottom: 1px solid\") %>%\n",
    "    row_spec(0, background = 'gray', extra_css = \"border-bottom: 2px solid\") %>%\n",
    "    kable_styling(position = 'center', full_width = FALSE)\n",
    "\n",
    "kb %>% toString() %>%\n",
    "display_html()\n",
    "\n",
    "#save_kable(kb, './figures/studies.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "give.n <- function(x){\n",
    "  return(c(y = max(x), label = length(x))) \n",
    "}\n",
    "\n",
    "p = segs_all %>%\n",
    "    mutate(size = size/1e6) %>%\n",
    "    filter(chrom != '23') %>%\n",
    "    filter(phi >= 0.1 | is.na(phi)) %>%\n",
    "    ggplot(\n",
    "        aes(x = type2, y = size, fill = type2)\n",
    "    ) +\n",
    "    geom_violin(alpha = 0.8, trim = F, draw_quantiles = c(0.5), scale = 'width', width = 0.8) +\n",
    "    theme_classic() +\n",
    "    scale_fill_manual(values = cnv_pal2) +\n",
    "    scale_color_manual(values = cnv_pal2) +\n",
    "    scale_y_continuous(trans = 'log10', limits = c(0.05, NA)) +\n",
    "    facet_wrap(~study, nrow = 1) +\n",
    "    # stat_summary(fun.data = give.n, geom = \"text\", hjust = 1.5, vjust = -1, size = 2.5) +\n",
    "    theme(panel.spacing = unit(1, 'mm'), legend.position = 'none') +\n",
    "    ylab('size in Mb') +\n",
    "    xlab('Event type')\n",
    "\n",
    "do_plot(p, 'size.pdf', w = 8, h = 2.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "give.min <- function(x){\n",
    "  return(c(y = min(x), label = round(min(x), 2))) \n",
    "}\n",
    "\n",
    "p = segs_all %>%\n",
    "    filter(!is.na(phi)) %>%\n",
    "    filter(study != 'Loh') %>%\n",
    "    ggplot(\n",
    "        aes(x = type2, y = phi, fill = type2, color = type2)\n",
    "    ) +\n",
    "    geom_violin(size = 0.3, position = position_dodge(width = 0.75), alpha = 0.2, scale = 'width') +\n",
    "    theme_classic() +\n",
    "    geom_quasirandom(width = 0.1, method = 'smiley', size = 0.5) +\n",
    "    scale_fill_manual(values = cnv_pal2) +\n",
    "    scale_color_manual(values = cnv_pal2) +\n",
    "    scale_y_continuous(expand = expansion(add = 0.15)) +\n",
    "    theme(\n",
    "        axis.title = element_text(size = 8)\n",
    "    ) +\n",
    "    # scale_y_continuous()\n",
    "    ylab('Aberrant cell fraction') +\n",
    "    xlab('Event type') +\n",
    "    stat_summary(fun.data = give.min, geom = \"text\", vjust = 1.5, size = 2.5) +\n",
    "    facet_wrap(~study, nrow = 1) +\n",
    "    theme(panel.spacing = unit(2, 'mm'), legend.position = 'none')\n",
    "\n",
    "do_plot(p, 'cell_frac.pdf', w = 5, h = 2.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = segs_all %>%\n",
    "    filter(chrom != '23') %>%\n",
    "    filter(phi >= 0.1 | is.na(phi)) %>%\n",
    "    group_by(study, chrom, type2) %>%\n",
    "    summarise(\n",
    "        n = length(unique(patient_id)),\n",
    "        prop = n/unique(n_total),\n",
    "        .groups = 'drop'\n",
    "    ) %>%\n",
    "    tidyr::complete(chrom, type2, study, fill = list(n = 0, prop = 0)) %>% \n",
    "    ggplot(\n",
    "        aes(x = type2, y = prop, fill = study)\n",
    "    ) +\n",
    "    scale_x_discrete(drop=FALSE, expand = expansion(add = 0.5)) +\n",
    "    geom_col(width = 0.5, position = position_dodge(width = 0.5)) +\n",
    "    facet_wrap(~chrom, scale = 'free') +\n",
    "    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        axis.text.y = element_text(size = 6)\n",
    "    ) +\n",
    "    ylab('Proportion of subjects') +\n",
    "    xlab('Event type')\n",
    "                   \n",
    "do_plot(p, 'comparison_incidence.pdf', w = 11, h = 6.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acen = fread('/home/gaot/ref/acen.tsv')\n",
    "colnames(acen) = c('chrom', 'start', 'end', 'arm', 'type')\n",
    "chrom_lens = fread('/home/gaot/ref/chrom_lens.tsv') %>% filter(chrom != 24)\n",
    "\n",
    "acen = acen %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "    filter(chrom != 'Y') %>%\n",
    "    mutate(\n",
    "        chrom = ifelse(chrom == 'X', 23, chrom),\n",
    "        chrom = as.integer(chrom)\n",
    "    ) %>%\n",
    "    arrange(chrom) %>%\n",
    "    group_by(chrom) %>%\n",
    "    summarise(cen_start = min(start), cen_end = max(end), .groups = 'drop') %>%\n",
    "    ungroup()\n",
    "\n",
    "cytoband = fread('./cytoBand.txt') %>% setNames(c('chrom', 'start', 'end', 'name', 'stain'))\n",
    "\n",
    "cytoband = cytoband %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "    mutate(chrom = ifelse(chrom == 'X', 23, chrom)) %>%\n",
    "    filter(chrom != 'Y') %>%\n",
    "    mutate(chrom = as.integer(chrom))\n",
    "\n",
    "cyto_colors = c(\n",
    "    'gpos100'= rgb(0/255.0,0/255.0,0/255.0),\n",
    "    'gpos'   = rgb(0/255.0,0/255.0,0/255.0),\n",
    "    'gpos75' = rgb(130/255.0,130/255.0,130/255.0),\n",
    "    'gpos66' = rgb(160/255.0,160/255.0,160/255.0),\n",
    "    'gpos50' = rgb(200/255.0,200/255.0,200/255.0),\n",
    "    'gpos33' = rgb(210/255.0,210/255.0,210/255.0),\n",
    "    'gpos25' = rgb(200/255.0,200/255.0,200/255.0),\n",
    "    'gvar'   = rgb(220/255.0,220/255.0,220/255.0),\n",
    "    'gneg'  = rgb(255/255.0,255/255.0,255/255.0),\n",
    "    'acen'  = rgb(217/255.0,47/255.0,39/255.0),\n",
    "    'stalk' = rgb(100/255.0,127/255.0,164/255.0)\n",
    ")\n",
    "\n",
    "window = 1e7\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "            start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        T ~ 'coding'\n",
    "    ))\n",
    "\n",
    "p_segs = segs_all %>%\n",
    "#     arrange(study != 'Loh') %>%\n",
    "    mutate(study = factor(study, unique(study))) %>%\n",
    "    filter(phi >= 0.1 | is.na(phi)) %>%\n",
    "    filter(chrom != 23) %>%\n",
    "    group_by(patient_id, chrom) %>%\n",
    "    mutate(total_length = sum(size)) %>%\n",
    "    mutate(type2 = factor(type2, c('AMP', 'CNLOH', 'DEL'))) %>%\n",
    "    arrange(type2, round(start/1e7), -total_length) %>%\n",
    "    group_by(chrom, study) %>%\n",
    "    mutate(index = as.integer(factor(patient_id, unique(patient_id)))) %>% \n",
    "    filter(index < 100) %>%\n",
    "    ungroup() %>%\n",
    "    ggplot(\n",
    "        aes(x = start, xend = end, y = index, yend = index, color = type2)\n",
    "    ) +\n",
    "    geom_rect(\n",
    "        inherit.aes = F,\n",
    "        data = chrom_lens %>% filter(chrom != 23),\n",
    "        aes(xmin = 0, xmax = length, ymin = -2, ymax = -0.5),\n",
    "        color = 'black',\n",
    "        fill = 'white',\n",
    "        size = 1\n",
    "    ) +\n",
    "    geom_rect(\n",
    "        data = cytoband %>% filter(chrom != 23),\n",
    "        inherit.aes = F,\n",
    "        aes(xmin = start, xmax = end, ymin = -2, ymax = -0.5, fill = stain),\n",
    "        size = 1,\n",
    "        show.legend = FALSE\n",
    "    ) +\n",
    "    geom_segment(\n",
    "        lineend = 'round',\n",
    "        size = 0.75\n",
    "    ) +\n",
    "    scale_fill_manual(values = cyto_colors) +\n",
    "    theme_void() +\n",
    "    theme(\n",
    "        axis.text = element_blank(),\n",
    "        axis.ticks = element_blank(),\n",
    "        axis.title = element_blank(),\n",
    "        legend.position = 'right',\n",
    "        panel.spacing.y = unit(0.5, \"lines\"),\n",
    "        plot.margin = unit(c(0,0,0,0), \"lines\"),\n",
    "        panel.border = element_blank(),\n",
    "        strip.text.y = element_text(size = 14, hjust = 0.5, angle = 0),\n",
    "        strip.background.y = element_rect(size = 1, fill = NA),\n",
    "        strip.background.x = element_blank()\n",
    "    ) +\n",
    "    scale_y_discrete(expand = expansion(add = 1)) +\n",
    "    scale_x_discrete(expand = c(0.02,0)) +\n",
    "    facet_grid(study~chrom, space = 'free', scales = 'free', switch=\"both\") +\n",
    "    xlab('') +\n",
    "    ylab('') +\n",
    "    scale_color_manual(values = cnv_pal2, name = '')\n",
    "\n",
    "do_plot(p_segs, 'comparison_genomewide.pdf', w = 16, h = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  EDITED MS - Reviewer 2 \n",
    "acen = fread('/home/gaot/ref/acen.tsv')\n",
    "colnames(acen) = c('chrom', 'start', 'end', 'arm', 'type')\n",
    "chrom_lens = fread('/home/gaot/ref/chrom_lens.tsv') %>% filter(chrom != 24)\n",
    "\n",
    "acen = acen %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "    filter(chrom != 'Y') %>%\n",
    "    mutate(\n",
    "        chrom = ifelse(chrom == 'X', 23, chrom),\n",
    "        chrom = as.integer(chrom)\n",
    "    ) %>%\n",
    "    arrange(chrom) %>%\n",
    "    group_by(chrom) %>%\n",
    "    summarise(cen_start = min(start), cen_end = max(end), .groups = 'drop') %>%\n",
    "    ungroup()\n",
    "\n",
    "cytoband = fread('./cytoBand.txt') %>% setNames(c('chrom', 'start', 'end', 'name', 'stain'))\n",
    "\n",
    "cytoband = cytoband %>% \n",
    "    mutate(chrom = str_remove(chrom, 'chr')) %>%\n",
    "    mutate(chrom = ifelse(chrom == 'X', 23, chrom)) %>%\n",
    "    filter(chrom != 'Y') %>%\n",
    "    mutate(chrom = as.integer(chrom))\n",
    "\n",
    "cyto_colors = c(\n",
    "    'gpos100'= rgb(0/255.0,0/255.0,0/255.0),\n",
    "    'gpos'   = rgb(0/255.0,0/255.0,0/255.0),\n",
    "    'gpos75' = rgb(130/255.0,130/255.0,130/255.0),\n",
    "    'gpos66' = rgb(160/255.0,160/255.0,160/255.0),\n",
    "    'gpos50' = rgb(200/255.0,200/255.0,200/255.0),\n",
    "    'gpos33' = rgb(210/255.0,210/255.0,210/255.0),\n",
    "    'gpos25' = rgb(200/255.0,200/255.0,200/255.0),\n",
    "    'gvar'   = rgb(220/255.0,220/255.0,220/255.0),\n",
    "    'gneg'  = rgb(255/255.0,255/255.0,255/255.0),\n",
    "    'acen'  = rgb(217/255.0,47/255.0,39/255.0),\n",
    "    'stalk' = rgb(100/255.0,127/255.0,164/255.0)\n",
    ")\n",
    "\n",
    "window = 1e7\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "            start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        T ~ 'coding'\n",
    "    ))\n",
    "\n",
    "p_segs = segs_all %>%\n",
    "filter(study %in% c(\"This study\", \"Loh\")) %>%\n",
    "#     arrange(study != 'Loh') %>%\n",
    "    mutate(study = case_when(\n",
    "        study == \"Loh\" & phi < 0.1 ~ \"Loh <10%\",\n",
    "        study == \"Loh\" & phi >= 0.1 ~ \"Loh >=10%\",\n",
    "        T ~ \"This study\")  ) %>%\n",
    "    mutate(study = factor(study, unique(study))) %>%\n",
    "    #filter(phi >= 0.1 | is.na(phi)) %>%\n",
    "    filter(chrom != 23) %>%\n",
    "    group_by(patient_id, chrom) %>%\n",
    "    mutate(total_length = sum(size)) %>%\n",
    "    mutate(type2 = factor(type2, c('AMP', 'CNLOH', 'DEL'))) %>%\n",
    "    arrange(type2, round(start/1e7), -total_length) %>%\n",
    "    group_by(chrom, study) %>%\n",
    "    mutate(index = as.integer(factor(patient_id, unique(patient_id)))) %>% \n",
    "#     filter(index < 100) %>%\n",
    "    ungroup() %>%\n",
    "    ggplot(\n",
    "        aes(x = start, xend = end, y = index, yend = index, color = type2)\n",
    "    ) +\n",
    "    geom_rect(\n",
    "        inherit.aes = F,\n",
    "        data = chrom_lens %>% filter(chrom != 23),\n",
    "        aes(xmin = 0, xmax = length, ymin = -2, ymax = -0.5),\n",
    "        color = 'black',\n",
    "        fill = 'white',\n",
    "        size = 1\n",
    "    ) +\n",
    "    geom_rect(\n",
    "        data = cytoband %>% filter(chrom != 23),\n",
    "        inherit.aes = F,\n",
    "        aes(xmin = start, xmax = end, ymin = -2, ymax = -0.5, fill = stain),\n",
    "        size = 1,\n",
    "        show.legend = FALSE\n",
    "    ) +\n",
    "    geom_segment(\n",
    "        lineend = 'round',\n",
    "        size = 0.75\n",
    "    ) +\n",
    "    scale_fill_manual(values = cyto_colors) +\n",
    "    theme_void() +\n",
    "    theme(\n",
    "        axis.text = element_blank(),\n",
    "        axis.ticks = element_blank(),\n",
    "        axis.title = element_blank(),\n",
    "        legend.position = 'right',\n",
    "        panel.spacing.y = unit(0.5, \"lines\"),\n",
    "        plot.margin = unit(c(0,0,0,0), \"lines\"),\n",
    "        panel.border = element_blank(),\n",
    "        strip.text.y = element_text(size = 14, hjust = 0.5, angle = 0),\n",
    "        strip.background.y = element_rect(size = 1, fill = NA),\n",
    "        strip.background.x = element_blank()\n",
    "    ) +\n",
    "    scale_y_discrete(expand = expansion(add = 1)) +\n",
    "    scale_x_discrete(expand = c(0.02,0)) +\n",
    "    facet_grid(study~chrom, space = 'free', scales = 'free', switch=\"both\") +\n",
    "    xlab('') +\n",
    "    ylab('') +\n",
    "    scale_color_manual(values = cnv_pal2, name = '')\n",
    "\n",
    "do_plot(p_segs, 'comparison_genomewide_MS_R2_10p_h9.pdf', w = 16, h = 9)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GM characterstics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = M_long %>% \n",
    "    filter(ch_nonsilent == 1) %>%\n",
    "    rowwise() %>%\n",
    "    mutate(\n",
    "        cis_cnv = any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "            start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    filter(cis_cnv == 0)\n",
    "\n",
    "p = ggplot(\n",
    "        D,\n",
    "        aes(x = VAF_N * 2)\n",
    "    ) +\n",
    "    geom_density(\n",
    "        trim = TRUE,\n",
    "        alpha = 0.5,\n",
    "        fill = sm_col,\n",
    "        size = 0.5\n",
    "    ) +\n",
    "    geom_density(\n",
    "        trim = TRUE,\n",
    "        inherit.aes = F,\n",
    "        data = segs_filtered,\n",
    "        mapping = aes(x = phi),\n",
    "        fill = cnv_col,\n",
    "        alpha = 0.5,\n",
    "        size = 0.5\n",
    "    ) +\n",
    "    xlim(0,1) +\n",
    "    theme_classic() +\n",
    "    xlab('Estimated cell fraction')\n",
    "\n",
    "do_plot(p, 'sensitivity.pdf', w = 2, h = 1.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_long %>% filter(ch_nonsilent == 1) %>% dim\n",
    "M_long %>% filter(ch_nonsilent == 1) %>% pull(VAF_N) %>% median\n",
    "M_long %>% filter(ch_nonsilent == 1) %>% pull(VAF_N) %>% range\n",
    "D %>% pull(VAF_N) %>% {.*2} %>% range\n",
    "M_wide %>% count(ch_nonsilent) %>% mutate(prop = n/sum(n))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "font_size = 8\n",
    "\n",
    "M_long_nonsilent = M_long %>% \n",
    "    filter(!VariantClass %in% c('Silent', 'Intron')) %>%\n",
    "    mutate(\n",
    "        VariantClass3 = case_when(\n",
    "            VariantClass == 'Frame_Shift_Del' ~ 'Frameshift indel',\n",
    "            VariantClass == 'Frame_Shift_Ins' ~ 'Frameshift indel',\n",
    "            VariantClass == 'In_Frame_Del' ~ 'Inframe indel',\n",
    "            VariantClass == 'In_Frame_Ins' ~ 'Inframe indel',\n",
    "            VariantClass == 'Splice_Site' ~ 'Splice or other',\n",
    "            VariantClass == 'Splice_Region' ~ 'Splice or other',\n",
    "            VariantClass == \"5'Flank\" ~ 'Splice or other',\n",
    "            VariantClass == 'Translation_Start_Site' ~ 'Splice or other',\n",
    "            VariantClass == 'Nonstop_Mutation' ~ 'Splice or other',\n",
    "            VariantClass == 'Missense_Mutation' ~ 'Missense',\n",
    "            VariantClass == 'Nonsense_Mutation' ~ 'Nonsense',\n",
    "            T ~ VariantClass)\n",
    "    ) %>%\n",
    "    mutate(VariantClass3 = factor(VariantClass3))\n",
    "\n",
    "panel_theme = theme(\n",
    "    legend.direction = 'vertical',\n",
    "    legend.position=\"top\", \n",
    "    legend.text = element_text(size=12),\n",
    "    plot.margin = margin(t = 0, r = 0, b = 0, l = 0),\n",
    "    axis.text = element_text(size = font_size)\n",
    "  )\n",
    "\n",
    "# All patients\n",
    "\n",
    "gene_list = M_long_nonsilent %>% count(Gene) %>% arrange(-n) %>% .$Gene %>% unique %>% .[1:30]\n",
    "\n",
    "p_all = ggplot(\n",
    "        M_long_nonsilent %>% filter(Gene %in% gene_list) %>% \n",
    "        mutate(Gene = as.integer(factor(Gene, gene_list))) %>%\n",
    "        count(Gene, VariantClass3),\n",
    "        aes(x = Gene, y = n, fill = VariantClass3)\n",
    "    ) + \n",
    "    geom_bar(stat = 'identity', color = 'black', size = 0.2, width = 0.9) +\n",
    "    ylab(\"Number of Mutations\") +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        panel.grid.major = element_blank(), \n",
    "        panel.border = element_blank(),\n",
    "        axis.line = element_line(colour = \"white\"),\n",
    "        legend.title = element_blank()\n",
    "    ) +\n",
    "    theme(\n",
    "        axis.text.x = element_text(angle = 45, hjust = 1),\n",
    "        legend.position = 'right',\n",
    "        strip.background = element_rect(fill = NA, size = 1, color = 'gray')\n",
    "    ) +\n",
    "    facet_zoom(xlim = c(5.5 + 0.1, 29.5 - 0.1), zoom.size = 1, ylim = c(0,500), horizontal = F, show.area = F) +\n",
    "    scale_x_continuous(labels = gene_list, breaks = 1:length(gene_list), expand = expansion(add=0.1)) +\n",
    "    scale_fill_discrete(drop = F)\n",
    "\n",
    "# CNV subset\n",
    "\n",
    "gene_list = M_long_nonsilent %>% \n",
    "    filter(!VariantClass %in% c('Silent', 'Intron')) %>%\n",
    "    filter(dmp_patient_id %in% (M_wide %>% filter(ch_cnv == 1) %>% pull(dmp_patient_id))) %>% \n",
    "    count(Gene) %>% arrange(-n) %>% .$Gene %>% unique %>% .[1:30]\n",
    "\n",
    "p_cnv = ggplot(\n",
    "    M_long_nonsilent %>% \n",
    "    filter(Gene %in% gene_list) %>% \n",
    "    filter(dmp_patient_id %in% (M_wide %>% filter(ch_cnv == 1) %>% pull(dmp_patient_id))) %>% \n",
    "    mutate(Gene = as.integer(factor(Gene, gene_list))) %>%\n",
    "    count(Gene, VariantClass3),\n",
    "    aes(x = Gene, y = n, fill = VariantClass3)\n",
    ") + \n",
    "geom_bar(stat = 'identity', color = 'black', size = 0.2) +\n",
    "ylab(\"Number of Mutations\") +\n",
    "theme_classic() +\n",
    "theme(\n",
    "    panel.grid.major = element_blank(), \n",
    "    panel.border = element_blank(),\n",
    "    axis.line = element_line(colour = \"white\"),\n",
    "    legend.title = element_blank()\n",
    ") +\n",
    "theme(\n",
    "    axis.text.x = element_text(angle = 45, hjust = 1),\n",
    "    legend.position = 'right',\n",
    "    strip.background = element_rect(fill = NA, size = 1, color = 'gray')\n",
    ") +\n",
    "facet_zoom(xlim = c(5.5 + 0.1, 29.5 - 0.1), zoom.size = 1, ylim = c(0,15), horizontal = F, show.area = F) +\n",
    "scale_x_continuous(labels = gene_list, breaks = 1:length(gene_list), expand = expansion(add=0.1)) +\n",
    "scale_fill_discrete(drop = F)\n",
    "\n",
    "do_plot(p_all, 'sm_bar_all.pdf', w = 8, h = 4)\n",
    "do_plot(p_cnv, 'sm_bar_ai.pdf', w = 8, h = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = M_wide %>%\n",
    "    filter(ch_nonsilent == 1) %>%\n",
    "    mutate(mutnum_nonsilent = case_when(\n",
    "        mutnum_nonsilent < 5 ~ as.character(mutnum_nonsilent), T ~ '5+')\n",
    "    ) %>%\n",
    "    count(mutnum_nonsilent) %>%\n",
    "    mutate(prop = n/sum(n))\n",
    "\n",
    "display(D)\n",
    "\n",
    "p_n_all = ggplot(\n",
    "        D,\n",
    "        aes(x = mutnum_nonsilent, y = n)\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    geom_bar(stat = 'identity', position = 'stack') +\n",
    "    scale_fill_manual(values = cnv_pal) +\n",
    "    theme(\n",
    "        legend.position = 'none',\n",
    "        axis.title = element_text(size = 8)\n",
    "    ) +\n",
    "    xlab('Events per patient') +\n",
    "    ylab('Number of patients')\n",
    "#     scale_y_continuous(expand = expansion(add = 0))\n",
    "\n",
    "\n",
    "D = M_wide %>% \n",
    "    filter(ch_nonsilent == 1) %>%\n",
    "    filter(ch_cnv == 1) %>%\n",
    "    mutate(mutnum_nonsilent = case_when(\n",
    "        mutnum_nonsilent < 5 ~ as.character(mutnum_nonsilent), T ~ '5+')\n",
    "    ) %>%\n",
    "    count(mutnum_nonsilent)\n",
    "\n",
    "p_n_ai = ggplot(\n",
    "        D,\n",
    "        aes(x = mutnum_nonsilent, y = n)\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    geom_bar(stat = 'identity', position = 'stack') +\n",
    "    scale_fill_manual(values = cnv_pal) +\n",
    "    theme(\n",
    "        legend.position = 'none',\n",
    "        axis.title = element_text(size = 8)\n",
    "    ) +\n",
    "    xlab('Events per patient') +\n",
    "    ylab('Number of patients')\n",
    "#     scale_y_continuous(expand = expansion(add = 0))\n",
    "\n",
    "(p_n_all / p_n_ai) %>%\n",
    "do_plot('sm_num.pdf', w = 2, h = 3.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 2, repr.plot.res = 300)\n",
    "\n",
    "D = M_wide %>% \n",
    "    mutate(age_bin = cut(age, c(0, seq(20, 80, 10), 100))) %>%\n",
    "    group_by(age_bin) %>%\n",
    "    summarise(\n",
    "        `GM VAF>2%` = sum(ch_nonsilent & VAF_nonsilent > 0.02)/n(),\n",
    "        `GM VAF>5%` = sum(ch_nonsilent & VAF_nonsilent > 0.05)/n(),\n",
    "        `GM VAF>10%` = sum(ch_nonsilent & VAF_nonsilent > 0.1)/n(),\n",
    "        n = n()\n",
    "    ) %>% \n",
    "    melt(measure.vars = c('GM VAF>2%', 'GM VAF>5%', 'GM VAF>10%'), variable.name = 'threshold', value.name = 'prop') %>%\n",
    "    rbind(\n",
    "        M_wide %>% \n",
    "        mutate(age_bin = cut(age, c(0, seq(20, 80, 10), 100))) %>%\n",
    "        group_by(age_bin) %>%\n",
    "        summarise(\n",
    "            prop = sum(ch_cnv)/n(),\n",
    "            n = n()\n",
    "        ) %>%\n",
    "        mutate(threshold = 'mCA')\n",
    "    )\n",
    "\n",
    "p_age = ggplot(\n",
    "    D,\n",
    "    aes(x = age_bin, y = prop, group = threshold, color = threshold)\n",
    ") +\n",
    "geom_line() +\n",
    "theme_classic() +\n",
    "geom_point() +\n",
    "ylab('Proportion of subjects') +\n",
    "xlab('Age') +\n",
    "labs(color = '')\n",
    "\n",
    "p_age %>%\n",
    "do_plot('sm_age.pdf', w = 5, h = 2)\n",
    "\n",
    "D %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/SFig_12B.tsv\", sep = \"\\t\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 2, repr.plot.res = 300)\n",
    "D = M_wide %>% \n",
    "    mutate(age_bin = cut(age, c(0, seq(20, 80, 10), 100))) %>%\n",
    "    group_by(age_bin) %>%\n",
    "    summarise(\n",
    "        `GM VAF>12.5%` = sum(ch_nonsilent & VAF_nonsilent > 0.125)/n(),\n",
    "        n = n()\n",
    "    ) %>% \n",
    "    melt(measure.vars = c('GM VAF>12.5%'), variable.name = 'threshold', value.name = 'prop') %>%\n",
    "    rbind(\n",
    "        M_wide %>% \n",
    "        mutate(age_bin = cut(age, c(0, seq(20, 80, 10), 100))) %>%\n",
    "        group_by(age_bin) %>%\n",
    "        summarise(\n",
    "            prop = sum(ch_cnv )/n(),\n",
    "            n = n()\n",
    "        ) %>%\n",
    "        mutate(threshold = 'mCA Cell Fraction>25%')\n",
    "    )\n",
    "p_age = ggplot(\n",
    "    D,\n",
    "    aes(x = age_bin, y = prop,  group = threshold , color = threshold)\n",
    ") +\n",
    "geom_line() +\n",
    "theme_classic() +\n",
    "geom_point() + theme(legend.position=\"top\") + theme(legend.title = element_blank()) + \n",
    "ylab('Proportion of subjects') +\n",
    "xlab('Age') + ylim(0,0.2)\n",
    "labs(color = '')\n",
    "p_age %>%\n",
    "do_plot('sm_age_MS_edit.pdf', w = 4, h = 3)\n",
    "\n",
    "D %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/SFig_12C.tsv\", sep = \"\\t\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Associations with exposures and gene mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Venn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(VennDiagram)\n",
    "library(RColorBrewer)\n",
    "futile.logger::flog.threshold(futile.logger::ERROR, name = \"VennDiagramLogger\")\n",
    "\n",
    "venn.diagram(\n",
    "    x = list(M_wide[M_wide$ch_cnv == 1,][['dmp_patient_id']], M_wide[M_wide$ch_nonsilent == 1,][['dmp_patient_id']]),\n",
    "    category.names = c('mCA', 'GM'),\n",
    "    \n",
    "    # output features\n",
    "    filename = './figures/venn.tiff',\n",
    "    imagetype=\"tiff\",\n",
    "    height = 800, \n",
    "    width = 800, \n",
    "#     resolution = 300,\n",
    "    compression = \"lzw\",\n",
    "    output = TRUE,\n",
    "    margin = 0.1,\n",
    "    \n",
    "    # circles\n",
    "    lwd = 1,\n",
    "    fill = c(cnv_col, sm_col),\n",
    "\n",
    "    # Numbers\n",
    "    cex = 0.6,\n",
    "    fontface = \"bold\",\n",
    "    fontfamily = \"sans\",\n",
    "    ext.dist = 0.06,\n",
    "    ext.line.lwd = 0.8,\n",
    "    \n",
    "    # Set names\n",
    "    cat.cex = 0.55,\n",
    "    cat.fontface = \"bold\",\n",
    "    cat.default.pos = \"outer\",\n",
    "    cat.fontfamily = \"sans\",\n",
    "    cat.dist = 0.1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(table(M_wide$ch_cnv, M_wide$ch_nonsilent))\n",
    "fisher.test(table(M_wide$ch_cnv, M_wide$ch_nonsilent))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "glm(\n",
    "    formula = ch_cnv ~ ch_nonsilent + age + age_sq,\n",
    "    data = M_wide %>% mutate(age_sq = age^2)\n",
    ") %>%\n",
    "sjPlot::get_model_data(type = 'est')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Volcano"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = M_long %>% count(Gene) %>% filter(n >= 30) %>% pull(Gene)\n",
    "\n",
    "res = data.frame()\n",
    "\n",
    "for (gene in genes) {\n",
    "    \n",
    "    test = fisher.test(table(M_wide[[gene]], M_wide[['ch_cnv']]))\n",
    "    \n",
    "    res = rbind(\n",
    "        res,\n",
    "        data.frame(\n",
    "            'p.value' = test$p.value,\n",
    "            'Gene' = gene,\n",
    "            'OR' = test$estimate\n",
    "        )\n",
    "    )\n",
    "}\n",
    "\n",
    "res = res %>% mutate(q = p.adjust(p.value, n = length(genes), method = 'BH'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_volcano = ggplot(\n",
    "    res %>% mutate(label = ifelse(q < 0.05, as.character(Gene), NA)),\n",
    "    aes(x = OR, y = -log(q), color = q < 0.05, label = label)\n",
    ") +\n",
    "geom_hline(yintercept = -log(0.05), linetype = 'dashed') +\n",
    "geom_point() +\n",
    "geom_text_repel(size = 2.5, seed = 3) +\n",
    "theme_classic() +\n",
    "scale_color_manual(values = c('gray', 'darkred')) +\n",
    "theme(\n",
    "    legend.position = 'none'\n",
    ")\n",
    "\n",
    "n_cnv = sum(M_wide$ch_cnv)\n",
    "n_normal = nrow(M_wide) - sum(M_wide$ch_cnv)\n",
    "\n",
    "genes = M_long %>% count(Gene) %>% arrange(-n) %>% filter(n > 30) %>% pull(Gene)\n",
    "\n",
    "D = M_long %>%\n",
    "    filter(ch_nonsilent == 1) %>%\n",
    "    count(Gene, ch_cnv) %>%\n",
    "    tidyr::complete(Gene, ch_cnv, fill = list(n = 0)) %>% \n",
    "    mutate(n_total = ifelse(ch_cnv == 1, n_cnv, n_normal)) %>% \n",
    "    group_by(ch_cnv) %>%\n",
    "    mutate(prop = n/n_total) %>%\n",
    "    left_join(\n",
    "        res,\n",
    "        by = 'Gene'\n",
    "    ) %>%\n",
    "    filter(q < 0.05) %>%\n",
    "    arrange(desc(n)) %>%\n",
    "    mutate(Gene = factor(Gene, Gene))\n",
    "\n",
    "p_genes = ggplot(\n",
    "    D %>% mutate(ch_cnv = ifelse(ch_cnv == 1, 'mCA+', 'mCA-')),\n",
    "    aes(x = Gene, y = prop, fill = ch_cnv)\n",
    ") +\n",
    "theme_classic() +\n",
    "geom_col(\n",
    "    position = position_dodge(),\n",
    "    width = 0.75,\n",
    "    size = 0.2,\n",
    "    color = 'black'\n",
    ") +\n",
    "scale_y_continuous(expand = expansion(add = 0.001), labels = abs) +\n",
    "scale_fill_manual(values = c(cnv_col, 'gainsboro')) +\n",
    "theme(\n",
    "    axis.text.x = element_text(angle = 30, hjust = 1, size = 7),\n",
    "    legend.title = element_blank(),\n",
    "    legend.position = 'right'\n",
    ") +\n",
    "ylab('Proportion of patients') +\n",
    "xlab('')\n",
    "\n",
    "panel = (p_volcano | p_genes) + plot_layout(widths = c(1,3))\n",
    "\n",
    "do_plot(panel, 'sm_enrichment.pdf', w = 7.5, h = 2.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mCA ~ VAF & Mutnum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = M_wide %>% \n",
    "    mutate(mutnum_bin = ifelse(mutnum_nonsilent >= 4, '4+', as.character(mutnum_nonsilent))) %>%\n",
    "    group_by(mutnum_bin) %>%\n",
    "    summarise(\n",
    "        n_cnv = sum(ch_cnv),\n",
    "        total = n(),\n",
    "        prop = n_cnv/total,\n",
    "        `.groups` = 'drop'\n",
    "    ) %>%\n",
    "    mutate(\n",
    "        prop_lower = qbinom(p = 0.05, size = total, prob = prop)/total,\n",
    "        prop_upper = qbinom(p = 0.95, size = total, prob = prop)/total,\n",
    "    )\n",
    "\n",
    "p_mutnum = ggplot(\n",
    "        D,\n",
    "        aes(x = mutnum_bin, y = prop, group = '', ymin = prop_lower, ymax = prop_upper)\n",
    "    ) +\n",
    "    geom_ribbon(fill = 'royalblue', alpha = 0.3) +\n",
    "    geom_line(color = 'black') +\n",
    "    geom_point(color = 'black') +\n",
    "    theme_classic() +\n",
    "    ylab('Prop. mCA') +\n",
    "    xlab('Number of mutations') +\n",
    "    scale_y_continuous(expand = expansion(mult = 0.05), limits = c(0,NA))\n",
    "\n",
    "display(D)\n",
    "D %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/Fig_2B.tsv\", sep = \"\\t\", quote = FALSE)\n",
    "\n",
    "D = M_wide %>%\n",
    "    mutate(VAF_bin = case_when(\n",
    "        VAF_trans == 0 ~ '<2%',\n",
    "        VAF_trans < 0.05 ~ '2-5%',\n",
    "        VAF_trans < 0.1 ~ '5-10%',\n",
    "        VAF_trans < 0.2 ~ '10-20%',\n",
    "        T ~ '>20%'\n",
    "    )) %>%\n",
    "    arrange(VAF_trans) %>%\n",
    "    mutate(VAF_bin = factor(VAF_bin, unique(VAF_bin))) %>%\n",
    "    group_by(VAF_bin) %>%\n",
    "    summarise(\n",
    "        n_cnv = sum(ch_cnv),\n",
    "        total = n(),\n",
    "        prop = n_cnv/total,\n",
    "        `.groups` = 'drop'\n",
    "    ) %>%\n",
    "    mutate(\n",
    "        prop_lower = qbinom(p = 0.05, size = total, prob = prop)/total,\n",
    "        prop_upper = qbinom(p = 0.95, size = total, prob = prop)/total,\n",
    "    )\n",
    "\n",
    "display(D)\n",
    "D %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/Fig_2C.tsv\", sep = \"\\t\", quote = FALSE)\n",
    "             \n",
    "p_vaf = ggplot(\n",
    "        D,\n",
    "        aes(x = VAF_bin, y = prop, ymin = prop_lower, ymax = prop_upper, group = '')\n",
    "    ) +\n",
    "    geom_ribbon(fill = 'royalblue', alpha = 0.3) +\n",
    "    geom_line(color = 'black') +\n",
    "    geom_point(color = 'black') +\n",
    "    theme_classic() +\n",
    "    ylab('Prop. mCA') +\n",
    "    xlab('Max VAF') +\n",
    "    scale_y_continuous(expand = expansion(mult = 0.05), limits = c(0,NA)) +\n",
    "    theme(axis.text.x = element_text(angle = 30, hjust = 1))\n",
    "\n",
    "panel = p_mutnum | p_vaf\n",
    "\n",
    "do_plot(panel, 'mutnum_vaf.pdf', w = 4, h = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = M_wide %>% \n",
    "    mutate(mutnum_bin = ifelse(mutnum_nonsilent >= 3, '3+', as.character(mutnum_nonsilent))) %>%\n",
    "    group_by(mutnum_bin) %>%\n",
    "    summarise(\n",
    "        n_cnv = sum(ch_cnv),\n",
    "        total = n(),\n",
    "        prop = n_cnv/total,\n",
    "        `.groups` = 'drop'\n",
    "    ) %>%\n",
    "    mutate(\n",
    "        prop_lower = qbinom(p = 0.05, size = total, prob = prop)/total,\n",
    "        prop_upper = qbinom(p = 0.95, size = total, prob = prop)/total,\n",
    "    )\n",
    "\n",
    "D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(M_wide$ch_cnv)/nrow(M_wide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>%\n",
    "    mutate(mutnum_bin = ifelse(mutnum_trans >= 4, '4+', as.character(mutnum_trans))) %>%\n",
    "    filter(smoke != 'missing') %>%\n",
    "    filter(mutnum_bin != '0') %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ age10 + mutnum_trans + VAF_trans,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est') %>%\n",
    "    mutate(feature = 'mutnum')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 1.5, repr.plot.res = 300)\n",
    "\n",
    "M_wide %>%\n",
    "mutate_at(\n",
    "    c('ch_my_pd', 'ch_cnv', 'ch_all'),\n",
    "    as.character\n",
    ") %>%\n",
    "mutate(ch_status = case_when(\n",
    "    ch_my_pd == 1 ~ 'ch_my_pd',\n",
    "    ch_all == 1 ~ 'ch_other',\n",
    "    T ~ 'none'\n",
    ")) %>%\n",
    "mutate(ch_status = factor(ch_status, rev(unique(ch_status)))) %>%\n",
    "count(ch_status, ch_cnv) %>%\n",
    "group_by(ch_cnv) %>%\n",
    "mutate(prop = n/sum(n)) %>%\n",
    "ungroup() %>%\n",
    "ggplot(\n",
    "    aes(x = ch_cnv, y = prop, fill = ch_status, label = n)\n",
    ") +\n",
    "geom_col() +\n",
    "theme_classic() +\n",
    "coord_flip() + \n",
    "geom_text(position = position_stack(vjust = 0.5), color = 'white') +\n",
    "scale_y_continuous(expand = c(0,0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use stacked bar instead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 7, repr.plot.height = 3, repr.plot.res = 300)\n",
    "\n",
    "germline_cause = c('11=', '1=', '15+', '23-', '14=', '16=', '8=', '10-', '15-', '15=')\n",
    "\n",
    "# https://www.nejm.org/doi/full/10.1056/NEJMoa1516192\n",
    "aml_cnvs = c('7-', '8+', '5-', '17-', '9-', '12-', '21+')\n",
    "\n",
    "# https://www.nature.com/articles/nature14666\n",
    "cll_cnvs = c('12+', '13-', '11-', '17-', '6-')\n",
    "\n",
    "# elsa's data\n",
    "mds_cnvs = c('5-', '8+', '7-', '20-', '11-', '17-', '17=', '4=', '13-', '21+')\n",
    "\n",
    "# \n",
    "mpn_cnvs = c('9=', '20-', '1=', '17=', '7=', '14=')\n",
    "\n",
    "D = segs_filtered %>% \n",
    "    mutate_at(\n",
    "        c('ch_sm'),\n",
    "        as.character\n",
    "    ) %>%\n",
    "    mutate(mutnum = ifelse(mutnum_nonsilent > 2, '3+', as.character(mutnum_nonsilent))) %>%\n",
    "    mutate(germline_cause = cnv_label %in% germline_cause) %>%\n",
    "    group_by(cnv_label) %>%\n",
    "    mutate(\n",
    "        prop_sm = sum(ch_nonsilent)/n(),\n",
    "        mutnum_median = mean(mutnum_nonsilent),\n",
    "        total = n()\n",
    "    ) %>%\n",
    "    filter(total >= 4) %>%\n",
    "    ungroup() %>%\n",
    "    arrange(mutnum_median) %>%\n",
    "    mutate(cnv_label = factor(cnv_label, unique(cnv_label))) %>%\n",
    "    count(mutnum, cnv_label) %>%\n",
    "    group_by(cnv_label) %>%\n",
    "    mutate(\n",
    "        total = sum(n),\n",
    "        prop = n/total\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    arrange(cnv_label) %>%\n",
    "    mutate(\n",
    "        category = case_when(\n",
    "            cnv_label %in% germline_cause ~ 'germline',\n",
    "            cnv_label %in% c(aml_cnvs, cll_cnvs) ~ 'CLL/AML',\n",
    "            T ~ 'none')\n",
    "    )\n",
    "\n",
    "ggplot(\n",
    "    D,\n",
    "    aes(x = cnv_label, y = prop, fill = mutnum, label = total)\n",
    ") +\n",
    "theme_classic() +\n",
    "geom_col(size = 0) +\n",
    "scale_y_continuous(expand = c(0,0), limits = c(0, 1.05)) +\n",
    "theme(\n",
    "    axis.text.x = element_text(\n",
    "        angle = 30, hjust = 0.5, size = 9,\n",
    "#         color = case_when(\n",
    "#             unique(D$cnv_label) %in% germline_cause ~ 'blue',\n",
    "#             unique(D$cnv_label) %in% c(aml_cnvs, cll_cnvs) ~ 'red',\n",
    "#             T ~ 'black'),\n",
    "        face = 'bold'),\n",
    "    legend.box = \"horizontal\",\n",
    "    legend.position = 'top'\n",
    ") +\n",
    "# guides(\n",
    "#     colour = guide_legend(override.aes = list(size = 1, fill = NA)),\n",
    "#     fill = guide_legend(override.aes = list(size = 2))\n",
    "# ) +\n",
    "# annotate('text', label = D$total, x = D$cnv_label, y = 1, vjust = -0.1, size = 2) +\n",
    "xlab('') +\n",
    "ylab('proportion') +\n",
    "scale_fill_brewer(palette = 'Reds', name = '#GM')\n",
    "# scale_color_manual(values = c('germline' = 'blue', 'CLL/AML' = 'red', 'none' = 'black'), name = 'etiology')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exposures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_c = fread('./data/clinicalreview/chemoclass_jan2019_revised.csv') %>% pull(drugclass_c) %>% unique\n",
    "class_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 4, repr.plot.height = 2.5, repr.plot.res = 500)\n",
    "\n",
    "options(warn=-1)\n",
    "\n",
    "summ_cnv = M_wide %>%\n",
    "    filter(smoke != 'missing') %>%\n",
    "    mutate(smoke = as.integer(smoke)) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "summ_cnv_therapy = M_wide %>%\n",
    "    filter(therapy_detailed & smoke != 'missing') %>%\n",
    "    filter(DateofProcedure == DOP_kelly) %>%\n",
    "    group_by(tumor_type) %>% mutate(n_tumor_type = n()) %>% ungroup() %>%\n",
    "    mutate(tumor_type = ifelse(n_tumor_type < 100, 'Other', tumor_type)) %>%\n",
    "    mutate(\n",
    "#         eqd_3_t = as.character(eqd_3_t),\n",
    "#         pct_cytotoxic_therapy = as.character(pct_cytotoxic_therapy),\n",
    "        smoke = as.integer(smoke)\n",
    "    ) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke + XRT + ind_cytotoxic_therapy,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "options(warn=0)\n",
    "\n",
    "rbind(\n",
    "    summ_cnv %>% mutate(cohort = 'full'),\n",
    "    summ_cnv_therapy %>% mutate(cohort = 'therapy')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### therapy mCA \n",
    "\n",
    "options(repr.plot.width = 4, repr.plot.height = 2.5, repr.plot.res = 500)\n",
    "\n",
    "options(warn=-1)\n",
    "\n",
    "summ_cnv = M_wide %>%\n",
    "    filter(smoke != 'missing') %>%\n",
    "    mutate(smoke = as.integer(smoke)) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke ,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "\n",
    "\n",
    "summ_cnv_therapy = M_wide %>%\n",
    "    filter(therapy_detailed & smoke != 'missing') %>%\n",
    "    filter(DateofProcedure == DOP_kelly) %>%\n",
    "    group_by(tumor_type) %>% mutate(n_tumor_type = n()) %>% ungroup() %>%\n",
    "    mutate(tumor_type = ifelse(n_tumor_type < 100, 'Other', tumor_type)) %>%\n",
    "    mutate(\n",
    "#         eqd_3_t = as.character(eqd_3_t),\n",
    "#         pct_cytotoxic_therapy = as.character(pct_cytotoxic_therapy),\n",
    "        smoke = as.integer(smoke)\n",
    "    ) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke + XRT + ind_cytotoxic_therapy,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "options(warn=0)\n",
    "\n",
    "summ_cnv_therapy\n",
    "\n",
    "p_forest = rbind(\n",
    "        summ_cnv %>% mutate(cohort = paste0('Full (n=', nrow(M_wide[M_wide$smoke != 'missing',]), ')')),\n",
    "        summ_cnv_therapy %>% mutate(cohort = paste0('Therapy (n=', sum(M_wide[M_wide$smoke != 'missing',]$therapy_detailed), ')'))\n",
    "    ) %>%\n",
    "    arrange(cohort) %>%\n",
    "    mutate(cohort = factor(cohort, rev(unique(cohort)))) %>%\n",
    "    filter(!str_detect(term, 'tumor_type')) %>%\n",
    "    mutate(term = factor(term, c('age10', 'GenderMale', 'race_b', 'smoke', 'XRT', 'ind_cytotoxic_therapy'))) %>%\n",
    "    arrange(term) %>%\n",
    "    mutate(\n",
    "        term = c('age10' = 'Age', 'GenderMale' = 'Male Gender', 'race_b' = 'White Race',\n",
    "                'smoke' = 'Ever Smoker', 'XRT' = 'External Beam Radiation', 'ind_cytotoxic_therapy' = 'Cytotoxic Therapy')[term] \n",
    "    ) %>%\n",
    "    mutate(term = factor(term, rev(unique(term)))) %>%\n",
    "    mutate(format.p.value = format.pval(p.value, digits = 2)) %>% \n",
    "    plot_forest(\n",
    "        x = 'term',\n",
    "        label = 'format.p.value',\n",
    "        eb_w = 0,\n",
    "        eb_s = 1,\n",
    "        ps = 2,\n",
    "        or_s = 3,\n",
    "        nudge = -0.45,\n",
    "        OR = T,\n",
    "        col = 'cohort',\n",
    "        dodge_width = 0.7\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    ylab('OR') +\n",
    "    xlab('') +\n",
    "    scale_color_discrete(name = 'Cohort')\n",
    "\n",
    "do_plot(p_forest, 'MS_R3_forest_therapy_pvalue.pdf', w = 6, h = 4.3)\n",
    "\n",
    "\n",
    "rbind(\n",
    "        summ_cnv %>% mutate(cohort = paste0('Full (n=', nrow(M_wide[M_wide$smoke != 'missing',]), ')')),\n",
    "        summ_cnv_therapy %>% mutate(cohort = paste0('Therapy (n=', sum(M_wide[M_wide$smoke != 'missing',]$therapy_detailed), ')'))\n",
    "    ) %>%\n",
    "    arrange(cohort) %>%\n",
    "    mutate(cohort = factor(cohort, rev(unique(cohort)))) %>%\n",
    "    filter(!str_detect(term, 'tumor_type')) %>%\n",
    "    mutate(term = factor(term, c('age10', 'GenderMale', 'race_b', 'smoke', 'XRT', 'ind_cytotoxic_therapy'))) %>%\n",
    "    arrange(term) %>%\n",
    "    mutate(\n",
    "        term = c('age10' = 'Age', 'GenderMale' = 'Male Gender', 'race_b' = 'White Race',\n",
    "                'smoke' = 'Ever Smoker', 'XRT' = 'External Beam Radiation', 'ind_cytotoxic_therapy' = 'Cytotoxic Therapy')[term]\n",
    "    ) %>%\n",
    "    mutate(term = factor(term, rev(unique(term)))) %>%\n",
    "    mutate(format.p.value = format.pval(p.value, digits = 2)) %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/SFig10.csv\", sep = \"\\t\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS edit for R3 rebuttal\n",
    "\n",
    "options(repr.plot.width = 4, repr.plot.height = 2.5, repr.plot.res = 500)\n",
    "\n",
    "options(warn=-1)\n",
    "\n",
    "M_wide = M_wide %>% mutate(ch_ddr = TP53 | PPM1D | CHEK2)\n",
    "\n",
    "summ_cnv = M_wide %>%\n",
    "    filter(smoke != 'missing') %>%\n",
    "    mutate(smoke = as.integer(smoke)) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke + ch_ddr,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "summ_cnv_therapy = M_wide %>%\n",
    "    filter(therapy_detailed & smoke != 'missing') %>%\n",
    "    filter(DateofProcedure == DOP_kelly) %>%\n",
    "    group_by(tumor_type) %>% mutate(n_tumor_type = n()) %>% ungroup() %>%\n",
    "    mutate(tumor_type = ifelse(n_tumor_type < 100, 'Other', tumor_type)) %>%\n",
    "    mutate(\n",
    "#         eqd_3_t = as.character(eqd_3_t),\n",
    "#         pct_cytotoxic_therapy = as.character(pct_cytotoxic_therapy),\n",
    "        smoke = as.integer(smoke)\n",
    "    ) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke + XRT + ind_cytotoxic_therapy + ch_ddr,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "options(warn=0)\n",
    "\n",
    "rbind(\n",
    "    summ_cnv %>% mutate(cohort = 'full'),\n",
    "    summ_cnv_therapy %>% mutate(cohort = 'therapy')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS edit for R3 rebuttal\n",
    "\n",
    "options(repr.plot.width = 4, repr.plot.height = 2.5, repr.plot.res = 500)\n",
    "\n",
    "options(warn=-1)\n",
    "\n",
    "M_wide = M_wide %>% mutate(ch_ddr = TP53 | PPM1D | CHEK2) %>% mutate( ch_ddr = as.integer(ch_ddr))\n",
    "\n",
    "summ_cnv = M_wide %>%\n",
    "    filter(smoke != 'missing') %>%\n",
    "    mutate(smoke = as.integer(smoke)) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke + ch_ddr,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "summ_cnv_therapy = M_wide %>%\n",
    "    filter(therapy_detailed & smoke != 'missing') %>%\n",
    "    filter(DateofProcedure == DOP_kelly) %>%\n",
    "    group_by(tumor_type) %>% mutate(n_tumor_type = n()) %>% ungroup() %>%\n",
    "    mutate(tumor_type = ifelse(n_tumor_type < 100, 'Other', tumor_type)) %>%\n",
    "    mutate(\n",
    "#         eqd_3_t = as.character(eqd_3_t),\n",
    "#         pct_cytotoxic_therapy = as.character(pct_cytotoxic_therapy),\n",
    "        smoke = as.integer(smoke)\n",
    "    ) %>%\n",
    "    glm(\n",
    "        ch_cnv_auto ~ Gender + race_b + age10 + smoke + XRT + ind_cytotoxic_therapy + ch_ddr,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "options(warn=0)\n",
    "\n",
    "rbind(\n",
    "    summ_cnv %>% mutate(cohort = 'full'),\n",
    "    summ_cnv_therapy %>% mutate(cohort = 'therapy')\n",
    ")\n",
    "\n",
    "p_forest = rbind(\n",
    "#         summ_cnv %>% mutate(cohort = paste0('Full (n=', nrow(M_wide[M_wide$smoke != 'missing',]), ')')),\n",
    "        summ_cnv_therapy %>% mutate(cohort = paste0('Therapy (n=', sum(M_wide[M_wide$smoke != 'missing',]$therapy_detailed), ')'))\n",
    "    ) %>%\n",
    "    arrange(cohort) %>%\n",
    "    mutate(cohort = factor(cohort, rev(unique(cohort)))) %>%\n",
    "    filter(!str_detect(term, 'tumor_type')) %>%\n",
    "    mutate(term = factor(term, c('age10', 'GenderMale', 'race_b', 'smoke', 'XRT', 'ind_cytotoxic_therapy', 'ch_ddr'))) %>%\n",
    "    arrange(term) %>%\n",
    "    mutate(\n",
    "        term = c('age10' = 'Age', 'GenderMale' = 'Male Gender', 'race_b' = 'White Race',\n",
    "                'smoke' = 'Ever Smoker', 'XRT' = 'External Beam Radiation', 'ind_cytotoxic_therapy' = 'Cytotoxic Therapy',  'ch_ddr' = 'CH in DDR Gene')[term] # 'ch_nonsilent'= 'CH Nonsilent', 'tumor_type' = 'Tumor Type',\n",
    "    ) %>%\n",
    "    mutate(term = factor(term, rev(unique(term)))) %>%\n",
    "    mutate(format.p.value = format.pval(p.value, digits = 2)) %>% \n",
    "    plot_forest(\n",
    "        x = 'term',\n",
    "        label = 'format.p.value',\n",
    "        eb_w = 0,\n",
    "        eb_s = 1,\n",
    "        ps = 2,\n",
    "        or_s = 3,\n",
    "        nudge = -0.4,\n",
    "        OR = T,\n",
    "        col = 'cohort',\n",
    "        dodge_width = 0.7\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    ylab('OR') +\n",
    "    xlab('') +\n",
    "    scale_color_discrete(name = 'Cohort')\n",
    "\n",
    "do_plot(p_forest, 'MS_R3_forest_pvalue.pdf', w = 6, h = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DOP_kelly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### MS edit\n",
    "\n",
    "options(repr.plot.width = 4, repr.plot.height = 2.5, repr.plot.res = 500)\n",
    "\n",
    "options(warn=-1)\n",
    "\n",
    "\n",
    "summ_cnv_therapy = M_wide %>%\n",
    "    filter(therapy_detailed & smoke != 'missing') %>%\n",
    "    filter(DateofProcedure == DOP_kelly) %>%\n",
    "    group_by(tumor_type) %>% mutate(n_tumor_type = n()) %>% ungroup() %>%\n",
    "    mutate(tumor_type = ifelse(n_tumor_type < 100, 'Other', tumor_type)) %>%\n",
    "    mutate(\n",
    "#         eqd_3_t = as.character(eqd_3_t),\n",
    "#         pct_cytotoxic_therapy = as.character(pct_cytotoxic_therapy),\n",
    "        smoke = as.integer(smoke)\n",
    "    ) %>%\n",
    "    glm(\n",
    "        ch_nonsilent ~ Gender + race_b + age10 + smoke + XRT + ind_cytotoxic_therapy,\n",
    "        data = .,\n",
    "        family = 'binomial'\n",
    "    ) %>% \n",
    "    sjPlot::get_model_data(type = 'est')\n",
    "\n",
    "options(warn=0)\n",
    "\n",
    "summ_cnv_therapy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_genes = c('DNMT3A', 'TET2', 'ASXL1', 'EZH2', 'TP53', 'ATM', 'CHEK2', 'PPM1D',\n",
    "        'SRSF2', 'SF3B1', 'U2AF1', 'MPL', 'JAK2', 'TERT')\n",
    "\n",
    "global_effects = data.frame()\n",
    "\n",
    "for (gene in top_genes) {\n",
    "        \n",
    "    D = M_wide %>% \n",
    "        filter(!is.na(get(gene))) %>%\n",
    "        mutate(multi_cnv_chrom = as.integer(n_cnv_chrom > 1)) %>%\n",
    "        filter(ch_cnv_auto == 1)\n",
    "    \n",
    "    n_co = sum(D[[paste0(gene, '')]] == 1 & D[['multi_cnv_chrom']] == 1)\n",
    "    \n",
    "    if (n_co >= 2) {\n",
    "        \n",
    "        pval = fisher.test(table(D[[paste0(gene, '')]], D[['multi_cnv_chrom']]))$p.value\n",
    "        \n",
    "        global_effects = rbind(\n",
    "            global_effects,\n",
    "            data.frame(\n",
    "                'Gene' = gene,\n",
    "                'intersect_count' = n_co,\n",
    "                'pval' = pval\n",
    "            )\n",
    "        )\n",
    "    }\n",
    "}\n",
    "\n",
    "global_effects = global_effects %>% mutate(q = p.adjust(pval, method = 'BH', n = length(top_genes)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "global_effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "source('/work/isabl/home/gaot/facets-ch/R/facets-ch-utils.R')\n",
    "\n",
    "plots = list()\n",
    "complex_samples = M_wide %>% filter(n_cnv_chrom >= 2 & TP53) %>% arrange(VAF_TP53) %>% pull(dmp_sample_id)\n",
    "\n",
    "for (sid in (complex_samples)) {\n",
    "    outdir = '/work/isabl/home/gaot/ch_cnv/results'\n",
    "\n",
    "    seg = fread(glue('{outdir}/{sid}_seg.tsv'))\n",
    "    sig = fread(glue('{outdir}/{sid}_sig.tsv'))\n",
    "    \n",
    "    plots[[sid]] = facets_plot(sig, seg, title = sid, snp_distance = F)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "panel = wrap_plots(plots, ncol = 1)\n",
    "options(warn = -1)\n",
    "do_plot(panel, 'complex.pdf', w = 10, h = 8)\n",
    "options(warn = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc-hr-collapsed": false
   },
   "source": [
    "## Cis effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_genes = c('DNMT3A', 'TET2', 'ASXL1', 'EZH2', 'TP53', 'ATM', 'CHEK2', 'PPM1D',\n",
    "        'SRSF2', 'SF3B1', 'U2AF1', 'MPL', 'JAK2', 'TERT')\n",
    "\n",
    "cis_gene_bed = gene_bed %>% filter(Gene %in% cis_genes)\n",
    "\n",
    "overlap_gene = function(seg_chrom, seg_start, seg_end, window = 5e6) {\n",
    "    \n",
    "    res = cis_gene_bed %>% \n",
    "        mutate(overlap = chrom == seg_chrom & start > seg_start - window & end < seg_end + window) %>%\n",
    "        pull(overlap) %>% as.integer()\n",
    "    \n",
    "    return(res)\n",
    "}\n",
    "\n",
    "cnv_overlaps = mapply(\n",
    "        overlap_gene,\n",
    "        segs_filtered$chrom, segs_filtered$start, segs_filtered$end\n",
    "    ) %>%\n",
    "    t %>%\n",
    "    as.data.frame() %>%\n",
    "    setNames(paste0(cis_gene_bed$Gene, '_cnv'))\n",
    "\n",
    "display('overlapped gene with CNV')\n",
    "\n",
    "segs_filtered = cbind(\n",
    "    segs_filtered %>% \n",
    "        select_if((!colnames(.) %in% colnames(cnv_overlaps)) | colnames(.) == 'dmp_patient_id'), \n",
    "    cnv_overlaps)\n",
    "\n",
    "segs_filtered = segs_filtered %>%\n",
    "    mutate_at(colnames(cnv_overlaps), list(amp = function(x){x * (.[['type']] == 'amp')})) %>%\n",
    "    mutate_at(colnames(cnv_overlaps), list(deloh = function(x){x * (.[['type']] %in% c('del', 'loh'))})) %>%\n",
    "    mutate_at(colnames(cnv_overlaps), list(del = function(x){x * (.[['type']] == 'del')})) %>%\n",
    "    mutate_at(colnames(cnv_overlaps), list(loh = function(x){x * (.[['type']] == 'loh')}))\n",
    "\n",
    "GENE_CNV = segs_filtered %>% \n",
    "    group_by(dmp_patient_id) %>%\n",
    "    summarise_at(\n",
    "        str_subset(colnames(.), '_cnv'),\n",
    "        max\n",
    "    )\n",
    "\n",
    "M_wide = M_wide %>%\n",
    "    select_if((!colnames(.) %in% colnames(GENE_CNV)) | colnames(.) == 'dmp_patient_id') %>%\n",
    "    left_join(\n",
    "        GENE_CNV,\n",
    "        by = \"dmp_patient_id\",\n",
    "        fill = 0\n",
    "    ) %>%\n",
    "    mutate_at(colnames(GENE_CNV), function(x){ifelse(is.na(x), 0, x)})\n",
    "\n",
    "display('merging with main dataframe')\n",
    "\n",
    "GENEALL = M_long %>%\n",
    "    reshape2::dcast(\n",
    "      formula = dmp_patient_id ~ Gene,\n",
    "      value.var = 'Gene',\n",
    "      fun.aggregate = function(x) {min(1, length(x))}\n",
    "    ) %>%\n",
    "    setNames(c('dmp_patient_id', paste0(colnames(.)[-1], '_all')))\n",
    "\n",
    "M_wide = M_wide %>%\n",
    "    select_if((!colnames(.) %in% colnames(GENEALL)) | colnames(.) == 'dmp_patient_id') %>%\n",
    "    left_join(GENEALL, by = \"dmp_patient_id\") %>%\n",
    "    mutate_at(colnames(GENEALL), function(x){ifelse(is.na(x), 0, x)})\n",
    "\n",
    "segs_filtered = segs_filtered %>%\n",
    "    select_if((!colnames(.) %in% colnames(GENEALL)) | colnames(.) == 'dmp_patient_id') %>%\n",
    "    left_join(GENEALL, by = \"dmp_patient_id\") %>%\n",
    "    mutate_at(colnames(GENEALL), function(x){ifelse(is.na(x), 0, x)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_effects = data.frame()\n",
    "\n",
    "for (gene in cis_genes) {\n",
    "        \n",
    "    D = M_wide %>% filter(!is.na(get(gene)))\n",
    "    \n",
    "    sm_count = sum(D[[paste0(gene, '')]])\n",
    "    cnv_count = sum(D[[paste0(gene, '_cnv')]])\n",
    "    del_count = sum(D[[paste0(gene, '_del')]])\n",
    "    loh_count = sum(D[[paste0(gene, '_loh')]])\n",
    "    intersect_cnv = sum(D[[paste0(gene, '')]] == 1 & D[[paste0(gene, '_cnv')]] == 1)\n",
    "    intersect_del = sum(D[[paste0(gene, '')]] == 1 & D[[paste0(gene, '_cnv_del')]] == 1)\n",
    "    intersect_loh = sum(D[[paste0(gene, '')]] == 1 & D[[paste0(gene, '_cnv_loh')]] == 1)\n",
    "    \n",
    "    #  pval_all = fisher.test(table(D[[paste0(gene, '')]], D[[paste0(gene, '_cnv')]]))$p.value\n",
    "\n",
    "    if (intersect_del >= 2) {\n",
    "        \n",
    "        pval = fisher.test(table(D[[paste0(gene, '')]], D[[paste0(gene, '_cnv_del')]]))$p.value\n",
    "        \n",
    "        cis_effects = rbind(\n",
    "            cis_effects,\n",
    "            data.frame(\n",
    "                'Gene' = gene,\n",
    "                'cnv_type' = 'del',\n",
    "                'sm_count' = sm_count,\n",
    "                'cnv_count' = del_count,\n",
    "                'intersect_count' = intersect_del,\n",
    "                'pval' = pval\n",
    "            )\n",
    "        )\n",
    "    }\n",
    "    \n",
    "    if (intersect_loh >= 2) {\n",
    "        \n",
    "        pval = fisher.test(table(D[[paste0(gene, '')]], D[[paste0(gene, '_cnv_loh')]]))$p.value\n",
    "        \n",
    "        cis_effects = rbind(\n",
    "            cis_effects,\n",
    "            data.frame(\n",
    "                'Gene' = gene,\n",
    "                'cnv_type' = 'loh',\n",
    "                'sm_count' = sm_count,\n",
    "                'cnv_count' = del_count,\n",
    "                'intersect_count' = intersect_loh,\n",
    "                'pval' = pval\n",
    "            )\n",
    "        )\n",
    "    }\n",
    "}\n",
    "\n",
    "cis_effects = cis_effects %>% mutate(q = p.adjust(pval, method = 'BH', n = length(cis_genes) * 2))\n",
    "\n",
    "cis_effects = cis_effects %>%\n",
    "    mutate(Gene = as.character(Gene)) %>%\n",
    "    left_join(gene_bed, by = 'Gene') %>%\n",
    "    arrange(chrom) %>% mutate(Gene = factor(Gene, unique(Gene)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_cnvs = function(gene, qval) {\n",
    "    \n",
    "    all_colors = c(\n",
    "        'amp' = 'darkblue', 'loh' = 'darkgreen', 'del' = 'darkred',\n",
    "        'Regulatory' = 'darkturquoise', 'Missense' = 'salmon', 'Truncating' = 'purple') \n",
    "\n",
    "    gene_chrom = gene_bed %>% filter(Gene == gene) %>% pull(chrom)\n",
    "    \n",
    "    gene_structure = gene_structures %>% filter(str_detect(description, gene))\n",
    "\n",
    "    gene_locs = gene_bed %>% \n",
    "        filter(Gene == gene) %>%\n",
    "        as.data.frame(stringsAsFactors = F) %>%\n",
    "        mutate(loc = as.integer((start + end) / 2), chrom = as.integer(chrom))\n",
    "\n",
    "    this_chrom_lens = chrom_lens %>% filter(chrom == gene_chrom)\n",
    "\n",
    "    acen_chrom = acen %>% filter(chrom == gene_chrom)\n",
    "\n",
    "    D = segs_filtered %>%\n",
    "        filter(chrom == gene_chrom) %>%\n",
    "        mutate(chrom = factor(chrom)) %>%\n",
    "        group_by(dmp_sample_id, chrom) %>%\n",
    "        mutate(total_length = sum(size)) %>%\n",
    "        mutate(type = factor(type, c('amp', 'loh', 'del'))) %>%\n",
    "        arrange(desc(type), round(start/1e7), -total_length) %>%\n",
    "        group_by(chrom) %>%\n",
    "        mutate(index = as.integer(factor(dmp_sample_id, unique(dmp_sample_id)))) %>%\n",
    "        ungroup()\n",
    "\n",
    "    M_overlap = M_long %>%\n",
    "        filter(Gene == gene) %>%\n",
    "        filter(dmp_patient_id %in% D$dmp_patient_id) %>%\n",
    "        rowwise() %>%\n",
    "        filter(\n",
    "            any(chrom == D$chrom & dmp_patient_id == D$dmp_patient_id)\n",
    "#                 start >= D$start - window & start <= D$end + window)\n",
    "        ) %>%\n",
    "        left_join(\n",
    "            D %>% select(dmp_sample_id, index),\n",
    "            by = 'dmp_sample_id'\n",
    "        ) %>%\n",
    "        ungroup() %>%\n",
    "        mutate(VariantClass = case_when(\n",
    "            VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "            T ~ 'coding'\n",
    "        ))\n",
    "\n",
    "    p_segs = ggplot(\n",
    "            D,\n",
    "            aes(x = start, xend = end, y = index, yend = index, color = type)\n",
    "        ) +\n",
    "        geom_rect(\n",
    "            inherit.aes = F,\n",
    "            data = acen_chrom,\n",
    "            aes(xmin = cen_start, xmax = cen_end, ymin = -Inf, ymax = Inf),\n",
    "            fill = 'gray',\n",
    "            size = 0,\n",
    "            alpha = 0.3\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            size = 1,\n",
    "            alpha = 0.8,\n",
    "            lineend = 'round'\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = this_chrom_lens,\n",
    "            aes(x = 0, xend = length, y = 0, yend = 0),\n",
    "            color = 'black',\n",
    "            size = 0\n",
    "        ) +\n",
    "        # label the mutations\n",
    "#         geom_point(\n",
    "#             inherit.aes = F,\n",
    "#             mapping = aes(x = start, y = index, color = VariantClass2),\n",
    "#             data = M_overlap,\n",
    "#             size = 6.5,\n",
    "#             pch = '*',\n",
    "#             color = 'white'\n",
    "#         ) +\n",
    "        geom_point(\n",
    "            inherit.aes = F,\n",
    "            mapping = aes(x = start, y = index, fill = VariantClass2),\n",
    "            data = M_overlap,\n",
    "            size = 1.5,\n",
    "            color = 'black',\n",
    "            pch = 21\n",
    "        ) +\n",
    "#         geom_text_repel(\n",
    "#             inherit.aes = F,\n",
    "#             mapping = aes(x = start, y = index, color = VariantClass2, label = AAchange),\n",
    "#             data = M_overlap %>% mutate(zoom = TRUE),\n",
    "#             size = 1.5,\n",
    "#             segment.size = 0.2,\n",
    "#             color = 'black',\n",
    "#             ylim = c(0,NA)\n",
    "#         ) +\n",
    "        scale_alpha(\n",
    "            range = c(0, 1)\n",
    "        ) +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            axis.text = element_blank(),\n",
    "            axis.ticks = element_blank(),\n",
    "            axis.title = element_blank(),\n",
    "            legend.position = 'none',\n",
    "            panel.spacing.y = unit(1, \"mm\"),\n",
    "            plot.margin = unit(c(0.5,0.2,0.2,0.2), \"lines\"),\n",
    "            strip.text.y = element_blank(),\n",
    "            panel.border = element_rect(size = 0.2, fill = NA, color = 'gray'),\n",
    "            strip.background = element_rect(size = 0.5, fill = 'seashell3', color = 'seashell3'),\n",
    "            plot.title = element_text(size = 7)\n",
    "        ) +\n",
    "        scale_y_discrete(expand = expansion(add = 1)) +\n",
    "        xlab('') +\n",
    "        ylab('') +\n",
    "        scale_color_manual(values = all_colors) +\n",
    "        scale_fill_manual(values = all_colors) +\n",
    "        ggtitle(paste0(gene, ', chr', gene_chrom)) +\n",
    "        # exon structures\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = gene_structure %>% mutate(zoom = TRUE),\n",
    "            aes(x = min(start), xend = max(end), y = -1, yend = -1),\n",
    "            size = 0.1,\n",
    "            color = 'black'\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = gene_structure %>% mutate(zoom = TRUE),\n",
    "            aes(x = start, xend = end, y = -1, yend = -1),\n",
    "            size = 2,\n",
    "            color = 'black'\n",
    "        ) +\n",
    "        facet_zoom(\n",
    "            zoom.size = 1,\n",
    "            zoom.data = zoom,\n",
    "            xlim = c(gene_locs$start, gene_locs$end)\n",
    "        )\n",
    "\n",
    "    p_segs\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plots = list()\n",
    "\n",
    "for (i in 1:nrow(cis_effects)) {\n",
    "    gene = as.character(cis_effects[i,'Gene'])\n",
    "    qval = cis_effects[i,'q']\n",
    "    pval = cis_effects[i,'pval']\n",
    "    plots[[gene]] = gene_cnvs(gene, pval)\n",
    "}\n",
    "\n",
    "p_cis = plot_grid(plotlist = plots, nrow = 2)\n",
    "\n",
    "do_plot(p_cis, 'cis_loci.pdf', w = 6, h = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  add text labels \n",
    "gene_cnvs_MS = function(gene, qval) {\n",
    "    \n",
    "    all_colors = c(\n",
    "        'amp' = 'darkblue', 'loh' = 'darkgreen', 'del' = 'darkred',\n",
    "        'Regulatory' = 'darkturquoise', 'Missense' = 'salmon', 'Truncating' = 'purple') \n",
    "\n",
    "    gene_chrom = gene_bed %>% filter(Gene == gene) %>% pull(chrom)\n",
    "    \n",
    "    gene_structure = gene_structures %>% filter(str_detect(description, gene))\n",
    "\n",
    "    gene_locs = gene_bed %>% \n",
    "        filter(Gene == gene) %>%\n",
    "        as.data.frame(stringsAsFactors = F) %>%\n",
    "        mutate(loc = as.integer((start + end) / 2), chrom = as.integer(chrom))\n",
    "\n",
    "    this_chrom_lens = chrom_lens %>% filter(chrom == gene_chrom)\n",
    "\n",
    "    acen_chrom = acen %>% filter(chrom == gene_chrom)\n",
    "\n",
    "    D = segs_filtered %>%\n",
    "        filter(chrom == gene_chrom) %>%\n",
    "        mutate(chrom = factor(chrom)) %>%\n",
    "        group_by(dmp_sample_id, chrom) %>%\n",
    "        mutate(total_length = sum(size)) %>%\n",
    "        mutate(type = factor(type, c('amp', 'loh', 'del'))) %>%\n",
    "        arrange(desc(type), round(start/1e7), -total_length) %>%\n",
    "        group_by(chrom) %>%\n",
    "        mutate(index = as.integer(factor(dmp_sample_id, unique(dmp_sample_id)))) %>%\n",
    "        ungroup()\n",
    "\n",
    "    M_overlap = M_long %>%\n",
    "        filter(Gene == gene) %>%\n",
    "        filter(dmp_patient_id %in% D$dmp_patient_id) %>%\n",
    "        rowwise() %>%\n",
    "        filter(\n",
    "            any(chrom == D$chrom & dmp_patient_id == D$dmp_patient_id)\n",
    "#                 start >= D$start - window & start <= D$end + window)\n",
    "        ) %>%\n",
    "        left_join(\n",
    "            D %>% select(dmp_sample_id, index),\n",
    "            by = 'dmp_sample_id'\n",
    "        ) %>%\n",
    "        ungroup() %>%\n",
    "        mutate(VariantClass = case_when(\n",
    "            VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "            T ~ 'coding'\n",
    "        ))\n",
    "    \n",
    "    M_overlap_MS = M_overlap #%>% distinct(dmp_patient_id, start, .keep_all = TRUE)\n",
    "    \n",
    "    p_segs = ggplot(\n",
    "            D,\n",
    "            aes(x = start, xend = end, y = index, yend = index, color = type)\n",
    "        ) +\n",
    "        geom_rect(\n",
    "            inherit.aes = F,\n",
    "            data = acen_chrom,\n",
    "            aes(xmin = cen_start, xmax = cen_end, ymin = -Inf, ymax = Inf),\n",
    "            fill = 'gray',\n",
    "            size = 0,\n",
    "            alpha = 0.3\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            size = 1,\n",
    "            alpha = 0.8,\n",
    "            lineend = 'round'\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = this_chrom_lens,\n",
    "            aes(x = 0, xend = length, y = 0, yend = 0),\n",
    "            color = 'black',\n",
    "            size = 0\n",
    "        ) +\n",
    "        # label the mutations\n",
    "#         geom_point(\n",
    "#             inherit.aes = F,\n",
    "#             mapping = aes(x = start, y = index, color = VariantClass2),\n",
    "#             data = M_overlap,\n",
    "#             size = 6.5,\n",
    "#             pch = '*',\n",
    "#             color = 'white'\n",
    "#         ) +\n",
    "        geom_point(\n",
    "            inherit.aes = F,\n",
    "            mapping = aes(x = start, y = index, fill = VariantClass2),\n",
    "            data = M_overlap_MS,\n",
    "            size = 1.5,\n",
    "            color = 'black',\n",
    "            pch = 21\n",
    "        ) +\n",
    "        geom_text_repel(\n",
    "            inherit.aes = F,\n",
    "            mapping = aes(x = start, y = index, color = VariantClass2, label = paste(AAchange, start, variant_id)),\n",
    "            data = M_overlap_MS %>% mutate(zoom = TRUE),\n",
    "            size = 2,\n",
    "            force = 5,\n",
    "            segment.size = 0.2,\n",
    "            color = 'black',\n",
    "            ylim = c(0,NA)\n",
    "        ) +\n",
    "        scale_alpha(\n",
    "            range = c(0, 1)\n",
    "        ) +\n",
    "        theme_classic() +\n",
    "        theme(\n",
    "            axis.text = element_blank(),\n",
    "            axis.ticks = element_blank(),\n",
    "            axis.title = element_blank(),\n",
    "            legend.position = 'none',\n",
    "            panel.spacing.y = unit(1, \"mm\"),\n",
    "            plot.margin = unit(c(0.5,0.2,0.2,0.2), \"lines\"),\n",
    "            strip.text.y = element_blank(),\n",
    "            panel.border = element_rect(size = 0.2, fill = NA, color = 'gray'),\n",
    "            strip.background = element_rect(size = 0.5, fill = 'seashell3', color = 'seashell3'),\n",
    "            plot.title = element_text(size = 7)\n",
    "        ) +\n",
    "        scale_y_discrete(expand = expansion(add = 1)) +\n",
    "        xlab('') +\n",
    "        ylab('') +\n",
    "        scale_color_manual(values = all_colors) +\n",
    "        scale_fill_manual(values = all_colors) +\n",
    "        ggtitle(paste0(gene, ', chr', gene_chrom)) +\n",
    "        # exon structures\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = gene_structure %>% mutate(zoom = TRUE),\n",
    "            aes(x = min(start), xend = max(end), y = -1, yend = -1),\n",
    "            size = 0.1,\n",
    "            color = 'black'\n",
    "        ) +\n",
    "        geom_segment(\n",
    "            inherit.aes = F,\n",
    "            data = gene_structure %>% mutate(zoom = TRUE),\n",
    "            aes(x = start, xend = end, y = -1, yend = -1),\n",
    "            size = 2,\n",
    "            color = 'black'\n",
    "        ) +\n",
    "        facet_zoom(\n",
    "            zoom.size = 1,\n",
    "            zoom.data = zoom,\n",
    "            xlim = c(gene_locs$start, gene_locs$end)\n",
    "        )\n",
    "\n",
    "    p_segs\n",
    "#     M_overlap_MS\n",
    "#     M_overlap %>% #distinct(dmp_patient_id, start,.keep_all = TRUE) %>% \n",
    "# filter(Gene %in% c(\"EZH2\", \"TP53\")) %>% select(c(dmp_patient_id, dmp_sample_id, Gene, variant_id, start, AAchange,cDNAchange, VariantClass)) %>% arrange(dmp_patient_id)\n",
    "\n",
    "# M_overlap_MS %>% #distinct(dmp_patient_id, start,.keep_all = TRUE) %>% \n",
    "# filter(Gene %in% c(\"EZH2\", \"TP53\")) %>% select(c(dmp_patient_id, dmp_sample_id, Gene, variant_id, start, AAchange,cDNAchange, VariantClass)) %>% arrange(dmp_patient_id)\n",
    "}\n",
    "plots = list()\n",
    "\n",
    "for (i in 1:nrow(cis_effects)) {\n",
    "    gene = as.character(cis_effects[i,'Gene'])\n",
    "    qval = cis_effects[i,'q']\n",
    "    pval = cis_effects[i,'pval']\n",
    "    plots[[gene]] = gene_cnvs_MS(gene, pval)\n",
    "}\n",
    "\n",
    "p_cis = plot_grid(plotlist = plots, nrow = 2)\n",
    "\n",
    "do_plot(p_cis, 'MS_edit_labeled_cis_loci.pdf', w = 15, h = 12)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cis_total = sum(cis_effects$intersect_count)\n",
    "cis_total %>% paste0(' total cis events')\n",
    "(cis_total/nrow(segs_filtered_auto)) %>% signif(3) %>% {.*100} %>% paste0('% of total events')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 5e6\n",
    "\n",
    "segs_filtered_auto = segs_filtered %>% filter(chrom != 23)\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered_auto$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    mutate(\n",
    "        locality = ifelse(\n",
    "            any(chrom == segs_filtered_auto$chrom & dmp_patient_id == segs_filtered_auto$dmp_patient_id &\n",
    "                start >= segs_filtered_auto$start - window & start <= segs_filtered_auto$end + window\n",
    "            ),\n",
    "            'cis',\n",
    "            'trans'\n",
    "        )\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(variant_type = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        T ~ 'coding'\n",
    "    ))\n",
    "\n",
    "\n",
    "M_overlap_nonsilent = M_overlap %>% filter(variant_type == 'coding')\n",
    "\n",
    "segs_filtered_auto %>%\n",
    "    rowwise() %>%\n",
    "    mutate(\n",
    "        cis_coding = as.integer(\n",
    "            any(chrom == M_overlap_nonsilent$chrom & dmp_patient_id == M_overlap_nonsilent$dmp_patient_id &\n",
    "                M_overlap_nonsilent$start >= start - window & M_overlap_nonsilent$start <= end + window\n",
    "            ))\n",
    "    ) %>%\n",
    "    group_by(type) %>%\n",
    "    summarise(\n",
    "        n = n(),\n",
    "        prop_cis = sum(cis_coding)/n\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GM CNV VAF correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "window = 1e7\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "           start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        str_detect(VariantClass, 'Del|Ins') ~ 'indel',\n",
    "        T ~ 'snv'\n",
    "    )) %>%\n",
    "    filter(VariantClass != 'noncoding') %>%\n",
    "    filter(Gene %in% c(\"EZH2\", \"ATM\", \"TET2\", \"JAK2\", \"MPL\", 'DNMT3A', 'TP53')) %>%\n",
    "    arrange(desc(VAF_N)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T)\n",
    "    \n",
    "\n",
    "p_vaf = segs_filtered %>%\n",
    "    left_join(\n",
    "        M_overlap %>% select(dmp_patient_id, chrom, sm_pos = start, Gene, VariantClass, VAF_N),\n",
    "        by = c('chrom', 'dmp_patient_id')\n",
    "    ) %>%\n",
    "    filter(type != 'amp') %>%\n",
    "    filter(VAF_N > 0) %>%\n",
    "    mutate(\n",
    "        phi_sm = case_when(\n",
    "            type == 'del' ~ 2 * VAF_N/(1 + VAF_N),\n",
    "            type == 'loh' ~ VAF_N\n",
    "        )\n",
    "    ) %>% arrange(phi, ascending = FALSE) %>% distinct(dmp_patient_id, sm_pos, .keep_all = TRUE) %>% \n",
    "    mutate(pigeonhole = phi_sm + phi > 1) %>%\n",
    "    ggplot(\n",
    "        aes(x = VAF_N, y = phi, color = type2, label = Gene)\n",
    "    ) +\n",
    "    # stat_function(fun = function(v){2*v}, linetype = 'dashed', color = cnv_pal['err']) +\n",
    "    # stat_function(fun = function(v){2 * v/(1+v)}, linetype = 'dashed', color = cnv_pal['del'], alpha = 0.5) +\n",
    "    # geom_abline(linetype = 'dashed', color = cnv_pal['loh'], alpha = 0.5) +\n",
    "    stat_function(fun = function(v){2 * v}, linetype = 'dashed', color = cnv_pal['err']) +\n",
    "    stat_function(fun = function(v){2 * v/(1+v)}, linetype = 'dashed', color = cnv_pal['del'], alpha = 0.5) +\n",
    "    geom_abline(linetype = 'dashed', color = cnv_pal['loh'], alpha = 0.5) +\n",
    "#     annotate('text', x = 0.35, y = 1, label = 'trans', color = cnv_pal['err'], size = 3) +\n",
    "#     annotate('text', x = 0.75, y = 1, label = 'DEL', color = cnv_pal['del'], size = 3, alpha = 0.5) +\n",
    "#     annotate('text', x = 0.9, y = 0.7, label = 'CNLOH', color = cnv_pal['loh'], size = 3, alpha = 0.5) +\n",
    "    theme_classic() +\n",
    "    # stat_cor(label.y = 0.95, size = 3) +\n",
    "    scale_color_manual(values = cnv_pal2) +\n",
    "    geom_point(\n",
    "        aes(pch = pigeonhole),\n",
    "#         pch = 21\n",
    "    ) +\n",
    "    scale_shape_manual(values = c('TRUE' = 19, 'FALSE' = 21)) +\n",
    "    xlim(0,1) +\n",
    "    ylim(0,1) +\n",
    "    geom_text(size = 1, vjust = -1, hjust = -0.2) +\n",
    "    facet_wrap(~type2, nrow = 1, scale = 'free') +\n",
    "    theme(\n",
    "        legend.position = 'none',\n",
    "        strip.background = element_blank()\n",
    "    ) +\n",
    "    xlab('Gene mutation VAF') +\n",
    "    ylab('mCA cell fraction')\n",
    "\n",
    "do_plot(p_vaf, 'cis_vaf.pdf', w = 4.6, h = 2.5)\n",
    "# do_plot(p_vaf, 'cis_vaf.pdf', w = 8, h = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segs_filtered %>%\n",
    "    left_join(\n",
    "        M_overlap %>% select(dmp_patient_id, chrom, sm_pos = start, Gene, VariantClass, VAF_N),\n",
    "        by = c('chrom', 'dmp_patient_id')\n",
    "    ) %>%\n",
    "    filter(type != 'amp') %>%\n",
    "    filter(VAF_N > 0) %>%\n",
    "    mutate(\n",
    "        phi_sm = case_when(\n",
    "            type == 'del' ~ 2 * VAF_N/(1 + VAF_N),\n",
    "            type == 'loh' ~ VAF_N\n",
    "        )\n",
    "    ) %>% arrange(phi, ascending = FALSE) %>% distinct(dmp_patient_id, sm_pos, .keep_all = TRUE) %>% \n",
    "    mutate(pigeonhole = phi_sm + phi > 1) %>% count(type,pigeonhole)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "           start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        str_detect(VariantClass, 'Del|Ins') ~ 'indel',\n",
    "        T ~ 'snv'\n",
    "    )) %>%\n",
    "    filter(VariantClass != 'noncoding') %>%\n",
    "    filter(Gene %in% c(\"EZH2\", \"ATM\", \"TET2\", \"JAK2\", \"MPL\", 'DNMT3A', 'TP53')) %>%\n",
    "    arrange(desc(VAF_N)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T)\n",
    "M_overlap %>% count(is.na(N_TotalDepth))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### MS edited \n",
    "\n",
    "window = 1e7\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "           start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        str_detect(VariantClass, 'Del|Ins') ~ 'indel',\n",
    "        T ~ 'snv'\n",
    "    )) %>%\n",
    "    filter(VariantClass != 'noncoding') %>%\n",
    "    filter(Gene %in% c(\"EZH2\", \"ATM\", \"TET2\", \"JAK2\", \"MPL\", 'DNMT3A', 'TP53')) %>%\n",
    "    arrange(desc(VAF_N)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T)\n",
    "    \n",
    "\n",
    "M_overlap = segs_filtered %>%\n",
    "    left_join(\n",
    "        M_overlap %>% select(dmp_patient_id, chrom, sm_pos = start, Gene, VariantClass, VAF_N, N_RefCount, N_AltCount),\n",
    "        by = c('chrom', 'dmp_patient_id')\n",
    "    ) %>%\n",
    "    filter(type != 'amp') %>%\n",
    "    filter(VAF_N > 0) %>%\n",
    "    mutate(\n",
    "        phi_sm = case_when(\n",
    "            type == 'del' ~ 2 * VAF_N/(1 + VAF_N),\n",
    "            type == 'loh' ~ VAF_N\n",
    "        )\n",
    "    ) %>%\n",
    "    mutate(pigeonhole = phi_sm + phi > 1) #%>%\n",
    "#     ggplot(\n",
    "#         aes(x = VAF_N, y = phi, color = type2, label = Gene)\n",
    "#     ) +\n",
    "#     # stat_function(fun = function(v){2*v}, linetype = 'dashed', color = cnv_pal['err']) +\n",
    "#     # stat_function(fun = function(v){2 * v/(1+v)}, linetype = 'dashed', color = cnv_pal['del'], alpha = 0.5) +\n",
    "#     # geom_abline(linetype = 'dashed', color = cnv_pal['loh'], alpha = 0.5) +\n",
    "#     stat_function(fun = function(v){2 * v}, linetype = 'dashed', color = cnv_pal['err']) +\n",
    "#     stat_function(fun = function(v){2 * v/(1+v)}, linetype = 'dashed', color = cnv_pal['del'], alpha = 0.5) +\n",
    "#     geom_abline(linetype = 'dashed', color = cnv_pal['loh'], alpha = 0.5) +\n",
    "# #     annotate('text', x = 0.35, y = 1, label = 'trans', color = cnv_pal['err'], size = 3) +\n",
    "# #     annotate('text', x = 0.75, y = 1, label = 'DEL', color = cnv_pal['del'], size = 3, alpha = 0.5) +\n",
    "# #     annotate('text', x = 0.9, y = 0.7, label = 'CNLOH', color = cnv_pal['loh'], size = 3, alpha = 0.5) +\n",
    "#     theme_classic() +\n",
    "#     # stat_cor(label.y = 0.95, size = 3) +\n",
    "\n",
    "#     scale_color_manual(values = cnv_pal2) +\n",
    "#     geom_point(\n",
    "#         aes(pch = pigeonhole),\n",
    "# #         pch = 21\n",
    "#     ) +\n",
    "#     scale_shape_manual(values = c('TRUE' = 19, 'FALSE' = 21)) +\n",
    "#     xlim(0,1) +\n",
    "#     ylim(0,1) +\n",
    "#     geom_text(size = 1, vjust = -1, hjust = -0.2) +\n",
    "#     facet_wrap(~type2, nrow = 1, scale = 'free') +\n",
    "#     theme(\n",
    "#         legend.position = 'none',\n",
    "#         strip.background = element_blank()\n",
    "#     ) +\n",
    "#     xlab('Gene mutation VAF') +\n",
    "#     ylab('mCA cell fraction')\n",
    "\n",
    "# do_plot(p_vaf, 'cis_vaf.pdf', w = 4.6, h = 2.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cn_loh_df = M_overlap %>% filter(type == \"loh\") %>% distinct(dmp_patient_id, sm_pos, .keep_all = TRUE) %>% filter(!is.na(N_AltCount)) %>% select ( dmp_patient_id, start, end, Gene, VariantClass, VAF_N, phi, phi_sm, pigeonhole, N_AltCount, N_RefCount) \n",
    "cn_loh_df %>% rowwise() %>% mutate(binom_res = binom.test(N_AltCount, (N_AltCount+N_RefCount), 0.5)$p.value) \n",
    "\n",
    "                           "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## microarray validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "double_hits_cyto = M_wide %>% filter(\n",
    "    (MPL & chr1 & VAF_MPL < 0.35) |\n",
    "    (DNMT3A & chr2 & VAF_DNMT3A < 0.35) |\n",
    "    (TET2 & chr4 & VAF_TET2 < 0.35) |\n",
    "    (EZH2 & chr7 & VAF_EZH2 < 0.35) |\n",
    "    (JAK2 & chr9 & VAF_JAK2 < 0.35) |\n",
    "    (ATM & chr11 & VAF_ATM < 0.35) |\n",
    "    (TP53 & chr17 & VAF_TP53 < 0.35)\n",
    ") %>% \n",
    "mutate(hits = case_when(\n",
    "    MPL & chr1 ~ 'MPL',\n",
    "    DNMT3A & chr2 ~ 'DNMT3A',\n",
    "    TET2 & chr4 ~ 'TET2',\n",
    "    EZH2 & chr7 ~ 'EZH2',\n",
    "    JAK2 & chr9 ~ 'JAK2',\n",
    "    ATM & chr11 ~ 'ATM',\n",
    "    TP53 & chr17 ~ 'TP53'\n",
    ")) %>%\n",
    "select(dmp_patient_id, dmp_sample_id, hits) %>%\n",
    "arrange(hits)\n",
    "\n",
    "# double_hits_cyto %>% fwrite('./data/double_hits_cyto.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 2.3, repr.plot.res = 300)\n",
    "\n",
    "window = 1e7\n",
    "\n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        !any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "           start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        str_detect(VariantClass, 'Del|Ins') ~ 'indel',\n",
    "        T ~ 'snv'\n",
    "    )) %>%\n",
    "    filter(VariantClass != 'noncoding') %>%\n",
    "    arrange(desc(VAF_N)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T)\n",
    "\n",
    "p_vaf = segs_filtered %>%\n",
    "    arrange(desc(phi)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T) %>%\n",
    "    left_join(\n",
    "        M_overlap %>% select(dmp_patient_id, chrom, sm_pos = start, Gene, VariantClass, VAF_N),\n",
    "        by = c('dmp_patient_id')\n",
    "    ) %>%\n",
    "    mutate(pigeonhole = 2 * VAF_N + phi > 1) %>%\n",
    "    ggplot(\n",
    "        aes(x = VAF_N, y = phi, color = type2, label = Gene)\n",
    "    ) +\n",
    "    stat_function(fun = function(v){2*v}, linetype = 'dashed', color = cnv_pal['err']) +\n",
    "    stat_function(fun = function(v){2 * v/(1+v)}, linetype = 'dashed', color = cnv_pal['del'], alpha = 0.5) +\n",
    "    geom_abline(linetype = 'dashed', color = cnv_pal['loh'], alpha = 0.5) +\n",
    "#     annotate('text', x = 0.35, y = 1, label = 'trans', color = cnv_pal['err'], size = 3) +\n",
    "#     annotate('text', x = 0.75, y = 1, label = 'cis-del', color = cnv_pal['del'], size = 3, alpha = 0.5) +\n",
    "#     annotate('text', x = 0.9, y = 0.7, label = 'cis-loh', color = cnv_pal['loh'], size = 3, alpha = 0.5) +\n",
    "    theme_classic() +\n",
    "    # geom_abline(linetype = 'dashed') +\n",
    "    scale_shape_manual(values = c('TRUE' = 19, 'FALSE' = 21)) +\n",
    "    # stat_cor(label.y = 0.98, size = 3) +\n",
    "    scale_color_manual(values = cnv_pal2) +\n",
    "    geom_point(\n",
    "    #     pch = 21,\n",
    "        aes(pch = pigeonhole)\n",
    "    ) +\n",
    "    xlim(0,1) +\n",
    "    ylim(0,1) +\n",
    "    # geom_text(size = 1, vjust = -1, hjust = -0.2) +\n",
    "    # scale_shape_manual(values = c(1,5)) +\n",
    "    facet_wrap(~type2, nrow = 1, scale = 'free') +\n",
    "    theme(\n",
    "        legend.position = 'none',\n",
    "        strip.background = element_blank()\n",
    "    ) +\n",
    "    ylab('mCA cell fraction') +\n",
    "    xlab('Gene mutation VAF')\n",
    "\n",
    "do_plot(p_vaf, 'MS_edit_trans_vaf.pdf', w = 6.8, h = 2.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_trans_tocast = segs_filtered %>%\n",
    "    arrange(desc(phi)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T) %>%\n",
    "    left_join(\n",
    "        M_overlap %>% select(dmp_patient_id, chrom, sm_pos = start, Gene, VariantClass, VAF_N),\n",
    "        by = c('dmp_patient_id')\n",
    "    ) %>% filter(!is.na(VAF_N)) %>% \n",
    "    mutate(pigeonhole = 2 * VAF_N + phi > 1) %>% count(type2, pigeonhole) \n",
    "\n",
    "dcast(M_wide_trans_tocast, ( type2 ~ pigeonhole), value.var=\"n\", fill = 0) %>% mutate(total = rowSums(.[2:3]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segs_filtered %>%\n",
    "    arrange(desc(phi)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T) %>%\n",
    "    left_join(\n",
    "        M_overlap %>% select(dmp_patient_id, chrom, sm_pos = start, Gene, VariantClass, VAF_N),\n",
    "        by = c('dmp_patient_id')\n",
    "    ) %>%\n",
    "    mutate(pigeonhole = 2 * VAF_N + phi > 1) %>% count(is.na(VAF_N), is.na(phi))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### MS edited: EZH2 cis CHr 7 \n",
    "\n",
    "EZH2_mut_pts = M_long %>% filter (Gene == 'EZH2') %>% filter(chr7 | chr7_del | chr7_loh ) %>% select (dmp_patient_id, Gene, VAF_N, start,  AAchange, chr7 , chr7_del, chr7_loh)\n",
    "\n",
    "EZH2_mut_pts\n",
    "segs_chr7 = segs_filtered %>% filter(chrom == 7) %>% select(dmp_patient_id, chrom, start, end, phi, cnlr, cnlr_adj, valor)\n",
    "\n",
    "EZH2_chr7 = merge(EZH2_mut_pts, segs_chr7, by = \"dmp_patient_id\", suffixes = c(\"_GM\", \"_mCA\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EZH2_chr7 %>% mutate (overlapped = ifelse((EZH2_chr7$start_GM >= EZH2_chr7$start_mCA & EZH2_chr7$start_GM <EZH2_chr7$end), 1,0)) %>% filter(as.logical(overlapped ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"EZH2\" \n",
    "M_overlap = M_long %>%\n",
    "    filter(dmp_patient_id %in% segs_filtered$dmp_patient_id) %>%\n",
    "    rowwise() %>%\n",
    "    filter(\n",
    "        !any(chrom == segs_filtered$chrom & dmp_patient_id == segs_filtered$dmp_patient_id &\n",
    "           start >= segs_filtered$start - window & start <= segs_filtered$end + window)\n",
    "    ) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(VariantClass = case_when(\n",
    "        VariantClass %in% c('Silent', 'Intron') ~ 'noncoding',\n",
    "        str_detect(VariantClass, 'Del|Ins') ~ 'indel',\n",
    "        T ~ 'snv'\n",
    "    )) %>%\n",
    "    filter(VariantClass != 'noncoding') %>%\n",
    "    arrange(desc(VAF_N)) %>%\n",
    "    distinct(dmp_patient_id, `.keep_all` = T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_long %>% filter(Gene == \"EZH2\") %>% count(VariantClass)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Blood count phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "give.median <- function(x){\n",
    "  return(c(x = median(x), label = median(x))) \n",
    "}\n",
    "\n",
    "D = M_wide %>% \n",
    "    filter(anc < 30 & alc < 8 & amc < 3 & plt < 1000 | is.na(anc) | is.na(alc) | is.na(amc) | is.na(plt)) %>%\n",
    "    mutate(group = ifelse(ch_cnv == 1, 'mCA+', 'mCA-')) %>%\n",
    "    mutate(group = factor(group, c('mCA+', 'mCA-'))) %>%\n",
    "    group_by(group) %>%\n",
    "    mutate(n = n()) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(group_label = paste0(group, '\\n(n=', n, ')')) %>%\n",
    "    mutate(group_label = factor(group_label, unique(group_label))) %>%\n",
    "    reshape2::melt(measure.vars = cbcs[cbcs != 'rbc']) %>%\n",
    "    mutate(variable = str_wrap(cbc_labels[variable], width = 3))\n",
    "\n",
    "p_cbc = ggplot(\n",
    "    D,\n",
    "    aes(x = value, fill = group_label)\n",
    ") +\n",
    "geom_histogram(bins = 20, show.legend = F) +\n",
    "# stat_summary(aes(y = 0), fun.data = give.median, geom = \"text\", vjust = 0, size = 2) +\n",
    "facet_grid(group_label~variable, scale = 'free') +\n",
    "scale_y_continuous(expand = c(0.2,0)) +\n",
    "scale_x_continuous(expand = c(0.1,0)) +\n",
    "geom_text(\n",
    "    inherit.aes = F,\n",
    "    data = D %>% group_by(group_label, variable) %>% summarise(median = median(na.omit(value))),\n",
    "    mapping = aes(y = 0, x = median, label = median),\n",
    "    size = 3,\n",
    "    vjust = 1\n",
    ") +\n",
    "theme_bw() +\n",
    "theme(\n",
    "    panel.spacing = unit(1, \"mm\"),\n",
    "    legend.position = 'top',\n",
    "    legend.title = element_blank(),\n",
    "    axis.text.x = element_text(size = 8, angle = 30, hjust = 1),\n",
    "    strip.text = element_text(size = 6),\n",
    "    strip.background = element_blank()\n",
    ") +\n",
    "xlab('') +\n",
    "ylab('Number of subjects')\n",
    "\n",
    "do_plot(p_cbc, 'cbc_dist.pdf', w = 8, h = 2.5)\n",
    "\n",
    "D %>% write.csv(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/cbc_dist_histrogram.csv\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WHO criteria for cytopenia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide = M_wide %>% \n",
    "    mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia\n",
    "    )\n",
    "\n",
    "D = M_wide %>% \n",
    "    mutate(group = factor(ifelse(ch_cnv == 1, 'mCA+', 'mCA-'), c('mCA-', 'mCA+'))) %>%\n",
    "#     mutate(group = factor(group, c('none', 'sm', 'cnv', 'both'))) %>%\n",
    "    select(group, thrombocytopenia, anemia, neutropenia, cytopenia) %>%\n",
    "    filter(complete.cases(.)) %>%\n",
    "    group_by(group) %>%\n",
    "    mutate(n = n()) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(group_label = paste0(group, '\\n(n=', n, ')')) %>%\n",
    "    mutate(group_label = factor(group_label, unique(group_label))) %>%\n",
    "    reshape2::melt(measure.vars = c('thrombocytopenia', 'anemia', 'neutropenia', 'cytopenia')) %>%\n",
    "    mutate(variable = ifelse(variable == 'cytopenia', 'any cytopenia', as.character(variable))) %>%\n",
    "    mutate(variable = Hmisc::capitalize(variable)) %>%\n",
    "    mutate(variable = factor(variable, unique(variable))) %>%\n",
    "    group_by(group_label, group, variable) %>%\n",
    "    summarise(\n",
    "        total = n(),\n",
    "        prop = sum(value)/total,\n",
    "        lower = qbinom(p = 0.05, size = total, prob = prop)/total,\n",
    "        upper = qbinom(p = 0.95, size = total, prob = prop)/total,\n",
    "    )\n",
    "\n",
    "\n",
    "\n",
    "p_cytopenia = ggplot(\n",
    "    D,\n",
    "    aes(x = group_label, y = prop, ymin = lower, ymax = upper, color = group)\n",
    ") +\n",
    "geom_errorbar(position = position_dodge(width = 0.3), width = 0.2) +\n",
    "geom_point(position = position_dodge(width = 0.3)) +\n",
    "theme_classic() +\n",
    "facet_wrap(~variable, scale = 'free_x', nrow = 1) +\n",
    "ylab('Proportion of subjects') +\n",
    "xlab('') +\n",
    "theme(\n",
    "    legend.title = element_blank(),\n",
    "    strip.text = element_text(size = 7),\n",
    "#     legend.position = 'none'\n",
    ")\n",
    "\n",
    "do_plot(p_cytopenia, 'cytopenia.pdf', w = 6.5, h = 2.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D %>% write.csv(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/cytopenia_mCA.csv\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### R1 Blood counts mCA +/- GM violin plots \n",
    "\n",
    "M_wide_Bcounts = M_wide %>% mutate (category = ifelse( ch_cnv_auto  & ch_nonsilent_auto, \"mCA + GM\", \"Rest\" )) %>%\n",
    "filter(anc < 30 & alc < 8 & amc < 3 & plt < 1000 | is.na(anc) | is.na(alc) | is.na(amc) | is.na(plt))  %>% \n",
    "group_by(category) %>%\n",
    "    mutate(n = n()) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(category_label = paste0(category, '\\n(n=', n, ')')) %>%\n",
    "    mutate(category_label = factor(category_label, unique(category_label))) %>%\n",
    "reshape2::melt(measure.vars = cbcs, variable.name = 'bc', value.name = 'count')\n",
    "\n",
    "\n",
    "p = ggplot(M_wide_Bcounts, aes(x = category, y = count, fill = category_label)) + geom_violin(fill = \"lightgrey\") + facet_wrap(~bc, scale = \"free\", nrow = 1) + \n",
    "    geom_boxplot(width=0.1, outlier.alpha = 0)+  theme_classic() + labs(fill = \"\")\n",
    "do_plot(p, \"R1_bloodcounts_mCA_vs_rest_n.pdf\", w = 15, h = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## R1 blood counts trans violin plots MS edited"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### R1 Blood counts trans violin plots \n",
    "\n",
    "# segs_filtered_auto = segs_filtered_auto %>% mutate(\n",
    "#         trans_effect = \n",
    "#             (chrom == 3 & type == 'amp' & TP53) |\n",
    "#             (chrom == 4 & type == 'del' & (!TET2) & (SRSF2 | ATM | CHEK2 | ASXL1)) |\n",
    "#             (chrom == 5 & type == 'del' & TP53) |\n",
    "#             (chrom == 7 & type == 'del' & (TP53 | PPM1D)) |\n",
    "#             (chrom == 8 & type == 'amp' & TET2) |\n",
    "#             (chrom == 12 & type == 'amp' & (NOTCH1 | MYD88 | FBXW7)) |\n",
    "#             (chrom == 13 & type == 'del' & ATM) |\n",
    "#             (chrom == 20 & type == 'del' & (U2AF1 | TERT))\n",
    "#     )\n",
    "\n",
    "\n",
    "\n",
    "M_wide_trans = M_wide %>% mutate(\n",
    "        trans_effect = \n",
    "            (chr3_amp & TP53) |\n",
    "            (chr4_del & (!TET2) & (SRSF2 | ATM | CHEK2 | ASXL1)) |\n",
    "            (chr5_del & TP53) |\n",
    "            (chr7_del & (TP53 | PPM1D)) |\n",
    "            (chr8_amp & TET2) |\n",
    "            (chr12_amp & (NOTCH1 | MYD88 | FBXW7)) |\n",
    "            (chr13_del & ATM) |\n",
    "            (chr20_del & (U2AF1 | TERT))\n",
    "    ) %>% mutate(\n",
    "        cis_effect = \n",
    "            (chr1_loh & MPL) |\n",
    "            (chr2_del & DNMT3A) | \n",
    "            ((chr4_del |chr4_loh) & TET2) | \n",
    "            (chr7_loh & EZH2) | \n",
    "            (chr9_loh & JAK2) | \n",
    "            ((chr11_loh | chr11_del) & ATM) | \n",
    "            (chr17 & TP53)\n",
    ")  %>% filter(anc < 30 & alc < 8 & amc < 3 & plt < 1000 | is.na(anc) | is.na(alc) | is.na(amc) | is.na(plt))  %>%\n",
    "mutate(category = ifelse(trans_effect|cis_effect , \"cis/trans\", \"Rest\")) %>% group_by(category) %>%\n",
    "    mutate(n = n()) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(category_label = paste0(category, '\\n(n=', n, ')')) %>%\n",
    "    mutate(category_label = factor(category_label, unique(category_label))) %>%\n",
    "reshape2::melt(measure.vars = cbcs, variable.name = 'bc', value.name = 'count') \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# M_wide_trans %>% count(trans_effect | cis_effect)\n",
    "\n",
    "\n",
    "\n",
    "p = ggplot(M_wide_trans, aes(x = category , y = count)) + geom_violin(fill = \"lightgrey\") + facet_wrap(~bc, scale = \"free\", nrow = 1) + geom_boxplot(width=0.1, outlier.alpha = 0)+  theme_classic()\n",
    "do_plot(p, \"R1_bloodcounts_cistrans_vs_rest.pdf\", w = 15, h = 2)\n",
    "\n",
    "\n",
    "p = ggplot(M_wide_trans, aes(x = category, y = count, fill = category_label)) + geom_violin(fill = \"lightgrey\") + facet_wrap(~bc, scale = \"free\", nrow = 1) + \n",
    "    geom_boxplot(width=0.1, outlier.alpha = 0)+  theme_classic() + labs(fill = \"\")\n",
    "do_plot(p, \"R1_bloodcounts_cistrans_vs_rest_n.pdf\", w = 15, h = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  R1 _ cis effect pateints with multiple mutations \n",
    "cis_multiple = M_wide %>% mutate(\n",
    "        trans_effect = \n",
    "            (chr3_amp & TP53) |\n",
    "            (chr4_del & (!TET2) & (SRSF2 | ATM | CHEK2 | ASXL1)) |\n",
    "            (chr5_del & TP53) |\n",
    "            (chr7_del & (TP53 | PPM1D)) |\n",
    "            (chr8_amp & TET2) |\n",
    "            (chr12_amp & (NOTCH1 | MYD88 | FBXW7)) |\n",
    "            (chr13_del & ATM) |\n",
    "            (chr20_del & (U2AF1 | TERT))\n",
    "    ) %>% mutate(\n",
    "        cis_effect = \n",
    "            (chr1_loh & MPL) |\n",
    "            (chr2_del & DNMT3A) | \n",
    "            ((chr4_del |chr4_loh) & TET2) | \n",
    "            (chr7_loh & EZH2) | \n",
    "            (chr9_loh & JAK2) | \n",
    "            ((chr11_loh | chr11_del) & ATM) | \n",
    "            (chr17 & TP53)\n",
    ")  %>%\n",
    "mutate(category = ifelse(trans_effect|cis_effect , \"cis/trans\", \"Rest\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pts_cis = cis_multiple %>% filter(chr1_loh & MPL) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"MPL\") %>% count(Gene)\n",
    "\n",
    "\n",
    "pts_cis = cis_multiple %>% filter(chr2_del & DNMT3A) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"DNMT3A\") %>% count(Gene)\n",
    "\n",
    "pts_cis = cis_multiple %>% filter((chr4_del |chr4_loh) & TET2) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"TET2\") %>% count(Gene)\n",
    "\n",
    "pts_cis = cis_multiple %>% filter(chr7_loh & EZH2) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"EZH2\") %>% count(Gene)\n",
    "\n",
    "pts_cis = cis_multiple %>% filter(chr9_loh & JAK2) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"JAK2\") %>% count(Gene)\n",
    "\n",
    "pts_cis = cis_multiple %>% filter((chr11_loh | chr11_del) & ATM) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"ATM\") %>% count(Gene)\n",
    "\n",
    "pts_cis = cis_multiple %>% filter(chr17 & TP53) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"TP53\") %>% count(Gene)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pts_cis = cis_multiple %>% filter((chr4_del |chr4_loh) & TET2) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"TET2\") %>% count(Gene)\n",
    "\n",
    "\n",
    "M_long %>% filter(dmp_patient_id %in% c(\"P-0007239\")) %>% select(dmp_patient_id, Gene, variant_id, start, AAchange, chr4_del ,chr4_loh)\n",
    "# M_long$AAchange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pts_cis = cis_multiple %>% filter(chr7_loh & EZH2) %>% select(dmp_sample_id)\n",
    "M_long %>% filter(dmp_sample_id %in% pts_cis$dmp_sample_id) %>% group_by(dmp_patient_id) %>% filter(Gene == \"EZH2\") %>% count(Gene)\n",
    "\n",
    "\n",
    "M_long %>% filter(dmp_patient_id %in% c( \"P-0047371\")) %>% select(dmp_patient_id, Gene, variant_id, start, AAchange, chr7_loh)\n",
    "M_long %>% filter(dmp_patient_id %in% c(\"P-0002903\")) %>% select(dmp_patient_id, Gene, variant_id, start, AAchange, chr7_loh)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_muts_pt = M_long %>% group_by(dmp_patient_id) %>% subset(Gene %in% c(\"DNMT3A\", \"TET2\", \"ASXL1\", \"JAK2\", \"TP53\", \"PPM1D\", \"ATM\", \"CHEK2\",\"EZH2\", \"MPL\")) %>% filter(as.logical(ch_nonsilent)) %>% count(Gene)\n",
    "n_muts_pt$Gene <- factor(n_muts_pt$Gene, levels=c(\"DNMT3A\", \"TET2\", \"PPM1D\", \"ASXL1\", \"TP53\",\"CHEK2\", \"ATM\", \"JAK2\", \"EZH2\", \"MPL\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_long %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "length(n_muts_pt$dmp_patient_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS edit\n",
    "\n",
    "\n",
    "# M_wide_cytopenia = M_wide %>% \n",
    "#     mutate(\n",
    "#         thrombocytopenia = plt < 100,\n",
    "#         anemia = hgb < 10,\n",
    "#         neutropenia = anc < 1.8,\n",
    "#         cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "#         lymphocytosis = alc > 5\n",
    "#     ) %>% mutate(\n",
    "#         trans_effect = \n",
    "#             (chr3_amp & TP53) |\n",
    "#             (chr4_del & (!TET2) & (SRSF2 | ATM | CHEK2 | ASXL1)) |\n",
    "#             (chr5_del & TP53) |\n",
    "#             (chr7_del & (TP53 | PPM1D)) |\n",
    "#             (chr8_amp & TET2) |\n",
    "#             (chr12_amp & (NOTCH1 | MYD88 | FBXW7)) |\n",
    "#             (chr13_del & ATM) |\n",
    "#             (chr20_del & (U2AF1 | TERT))\n",
    "#     ) %>% mutate(\n",
    "#         cis_effect = \n",
    "#             (chr1_loh & MPL) |\n",
    "#             (chr2_del & DNMT3A) | \n",
    "#             ((chr4_del |chr4_loh) & TET2) | \n",
    "#             (chr7_loh & EZH2) | \n",
    "#             (chr9_loh & JAK2) | \n",
    "#             ((chr11_loh | chr11_del) & ATM) | \n",
    "#             (chr17 & TP53)\n",
    "# )  %>%\n",
    "# mutate(category = ifelse(trans_effect|cis_effect , \"cis/trans\", \"Rest\"))\n",
    "\n",
    "\n",
    "\n",
    "### make one column w small categories\n",
    "\n",
    "M_wide_cytopenia = M_wide %>% \n",
    "    mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "        lymphocytosis = alc > 5\n",
    "    ) %>% mutate(category = case_when((chr3_amp & TP53)  ~ \"chr3_amp_TP53\",\n",
    "                                      (chr4_del & (!TET2) & (SRSF2 | ATM | CHEK2 | ASXL1)) ~ \"chr4_del & (!TET2) & (SRSF2 | ATM | CHEK2 | ASXL1)\",\n",
    "                                      (chr5_del & TP53) ~ \"chr5_del & TP53\",\n",
    "                                        (chr7_del & (TP53 | PPM1D)) ~ \"chr7_del & (TP53 | PPM1D)\", \n",
    "                                        (chr8_amp & TET2) ~ \"chr8_amp & TET2\",\n",
    "                                        (chr12_amp & (NOTCH1 | MYD88 | FBXW7)) ~ \"chr12_amp & (NOTCH1 | MYD88 | FBXW7)\", \n",
    "                                        (chr13_del & ATM) ~ \"chr13_del_ATM\",\n",
    "                                        (chr20_del & (U2AF1 | TERT)) ~ \"chr20_del & (U2AF1 | TERT)\",\n",
    "                                      (chr1_loh & MPL) ~ \"chr1_loh & MPL\",\n",
    "                                        (chr2_del & DNMT3A) ~ \"chr2_del & DNMT3A\",\n",
    "                                        ((chr4_del |chr4_loh) & TET2) ~ \"(chr4_del |chr4_loh) & TET2\",\n",
    "                                        (chr7_loh & EZH2) ~ \"chr7_loh & EZH2\",\n",
    "                                        (chr9_loh & JAK2) ~ \"chr9_loh & JAK2\",\n",
    "                                        ((chr11_loh | chr11_del) & ATM) ~ \"(chr11_loh | chr11_del) & ATM\",\n",
    "                                        (chr17 & TP53) ~ \"chr17 & TP53\",\n",
    "                           \n",
    "                            TRUE ~ \"rest\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_cytopenia %>% filter(cytopenia | lymphocytosis) %>% count(category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_cytopenia %>% mutate(cbc_cat = case_when(cytopenia ~ \"cytopenia\",\n",
    "                                               lymphocytosis ~ \"lymphocytosis\",\n",
    "                                               TRUE ~ \"rest\")) %>% group_by(cbc_cat) %>% count(category) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_cytopenia %>% mutate(cbc_cat = case_when(cytopenia ~ \"cytopenia\",\n",
    "                                               cytopenia ~ \"lymphocytosis\",\n",
    "                                               TRUE ~ \"rest\")) %>% mutate(normal_cbc = (cbc_cat == \"rest\") )  %>% count(category) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_cytopenia_tocast = M_wide_cytopenia %>% mutate(cbc_cat = case_when(cytopenia ~ \"cytopenia\",\n",
    "                                               lymphocytosis ~ \"lymphocytosis\",\n",
    "                                               TRUE ~ \"normal\")) %>% mutate(normal_cbc = (cbc_cat == \"normal\") ) %>% group_by(cbc_cat) %>% count(category) \n",
    "    dcast(M_wide_cytopenia_tocast, ( category ~ cbc_cat), value.var=\"n\", fill = 0) %>% mutate(total = rowSums(.[2:4])) # %>% write.csv(\"/work/isabl/home/sirenkom/teng_mCA/cytopenia_lymph_cis_trans.csv\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_cytopenia_tocast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_cytopenia %>% mutate(cbc_cat = case_when(cytopenia ~ \"cytopenia\",\n",
    "                                               lymphocytosis ~ \"lymphocytosis\",\n",
    "                                               TRUE ~ \"rest\")) %>% group_by(category) %>% count(cbc_cat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS edit \n",
    "M_wide_cytopenia_hemecat = M_wide_cytopenia %>% mutate(cbc_cat = case_when(cytopenia ~ \"cytopenia\",\n",
    "                                               lymphocytosis ~ \"lymphocytosis\",\n",
    "                                               TRUE ~ \"normal\")) %>% mutate(normal_cbc = (cbc_cat == \"normal\") ) %>% \n",
    "mutate( heme_cat3 = case_when( \n",
    "                                    heme_cat2 %in%c( \"MDS\", \"MPN\", \"AML\", \"CML\") ~ \"Myeloid\", \n",
    "                                    heme_cat2 %in% \"CLL\" ~ c(\"CLL\"),\n",
    "                                    heme_cat2 %in% c(\"B-NHL\", \"T-NHL\") ~ \"Lymphoid\" , \n",
    "                                    T  ~ \"None\" \n",
    "\n",
    "\n",
    ")) %>% group_by(cbc_cat) %>% count(heme_cat3) \n",
    "    dcast(M_wide_cytopenia_hemecat, ( heme_cat3 ~ cbc_cat), value.var=\"n\", fill = 0) %>% mutate(total = rowSums(.[2:4]))  %>% write.csv(\"/work/isabl/home/sirenkom/teng_mCA/cytopenia_lymph_hemecat3.csv\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS edit add landmark \n",
    "landmark = c(90, 180, 270, 360)\n",
    "for (j in landmark){\n",
    "     print(j)\n",
    "\n",
    "M_wide_cytopenia_hemecat = M_wide_cytopenia %>% filter(time_anyblood > j) %>% mutate(cbc_cat = case_when(cytopenia ~ \"cytopenia\",\n",
    "                                               lymphocytosis ~ \"lymphocytosis\",\n",
    "                                               TRUE ~ \"normal\")) %>% mutate(normal_cbc = (cbc_cat == \"normal\") ) %>% \n",
    "mutate( heme_cat3 = case_when( \n",
    "                                    heme_cat2 %in%c( \"MDS\", \"MPN\", \"AML\", \"CML\") ~ \"Myeloid\", \n",
    "                                    heme_cat2 %in% \"CLL\" ~ c(\"CLL\"),\n",
    "                                    heme_cat2 %in% c(\"B-NHL\", \"T-NHL\") ~ \"Lymphoid\" , \n",
    "                                    T  ~ \"None\" \n",
    "\n",
    "\n",
    ")) %>% group_by(cbc_cat) %>% count(heme_cat3) \n",
    "    dcast(M_wide_cytopenia_hemecat, ( heme_cat3 ~ cbc_cat), value.var=\"n\", fill = 0) %>% mutate(total = rowSums(.[2:4]))  %>% write.csv(paste(\"/work/isabl/home/sirenkom/teng_mCA/cytopenia_lymph_hemecat3_landmark\", j, \".csv\", sep = \"\"), quote = FALSE)\n",
    "    \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS edit split by mCa +/- GM\n",
    "\n",
    "\n",
    "M_wide_mCA_GM = M_wide_cytopenia  %>% mutate(cbc_cat = case_when(cytopenia ~ \"cytopenia\",\n",
    "                                               lymphocytosis ~ \"lymphocytosis\",\n",
    "                                               TRUE ~ \"normal\")) %>% mutate(normal_cbc = (cbc_cat == \"normal\") ) %>% \n",
    "mutate( heme_cat3 = case_when( \n",
    "                                    heme_cat2 %in%c( \"MDS\", \"MPN\", \"AML\", \"CML\") ~ \"Myeloid\", \n",
    "                                    heme_cat2 %in% \"CLL\" ~ c(\"CLL\"),\n",
    "                                    heme_cat2 %in% c(\"B-NHL\", \"T-NHL\") ~ \"Lymphoid\" , \n",
    "                                    T  ~ \"None\" \n",
    "\n",
    "\n",
    ")) %>% mutate (ch_type = case_when( (ch_cnv & ch_nonsilent) ~ \"mCA + GM\",\n",
    "                                  (ch_cnv & !ch_nonsilent) ~ \"mCA\",\n",
    "                                   (!ch_cnv & ch_nonsilent) ~ \"GM\",\n",
    "                                   TRUE ~ \"rest\"\n",
    "                                  ) \n",
    ") %>% \n",
    "group_by(cbc_cat) %>% count(ch_type) \n",
    "    dcast(M_wide_mCA_GM, ( ch_type ~ cbc_cat), value.var=\"n\", fill = 0) %>% mutate(total = rowSums(.[2:4]))  %>% write.csv(paste(\"/work/isabl/home/sirenkom/teng_mCA/cytopenia_lymph_GM_mCA.csv\", sep = \"\"), quote = FALSE)\n",
    "    \n",
    "M_wide_mCA_GM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS edit\n",
    "\n",
    "# n_muts_pt\n",
    "p_muts = n_muts_pt %>% ggplot(aes(x = Gene)) + geom_bar(stat = \"count\", aes(fill = factor(n, levels = c(\"5\", \"4\", \"3\", \"2\", \"1\")))) + \n",
    " scale_fill_brewer( palette = rev(\"Dark2\"),\n",
    "                       name=\"Number Mutations\",\n",
    "                       labels=c(\"5\", \"4\", \"3\", \"2\", \"1\")) + theme_classic() + ylim(0,4000) + ylab(\"Number of Patients\")\n",
    "\n",
    "do_plot(p_muts, \"R1_muts_per_gene_patient\", h = 5, w = 7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### R1 Blood counts heme category disease violin plots MS edited\n",
    "\n",
    "M_wide_hemecat = M_wide %>% mutate( heme_cat3 = case_when( \n",
    "                                    heme_cat2 %in%c( \"MDS\", \"MPN\", \"AML\", \"CML\") ~ \"Myeloid\", \n",
    "                                    heme_cat2 %in% \"CLL\" ~ c(\"CLL\"),\n",
    "                                    heme_cat2 %in% c(\"B-NHL\", \"T-NHL\") ~ \"Lymphoid\" , \n",
    "                                    T  ~ \"None\" \n",
    "\n",
    "\n",
    ")) %>% filter(anc < 30 & alc < 8 & amc < 3 & plt < 1000 | is.na(anc) | is.na(alc) | is.na(amc) | is.na(plt))  %>% \n",
    "    group_by(heme_cat3) %>%\n",
    "    mutate(n = n()) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(category_label = paste0(heme_cat3, '\\n(n=', n, ')')) %>%\n",
    "    mutate(category_label = factor(category_label, unique(category_label))) %>%\n",
    "reshape2::melt(measure.vars = cbcs, variable.name = 'bc', value.name = 'count') \n",
    "\n",
    "\n",
    "p = ggplot(M_wide_hemecat, aes(x = heme_cat3 , y = count)) + geom_violin(fill = \"lightgrey\") + facet_wrap(~bc, scale = \"free\", nrow = 1) + \n",
    "    geom_boxplot(width=0.1, outlier.alpha = 0)+  theme_classic()+ theme(axis.text.x = element_text(angle = 45, hjust=1))\n",
    "do_plot(p, \"R1_bloodcounts_heme_disease_1row.pdf\", w = 20, h = 2) \n",
    "\n",
    "\n",
    "p = ggplot(M_wide_hemecat, aes(x = heme_cat3, y = count, fill = category_label)) + geom_violin(fill = \"lightgrey\") + facet_wrap(~bc, scale = \"free\", nrow = 1) + \n",
    "    geom_boxplot(width=0.1, outlier.alpha = 0)+  theme_classic() + labs(fill = \"\") + theme(axis.text.x = element_text(angle = 45, hjust=1))\n",
    "do_plot(p, \"R1_bloodcounts_heme_disease_1row_n.pdf\", w = 15, h = 2)\n",
    "\n",
    "p = ggplot(M_wide_hemecat, aes(x = heme_cat3 , y = count)) + geom_violin(fill = \"lightgrey\") + facet_wrap(~bc, scale = \"free\", nrow = 2) + \n",
    "    geom_boxplot(width=0.1, outlier.alpha = 0)+  theme_classic()\n",
    "do_plot(p, \"R1_bloodcounts_heme_disease_2row.pdf\", w = 15, h = 5) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Serial CBC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cbc_cnv = fread(\"/work/isabl/home/sirenkom/teng_mCA/input_files_to_upload/cbc_serial_cnv.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_cbc = function(cbc_serial) {\n",
    "    cbc_serial %>%\n",
    "    reshape2::melt(measure.vars = cbcs, variable.name = 'bc', value.name = 'count') %>%\n",
    "    mutate(\n",
    "        cytopenia = case_when(\n",
    "            bc == 'plt' ~ count < 100,\n",
    "            bc == 'hgb' ~ count < 10,\n",
    "            bc == 'anc' ~ count < 1.8,\n",
    "            T ~ F\n",
    "        )\n",
    "    ) %>%\n",
    "    ggplot(\n",
    "        aes(x = DAYSFROMIMPACT, y = count, group = bc, color = cytopenia)\n",
    "    ) +\n",
    "    geom_vline(xintercept = 0, linetype = 'dashed', color = 'gray') +\n",
    "    geom_line() +\n",
    "    geom_point(size = 1) +\n",
    "    theme_classic() +\n",
    "    facet_wrap(~bc, scale = 'free') +\n",
    "    theme(legend.position = 'none') +\n",
    "    scale_color_manual(values = c('royalblue', 'red'))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cbcs = c(\"anc\", \"alc\", \"amc\", \"hgb\", \"mcv\", \"rdw\", \"plt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CBC ~ CNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnvs = colnames(CNV_detailed)[-1][(CNV_detailed[,-1] %>% colSums) >= 5]\n",
    "\n",
    "cnvs = cnvs[cnvs != 'chr9_loh']\n",
    "cnvs = c(cnvs, 'chr9_p_loh', 'chr9_q_loh')\n",
    "cbcs = cbcs[cbcs != 'rbc']\n",
    "\n",
    "summs = data.frame()\n",
    "\n",
    "for (bc in paste0(cbcs[!cbcs %in% c('rbc', 'wbc')], '_n')) {\n",
    "    display(bc)\n",
    "    for (cnv in cnvs) {\n",
    "        \n",
    "        summ = M_wide %>%\n",
    "            lm(\n",
    "                glue('{bc} ~ Gender + race_b + age10 + smoke + {cnv}'),\n",
    "                data = .\n",
    "            ) %>%\n",
    "            sjPlot::get_model_data(type = 'est') %>%\n",
    "            mutate(bc = bc)\n",
    "\n",
    "        summs = rbind(summs, summ)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 1.8, repr.plot.res = 300)\n",
    "\n",
    "limit = 5\n",
    "\n",
    "D = summs %>% \n",
    "    filter(term %in% cnvs) %>%\n",
    "    rowwise() %>%\n",
    "    mutate(estimate = max(min(estimate, limit), -limit)) %>% \n",
    "    ungroup() %>%\n",
    "    mutate(q = p.adjust(p.value, method = 'BH')) %>%\n",
    "    mutate(signif_cat = case_when(\n",
    "        q < 0.01 ~ 'q<0.01',\n",
    "        q < 0.05 ~ 'q<0.05',\n",
    "        q < 0.1 ~ 'q<0.1',\n",
    "        T ~ 'ns'\n",
    "    )) %>%\n",
    "    mutate(\n",
    "        bc = str_remove(bc, '_n'),\n",
    "        term = str_remove_all(term, '_|chr'),\n",
    "        term = str_replace(str_replace(str_replace(term, 'del', '-'), 'amp', '+'), 'loh', '='),\n",
    "        chrom = as.integer(str_extract(term, '\\\\d+'))\n",
    "    ) %>%\n",
    "    arrange(chrom) %>%\n",
    "    mutate(term = factor(term, unique(term)))\n",
    "\n",
    "ggplot(\n",
    "    D,\n",
    "    aes(x = term, y = bc, fill = estimate)\n",
    ") +\n",
    "theme_classic() +\n",
    "geom_tile(width = 0.9, height = 0.9) +\n",
    "geom_point(\n",
    "    data = D %>% filter(q < 0.1),\n",
    "    aes(pch = signif_cat)\n",
    "#         aes(size = -log10(p.value))\n",
    ") +\n",
    "scale_shape_manual(values = c('q<0.01' = 8, 'q<0.05' = 3, 'q<0.1' = 20), name = 'FDR') +\n",
    "scale_fill_gradient2(low = 'lightblue', high = 'red', mid = 'white', na.value = 'white', midpoint = 0, limits = c(-limit, limit)) +\n",
    "theme(plot.margin = unit(c(0,0,0,0), 'mm'), legend.key.size = unit(0.5, 'cm')) +\n",
    "theme(\n",
    "    axis.text.x = element_text(angle = 30, hjust = 0.5, size = 8),\n",
    "    axis.text.y = element_text(size = 8),\n",
    "    legend.key.size = unit(2.5, 'mm'),\n",
    "    legend.text = element_text(size = 8),\n",
    "    legend.title = element_text(size = 8),\n",
    "    legend.spacing = unit(1, 'mm')\n",
    ") +\n",
    "xlab('') +\n",
    "ylab('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CBC ~ CNV + GM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combs = list(c('JAK2', 'chr9_p_loh'), c('EZH2', 'chr7_loh'),\n",
    "             c('TET2', 'chr4_del', 'chr4_loh'),\n",
    "             c('MPL', 'chr1_p_loh'), c('ATM', 'chr11_del', 'chr11_loh'),\n",
    "            c('DNMT3A', 'chr2_del'),\n",
    "            c(''))\n",
    "\n",
    "summs = data.frame()\n",
    "\n",
    "for (bc in paste0(cbcs[!cbcs %in% c('rbc', 'wbc')], '_n')) {\n",
    "    display(bc)\n",
    "    for (comb in combs) {\n",
    "        \n",
    "        comb = paste(comb, collapse = '+')\n",
    "        \n",
    "        summ = M_wide %>%\n",
    "            lm(\n",
    "                glue('{bc} ~ Gender + race_b + age10 + smoke + {comb}'),\n",
    "                data = .\n",
    "            ) %>%\n",
    "            sjPlot::get_model_data(type = 'est') %>%\n",
    "            mutate(bc = bc)\n",
    "\n",
    "        summs = rbind(summs, summ)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 6, repr.plot.height = 2, repr.plot.res = 300)\n",
    "\n",
    "limit = 5\n",
    "\n",
    "D = summs %>% \n",
    "    filter(!str_detect(term, 'Gender|race|smoke|age')) %>%\n",
    "    rowwise() %>%\n",
    "    mutate(estimate = max(min(estimate, limit), -limit)) %>% \n",
    "    ungroup() %>%\n",
    "    mutate(q = p.adjust(p.value, method = 'BH')) %>%\n",
    "#     mutate(q = p.value) %>%\n",
    "    mutate(signif_cat = case_when(\n",
    "        q < 0.01 ~ 'q<0.01',\n",
    "        q < 0.05 ~ 'q<0.05',\n",
    "        q < 0.1 ~ 'q<0.1',\n",
    "        T ~ 'ns'\n",
    "    )) %>%\n",
    "    mutate(term = factor(term, unlist(combs))) %>%\n",
    "    arrange(term) %>%\n",
    "    mutate(\n",
    "        bc = str_remove(bc, '_n'),\n",
    "        term = str_remove_all(term, '_|chr'),\n",
    "        term = str_replace(str_replace(str_replace(term, 'del', '-'), 'amp', '+'), 'loh', '='),\n",
    "        chrom = as.integer(str_extract(term, '\\\\d+'))\n",
    "    ) %>%\n",
    "    mutate(term = factor(term, unique(term)))\n",
    "\n",
    "ggplot(\n",
    "    D,\n",
    "    aes(x = term, y = bc, fill = estimate)\n",
    ") +\n",
    "theme_classic() +\n",
    "geom_tile(width = 0.9, height = 0.9) +\n",
    "geom_point(\n",
    "    data = D %>% filter(q < 0.1),\n",
    "    aes(pch = signif_cat)\n",
    "#         aes(size = -log10(p.value))\n",
    ") +\n",
    "scale_shape_manual(values = c('q<0.01' = 8, 'q<0.05' = 3, 'q<0.1' = 20), name = 'FDR') +\n",
    "scale_fill_gradient2(low = 'lightblue', high = 'red', mid = 'white', na.value = 'white', midpoint = 0, limits = c(-limit, limit)) +\n",
    "theme(plot.margin = unit(c(0,0,0,0), 'mm'), legend.key.size = unit(0.5, 'cm')) +\n",
    "theme(\n",
    "    axis.text.x = element_text(angle = 30, hjust = 1, size = 8),\n",
    "    axis.text.y = element_text(size = 8),\n",
    "    legend.key.size = unit(2.5, 'mm'),\n",
    "    legend.text = element_text(size = 8),\n",
    "    legend.title = element_text(size = 8),\n",
    "    legend.spacing = unit(1, 'mm')\n",
    ") +\n",
    "xlab('') +\n",
    "ylab('')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Patient Outcome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = ggplot(\n",
    "    M_wide %>% \n",
    "    filter(!is.na(heme_cat2)) %>% \n",
    "    group_by(heme_cat2) %>%\n",
    "    mutate(n_heme_cat = n()) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(heme_cat2 = glue('{heme_cat2}({n_heme_cat})')) %>%\n",
    "    mutate(seq_to_diag = seq_to_diag/365),\n",
    "    aes(x = seq_to_diag, fill = heme_cat2)\n",
    ") +\n",
    "geom_histogram(bins = 50, color = 'black', size = 0.1) +\n",
    "scale_y_continuous(expand = c(0,0)) +\n",
    "xlab('Years since blood draw') +\n",
    "theme_classic() +\n",
    "theme(\n",
    "    legend.title = element_blank(),\n",
    "    legend.position = 'top'\n",
    ") +\n",
    "ylab('Cases')\n",
    "\n",
    "do_plot(p, 'diagnosis.pdf', w = 5, h = 2.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(M_wide %>% filter(post_heme == 1) %>% pull(seq_to_diag) %>% mean)/30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% filter(post_heme == 1) %>% pull(seq_to_diag)  %>% range %>% {./30}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_cnv = M_wide %>% filter(post_heme == 1 & heme_cat == 'MDS' & (!ch_cnv_auto)) %>% pull(seq_to_diag)\n",
    "yes_cnv = M_wide %>% filter(post_heme == 1 & heme_cat == 'MDS' & ch_cnv_auto) %>% pull(seq_to_diag)\n",
    "\n",
    "t.test(no_cnv, yes_cnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% \n",
    "    filter(post_heme == 1) %>% \n",
    "    mutate(\n",
    "        heme_cat = ifelse(heme_cat %in% c('ET', 'PV'), 'MPN', heme_cat),\n",
    "        heme_cat = ifelse(heme_cat %in% c('ALCL', 'CTCL'), 'T-NHL', heme_cat),\n",
    "        heme_cat = ifelse(heme_cat %in% c('DLBCL', 'FL', 'MZL', 'PCNSL', 'WM'), 'B-NHL', heme_cat)\n",
    "    ) %>%\n",
    "    group_by(heme_cat) %>%\n",
    "    mutate(n_heme_cat = n()) %>%\n",
    "    ungroup() %>%\n",
    "    mutate(seq_to_diag = seq_to_diag/30) %>% \n",
    "    group_by(heme_cat) %>%\n",
    "    summarise(\n",
    "        n = n(),\n",
    "        seq_to_diag = mean(seq_to_diag)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% \n",
    "    filter(post_heme == 1) %>%\n",
    "    filter(heme_cat == 'MDS') %>%\n",
    "    group_by(ch_cnv_auto) %>%\n",
    "    summarise(round(median(seq_to_diag)/30, 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_set = fread('./data/gene_set.tsv')$Gene\n",
    "gene_set = c('DNMT3A', 'TET2', 'JAK2', 'ASXL1', 'TP53', 'ATM', 'PPM1D', 'CHEK2',\n",
    "             'SRSF2', 'SF3B1', 'U2AF1', 'EZH2', 'MPL', 'NOTCH1', 'TERT', 'SUZ12', 'CBL')\n",
    "\n",
    "patients_heme = M_wide %>% filter(post_heme == 1) %>% pull(dmp_patient_id)\n",
    "\n",
    "segs_auto = segs_filtered %>% filter(chrom != '23')\n",
    "\n",
    "M_co = rbind(\n",
    "        M_long %>% \n",
    "        filter(Gene %in% gene_set) %>%\n",
    "        filter(ch_nonsilent == 1 | VariantClass == 'Addition') %>%\n",
    "        mutate(\n",
    "            mutation = Gene,\n",
    "            type = VariantClass2,\n",
    "            arm = 'na',\n",
    "            focality = 'na'\n",
    "        ) %>%\n",
    "        rowwise() %>% mutate(phi = min(VAF_N * 2, 1)) %>% ungroup() %>%\n",
    "        select(mutation, phi, type, arm, focality, dmp_patient_id),\n",
    "        segs_auto %>% \n",
    "        mutate(type_label = c('loh' = '=', 'del' = '-', 'amp' = '+')[type]) %>%\n",
    "        mutate(mutation = paste0(chrom, ifelse(arm == 'p,q', '', arm), type_label)) %>%\n",
    "        select(mutation, phi, type = type2, arm, focality, dmp_patient_id)\n",
    "    ) %>%\n",
    "    filter(dmp_patient_id %in% patients_heme) %>%\n",
    "    full_join(\n",
    "        M_wide %>% filter(post_heme == 1) %>%\n",
    "            select(dmp_patient_id, heme_cat, seq_to_diag),\n",
    "        by = 'dmp_patient_id'\n",
    "    ) %>%\n",
    "    filter(heme_cat != 'LCH') %>%\n",
    "    mutate(\n",
    "        heme_cat = ifelse(heme_cat %in% c('ET', 'PV'), 'MPN', heme_cat),\n",
    "        heme_cat = ifelse(heme_cat %in% c('ALCL', 'CTCL'), 'T-NHL', heme_cat),\n",
    "        heme_cat = ifelse(heme_cat %in% c('DLBCL', 'FL', 'MZL', 'PCNSL', 'WM'), 'B-NHL', heme_cat)\n",
    "    ) %>%\n",
    "    mutate(heme_cat = factor(heme_cat, c('MDS', 'MPN', 'AML', 'CML', 'CLL', 'B-NHL', 'T-NHL', 'MM'))) %>%\n",
    "    mutate(mutation = ifelse(is.na(mutation), 'none', mutation))\n",
    "\n",
    "M_co = M_co %>% mutate(\n",
    "    class = case_when(\n",
    "        mutation %in% c('JAK2', 'MPL', 'CHEK2', '9ploh') ~ 'MPN',\n",
    "        mutation %in% c('TP53') ~ 'TP53',\n",
    "        mutation %in% c('EZH2', 'SRSF2', 'U2AF1', 'SF3B1', 'ASXL1', 'SUZ12') ~ 'MDS',\n",
    "        mutation %in% c('PPM1D', '7qdel') ~ 'CMML',\n",
    "        mutation %in% c('NOTCH1', '12amp', '19amp', 'ATM', 'MYD88') ~ 'CLL',\n",
    "        T ~ 'other'\n",
    "    ),\n",
    "    class_rank = as.integer(factor(class, c('MPN', 'TP53', 'MDS', 'CMML', 'CLL', 'other')))\n",
    ")\n",
    "\n",
    "mutation_order = c('DNMT3A', 'TET2', 'ASXL1', 'JAK2', 'MPL', 'CHEK2',\n",
    "                   'SRSF2', 'SF3B1', 'U2AF1', 'EZH2', 'SUZ12', 'PPM1D', 'NOTCH1','ATM',\n",
    "                   'MYD88', 'TERT', 'TP53')\n",
    "\n",
    "mutation_other = M_co %>% filter(!mutation %in% mutation_order) %>% pull(mutation) %>% unique\n",
    "\n",
    "mutation_order = c(mutation_order, mutation_other)\n",
    "\n",
    "D = M_co %>%\n",
    "#     filter(type == 'snv') %>%\n",
    "    distinct(mutation, dmp_patient_id, `.keep_all` = T) %>%\n",
    "    group_by(mutation) %>% mutate(mutation_n = n()) %>% ungroup() %>%\n",
    "    arrange(class_rank, mutation != 'EZH2', desc(mutation_n)) %>%\n",
    "    mutate(mutation = factor(mutation, rev(unique(mutation)))) %>% \n",
    "    rowwise() %>%\n",
    "    mutate(mut_weight = 2^which(mutation == levels(mutation))) %>%\n",
    "    ungroup() %>%\n",
    "    group_by(dmp_patient_id) %>% \n",
    "    mutate(patient_weight = sum(mut_weight)) %>%\n",
    "    ungroup() %>%\n",
    "    group_by(dmp_patient_id) %>% \n",
    "    mutate(patient_group = min(class_rank)) %>%\n",
    "    ungroup() %>%\n",
    "    arrange(heme_cat, seq_to_diag, desc(patient_weight), type) %>%\n",
    "    mutate(pid = as.integer(factor(dmp_patient_id, unique(dmp_patient_id)))) %>%\n",
    "    mutate(mutation = as.character(mutation)) %>%\n",
    "    mutate(mutation = ifelse(mutation == 'none', NA, mutation)) %>%\n",
    "    mutate(mutation = factor(mutation, rev(mutation_order))) %>%\n",
    "    mutate(mutation_i = as.integer(mutation))\n",
    "\n",
    "p_mat = ggplot(\n",
    "        D,\n",
    "        aes(x = pid, y = mutation, fill = type)\n",
    "    ) +\n",
    "    geom_tile(color = 'white', size = 0.5) +\n",
    "    geom_vline(\n",
    "        aes(xintercept = pid - 0.5), color = 'gray95', size = 0.3\n",
    "    ) +\n",
    "#     geom_hline(\n",
    "#         yintercept = length(levels(D$mutation)) - 1 - c(2.5), color = 'gray80',\n",
    "#     ) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        axis.text.x = element_blank(),\n",
    "        axis.title.x = element_blank(),\n",
    "        plot.margin = unit(c(0,0,0,0), 'mm'),\n",
    "        legend.key.size = unit(2, 'mm')\n",
    "    ) +\n",
    "    geom_vline(xintercept = cumsum(\n",
    "        D %>% distinct(pid, heme_cat) %>% count(heme_cat) %>% pull(n)\n",
    "    ) + 0.5, size = 0.2, linetype = 'dashed') +\n",
    "    scale_fill_manual(\n",
    "        values = c(cnv_pal2, sm_pal),\n",
    "        na.translate = FALSE\n",
    "    ) +\n",
    "    scale_color_manual(\n",
    "        values = c(cnv_pal2, sm_pal),\n",
    "        na.translate = FALSE\n",
    "    ) +\n",
    "    scale_x_discrete(expand = expansion(add = 0)) +\n",
    "    scale_y_discrete(expand = c(0.01,0), na.translate = FALSE) +\n",
    "    xlab('') +\n",
    "    scale_alpha_continuous(range = c(0.4,1)) +\n",
    "    labs(alpha = 'Cell fraction') +\n",
    "    labs(fill = 'Mutation') +\n",
    "    ylab('')\n",
    "\n",
    "p_heme = D %>% \n",
    "    mutate(seq_to_diag = seq_to_diag/365) %>%\n",
    "    distinct(dmp_patient_id, pid, heme_cat, seq_to_diag) %>% \n",
    "    group_by(heme_cat) %>%\n",
    "    mutate(label_pos = median(pid)) %>%\n",
    "    ungroup %>%\n",
    "    ggplot(\n",
    "        aes(x = pid, y = 1, fill = heme_cat, alpha = seq_to_diag)\n",
    "    ) +\n",
    "    geom_tile(color = 'white', size = 0.5, show.legend = T) +\n",
    "#     geom_vline(xintercept = cumsum(\n",
    "#         D %>% distinct(pid, heme_cat) %>% count(heme_cat) %>% pull(n)\n",
    "#     ) + 0.5, size = 0.2, linetype = 'dashed') +\n",
    "    geom_rect(xmin = 0, xmax = max(D$pid) + 0.5, ymin = 1.5, ymax = 3, fill = 'white', show.legend = F) +\n",
    "    geom_hline(yintercept = 0.5, size = 1) +\n",
    "    geom_text(\n",
    "        aes(x = label_pos, y = 2, label = heme_cat),\n",
    "        color = 'black',\n",
    "        vjust = 1,\n",
    "        size = 2,\n",
    "        show.legend = F\n",
    "    ) +\n",
    "    theme_void() +\n",
    "    theme(\n",
    "        axis.text.x = element_blank(),\n",
    "        legend.key.size = unit(2, 'mm'),\n",
    "        plot.margin = unit(c(0,0,0,0), 'mm'),\n",
    "        legend.margin = margin(l = 1.7,1,1,1, \"mm\")\n",
    "    ) +\n",
    "    scale_x_discrete(expand = expansion(add = 0)) +\n",
    "    scale_y_discrete(expand = expansion(add = 0)) +\n",
    "    ylab('') +\n",
    "    scale_fill_manual(\n",
    "        values = c(' ' = 'white', 'CLL' = 'royalblue', 'MDS' = 'tomato3', 'CML' = 'yellow',\n",
    "                   'AML' = 'red', 'MPN' = 'purple', 'B-NHL' = 'yellowgreen', 'T-NHL' = 'orchid', 'MM' = 'orange')\n",
    "    ) +\n",
    "    scale_alpha_continuous(trans = 'reverse') +\n",
    "    labs(alpha = \"Years to \\ndiagnosis\") +\n",
    "    guides(fill = FALSE)\n",
    "\n",
    "panel = (p_heme / p_mat) + plot_layout(heights = c(1,20), guides = 'collect')\n",
    "\n",
    "do_plot(panel, 'heme_heatmap.pdf', w = 7.7, h = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% filter(heme_cat2 == 'MPN' & chr9) %>% pull(seq_to_diag)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Risk models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "betas = data.frame()\n",
    "\n",
    "for (landmark in seq(90, 360, 90)) {\n",
    "    \n",
    "    D = M_wide %>% \n",
    "        filter(time_leukemia > landmark) %>%\n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        ) %>%\n",
    "        ungroup()\n",
    "    \n",
    "    fit = coxph(Surv(time_leukemia, post_leukemia) ~ ch_cnv_auto + VAF_pd_10 + mutnum_pd + age, data = D, na.action = 'na.omit')\n",
    "    \n",
    "    cox.sum = sjPlot::get_model_data(fit, type = 'est') %>% mutate(with_cnv = 0, landmark = landmark)\n",
    "    \n",
    "    betas = rbind(\n",
    "        betas,\n",
    "        cox.sum\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 4, repr.plot.height = 2, repr.plot.res = 300)\n",
    "\n",
    "p = betas %>%\n",
    "    filter(landmark == 270) %>%\n",
    "    filter(!(term %in% c('GenderMale', 'age'))) %>%\n",
    "    mutate(landmark = paste0(landmark, ' days')) %>%\n",
    "    mutate(landmark = factor(landmark, unique(landmark))) %>%\n",
    "    mutate(term = c('VAF_pd_10' = 'PD VAF', 'mutnum_pd' = 'PD Mutation Number', 'ch_cnv_auto' = 'mCA')[as.character(term)]) %>%\n",
    "    mutate(term = factor(term, unique(term))) %>%\n",
    "    mutate(format.p.value = format.pval(p.value, digits = 2)) %>% \n",
    "    plot_forest(\n",
    "        x = 'term',\n",
    "        label = 'format.p.value',\n",
    "        eb_w = 0,\n",
    "        eb_s = 1,\n",
    "        ps = 2,\n",
    "        or_s = 3,\n",
    "        nudge = -0.6\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    ylab('HR') +\n",
    "    xlab('')\n",
    "\n",
    "do_plot(p, 'cox_pvalue.pdf', w = 4, h = 1.8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "landmark = 30\n",
    "D = M_wide %>% \n",
    "        filter(time_leukemia > landmark) %>%\n",
    "        filter(therapy_detailed ) %>%\n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        ) %>%\n",
    "        ungroup()\n",
    "    \n",
    "    fit = coxph(Surv(time_leukemia, post_leukemia) ~  age +  ind_cytotoxic_therapy + XRT, data = D, na.action = 'na.omit')\n",
    "    \n",
    "    cox.sum = sjPlot::get_model_data(fit, type = 'est') %>% mutate(with_cnv = 0, landmark = landmark)\n",
    "    \n",
    "    \n",
    "cox.sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## filter cytopenia / lymph\n",
    "\n",
    "\n",
    "\n",
    "betas = data.frame()\n",
    "\n",
    "for (landmark in seq(90, 360, 90)) {\n",
    "    \n",
    "    D = M_wide %>% mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "        lymphocytosis = alc > 5\n",
    "    ) %>% filter(!(cytopenia|lymphocytosis) ) %>% \n",
    "        filter(time_leukemia > landmark) %>%\n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        ) %>%\n",
    "        ungroup()\n",
    "    \n",
    "    fit = coxph(Surv(time_leukemia, post_leukemia) ~ ch_cnv_auto + VAF_pd_10 + mutnum_pd + age, data = D, na.action = 'na.omit')\n",
    "    \n",
    "    cox.sum = sjPlot::get_model_data(fit, type = 'est') %>% mutate(with_cnv = 0, landmark = landmark)\n",
    "    \n",
    "    betas = rbind(\n",
    "        betas,\n",
    "        cox.sum\n",
    "    )\n",
    "}\n",
    "betas\n",
    "\n",
    "\n",
    "\n",
    "betas_unfilt = data.frame()\n",
    "\n",
    "for (landmark in seq(90, 360, 90)) {\n",
    "    \n",
    "    D = M_wide %>% mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "        lymphocytosis = alc > 5\n",
    "    ) %>% filter(!(cytopenia|lymphocytosis) ) %>% \n",
    "        filter(time_leukemia > landmark) %>%\n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        ) %>%\n",
    "        ungroup()\n",
    "    \n",
    "    fit = coxph(Surv(time_leukemia, post_leukemia) ~ ch_cnv_auto + VAF_pd_10 + mutnum_pd + age, data = D, na.action = 'na.omit')\n",
    "    \n",
    "    cox.sum = sjPlot::get_model_data(fit, type = 'est') %>% mutate(with_cnv = 0, landmark = landmark)\n",
    "    \n",
    "    betas_unfilt = rbind(\n",
    "        betas_unfilt,\n",
    "        cox.sum\n",
    "    )\n",
    "}\n",
    "betas_unfilt = betas_unfilt %>% mutate(filt_status = \"unfiltered\")\n",
    "betas = betas %>% mutate(filt_status = \"filtered\")\n",
    "\n",
    "betas_toplot = rbind(betas_unfilt, betas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(M_wide %>% mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "        lymphocytosis = alc > 5\n",
    "    ) %>% filter(!(cytopenia|lymphocytosis) ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = betas_toplot %>%\n",
    "    filter(!(term %in% c('GenderMale', 'age'))) %>%\n",
    "    mutate(landmark = paste0(landmark, ' days')) %>%\n",
    "    mutate(landmark = factor(landmark, unique(landmark))) %>%\n",
    "    mutate(term = c('VAF_pd_10' = 'PD VAF', 'mutnum_pd' = 'PD Mutation Number', 'ch_cnv_auto' = 'mCA')[as.character(term)]) %>%\n",
    "    mutate(term = factor(term, unique(term))) %>%\n",
    "    plot_forest(\n",
    "        x = 'term',\n",
    "        label = 'p.stars',\n",
    "        eb_w = 0,\n",
    "        eb_s = 1,\n",
    "        ps = 2,\n",
    "        or_s = 3,\n",
    "        nudge = 0.1,\n",
    "        col = 'landmark', \n",
    "        shape = 'filt_status'\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    ylab('HR') +\n",
    "    xlab('') +\n",
    "    scale_color_discrete(name = 'Landmark')\n",
    "\n",
    "do_plot(p, 'filter_unfiltered_landmark_pvalue.pdf', w = 4.5, h = 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = betas %>%\n",
    "    filter(!(term %in% c('GenderMale', 'age'))) %>%\n",
    "    mutate(landmark = paste0(landmark, ' days')) %>%\n",
    "    mutate(landmark = factor(landmark, unique(landmark))) %>%\n",
    "    mutate(term = c('VAF_pd_10' = 'PD VAF', 'mutnum_pd' = 'PD Mutation Number', 'ch_cnv_auto' = 'mCA')[as.character(term)]) %>%\n",
    "    mutate(term = factor(term, unique(term))) %>%\n",
    "    plot_forest(\n",
    "        x = 'term',\n",
    "        label = 'p.stars',\n",
    "        eb_w = 0,\n",
    "        eb_s = 1,\n",
    "        ps = 2,\n",
    "        or_s = 3,\n",
    "        nudge = 0.1,\n",
    "        col = 'landmark'\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    ylab('HR') +\n",
    "    xlab('') +\n",
    "    scale_color_discrete(name = 'Landmark')\n",
    "\n",
    "do_plot(p, 'filter_cytolymph_landmark.pdf', w = 4.5, h = 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "\n",
    "options(repr.plot.width = 4, repr.plot.height = 2, repr.plot.res = 300)\n",
    "betas_filt = data.frame()\n",
    "\n",
    "\n",
    "## filt - filter out cytopenias and lymphocytosis\n",
    "landmark = 270\n",
    "D = M_wide %>%  mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "        lymphocytosis = alc > 5\n",
    "    ) %>%   mutate( category = case_when( !(cytopenia | lymphocytosis )   ~ \"normal CBC\",\n",
    "                                         (cytopenia | lymphocytosis ) ~ \"abnormal\",\n",
    "                                                                                TRUE ~ \"exclude\")) %>% \n",
    "        \n",
    "        filter(time_leukemia > landmark) %>% \n",
    "        filter(category == \"normal CBC\") %>% \n",
    "        \n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        ) %>%\n",
    "        ungroup()\n",
    "    \n",
    "    fit = coxph(Surv(time_leukemia, post_leukemia) ~ ch_cnv_auto + VAF_pd_10 + mutnum_pd + age, data = D, na.action = 'na.omit')\n",
    "    \n",
    "    cox.sum = sjPlot::get_model_data(fit, type = 'est') %>% mutate(with_cnv = 0, landmark = landmark)\n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "betas_filt = rbind(\n",
    "        betas_filt,\n",
    "        cox.sum)\n",
    "\n",
    "## unfilt \n",
    "betas_unfilt = data.frame()\n",
    "    \n",
    "D = M_wide  %>% \n",
    "        filter(time_leukemia > landmark) %>%\n",
    "       \n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        ) %>%\n",
    "        ungroup()\n",
    "    \n",
    "    fit = coxph(Surv(time_leukemia, post_leukemia) ~ ch_cnv_auto + VAF_pd_10 + mutnum_pd + age, data = D, na.action = 'na.omit')\n",
    "    \n",
    "    cox.sum = sjPlot::get_model_data(fit, type = 'est') %>% mutate(with_cnv = 0, landmark = landmark)\n",
    "    betas_unfilt = rbind(\n",
    "        betas_unfilt,\n",
    "        cox.sum)\n",
    "\n",
    "## bind together \n",
    "betas_filt = betas_filt %>% mutate(filt_status = \"filtered\")\n",
    "betas_unfilt = betas_unfilt %>% mutate(filt_status = \"unfiltered\")\n",
    "betas = rbind(betas_filt, betas_unfilt)\n",
    "\n",
    "\n",
    "## compare filt and unfilt\n",
    "p = betas  %>% \n",
    "    filter(landmark == 270) %>%\n",
    "    filter(!(term %in% c('GenderMale', 'age'))) %>%\n",
    "    mutate(landmark = paste0(landmark, ' days')) %>%\n",
    "    mutate(landmark = factor(landmark, unique(landmark))) %>%\n",
    "    mutate(term = c('VAF_pd_10' = 'PD VAF', 'mutnum_pd' = 'PD Mutation Number', 'ch_cnv_auto' = 'mCA')[as.character(term)]) %>%\n",
    "    mutate(term = factor(term, unique(term))) %>% \n",
    "    mutate(format.p.value = format.pval(p.value, digits = 2)) %>% \n",
    "    plot_forest(\n",
    "        x = 'term',\n",
    "        label = 'format.p.value',\n",
    "        eb_w = 0,\n",
    "        eb_s = 1,\n",
    "        ps = 2,\n",
    "        or_s = 3,\n",
    "        nudge = -0.5, \n",
    "        col = 'filt_status'\n",
    "    ) +\n",
    "    theme_classic() +\n",
    "    ylab('HR') +\n",
    "    xlab('') +scale_color_discrete(labels = c(\"No cytopenia or lymphocytosis (n=17816)\", \"All (n=22428)\") )+  theme(legend.title = element_blank())+ theme(legend.position=\"top\", legend.direction = \"vertical\")\n",
    "\n",
    "do_plot(p, 'MS_edit_cox_all_normal_pvalue.pdf', w = 4.3, h = 3)\n",
    "\n",
    "betas\n",
    "\n",
    "betas  %>% \n",
    "    filter(landmark == 270) %>%\n",
    "    filter(!(term %in% c('GenderMale', 'age'))) %>%\n",
    "    mutate(landmark = paste0(landmark, ' days')) %>%\n",
    "    mutate(landmark = factor(landmark, unique(landmark))) %>%\n",
    "    mutate(term = c('VAF_pd_10' = 'PD VAF', 'mutnum_pd' = 'PD Mutation Number', 'ch_cnv_auto' = 'mCA')[as.character(term)]) %>%\n",
    "    mutate(term = factor(term, unique(term))) %>% \n",
    "    mutate(format.p.value = format.pval(p.value, digits = 2)) %>% fwrite(\"/work/isabl/home/sirenkom/teng_mCA/csv_dfs/SFig20.csv\", sep = \"\\t\", quote = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>%  mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "        lymphocytosis = alc > 5\n",
    "    ) %>%   mutate( category = case_when( !(cytopenia | lymphocytosis )   ~ \"normal CBC\",\n",
    "                                         (cytopenia | lymphocytosis ) ~ \"abnormal\",\n",
    "                                                                                TRUE ~ \"exclude\")) %>% \n",
    "        \n",
    "        filter(time_leukemia > landmark) %>% \n",
    "        filter(category == \"normal CBC\") %>% \n",
    "        \n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>% mutate(n_pts = length(unique(dmp_patient_id))) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        )    %>%\n",
    "        ungroup()\n",
    "\n",
    "\n",
    "M_wide  %>% \n",
    "        filter(time_leukemia > landmark) %>% mutate(n_pts = length(unique(dmp_patient_id)))  %>%\n",
    "       \n",
    "        mutate(time_leukemia = time_leukemia - landmark) %>%\n",
    "        rowwise() %>%\n",
    "        mutate(\n",
    "#             VAF_pd_10 = min(VAF_pd, 0.5)/0.1,\n",
    "            VAF_pd_10 = VAF_pd/0.1\n",
    "        )  %>%\n",
    "        ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_long %>% filter(dmp_patient_id %in% c(\"P-0006225\", \"P-0002903\")) %>% filter(Gene == \"EZH2\") %>%  arrange(dmp_patient_id) %>% select(dmp_patient_id, Gene, VAF_N, AAchange)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>%  mutate(\n",
    "        thrombocytopenia = plt < 100,\n",
    "        anemia = hgb < 10,\n",
    "        neutropenia = anc < 1.8,\n",
    "        cytopenia = thrombocytopenia | anemia | neutropenia,\n",
    "        lymphocytosis = alc > 5\n",
    "    ) %>%   mutate( category = case_when( !(cytopenia | lymphocytosis )   ~ \"normal CBC\",\n",
    "                                         (cytopenia | lymphocytosis ) ~ \"abnormal\",\n",
    "                                                                                TRUE ~ \"exclude\")) %>% \n",
    "        \n",
    "        filter(time_leukemia > landmark)  %>% count(category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "betas %>%\n",
    "filter(landmark == 270)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cumulative incidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data frame for Sean\n",
    "M_wide %>% \n",
    "select(dmp_patient_id, ch_cnv_auto, ch_nonsilent, DateofProcedure, date_death,\n",
    "       date_lastfu, post_leukemia, post_mn, seq_to_diag) %>%\n",
    "#fwrite('./data/survival_sean.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(prodlim)\n",
    "library(cmprsk)\n",
    "\n",
    "data001 = M_wide %>% mutate(seq_to_diag = ifelse(post_leukemia, seq_to_diag, NA))\n",
    "\n",
    "OSDays <- as.numeric(as.Date(data001$date_lastfu , \"%Y-%m-%d\") - as.Date(data001$DateofProcedure , \"%Y-%m-%d\"))\n",
    "Dead <- 1*(!is.na(as.Date(data001$date_death , \"%Y-%m-%d\")))\n",
    "\n",
    "TimetoAnalysis <- data001$seq_to_diag\n",
    "TimetoAnalysis[is.na(TimetoAnalysis)] <- OSDays[is.na(TimetoAnalysis)]\n",
    "\n",
    "YearstoAnalysis <- TimetoAnalysis/365\n",
    "data001$YearstoAnalysis <- YearstoAnalysis\n",
    "\n",
    "EventAnalysis <- 2*Dead\n",
    "EventAnalysis[!is.na(data001$seq_to_diag)] <- 1\n",
    "data001$EventAnalysis <- EventAnalysis\n",
    "\n",
    "CH <- rep(\"None\", length(data001$ch_nonsilent))\n",
    "CH[data001$ch_pd ==0 & data001$ch_cnv_auto==1] <- \"mCA ONLY\"\n",
    "CH[data001$ch_pd ==1 & data001$ch_cnv_auto==0] <- \"GM ONLY\"\n",
    "CH[data001$ch_pd ==1 & data001$ch_cnv_auto==1] <- \"mCA+GM\"\n",
    "\n",
    "data001$CH <- CH\n",
    "\n",
    "data001 = data001 %>% mutate(CH = factor(CH, c('None', 'GM ONLY', 'mCA ONLY', 'mCA+GM')))\n",
    "\n",
    "data002 <- data001[data001$YearstoAnalysis >0 , ] ## need to remove the <= 0 times\n",
    "dim(data002)\n",
    "\n",
    "cmv <- prodlim(Hist(YearstoAnalysis, EventAnalysis) ~CH, data=data002)\n",
    "summary(cmv,3) ### if the lower confidence interval =0 in the output just put at <0.001 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "par(mar = c(3,2,1,1))\n",
    "\n",
    "pdf(file = './figures/cum_inc.pdf', width = 6, height = 6)\n",
    "\n",
    "plot(cmv, \n",
    "     xlim=c(0, 4), ylim=c(0,0.25),\n",
    "     xlab = \"Years from blooddraw\", ylab = \"Cumulative Incidence\",\n",
    "     atRisk.at = seq(0, 4, 1),\n",
    "     axis1.at = seq(0, 4, 1),\n",
    "     atrisk.labels = \"\",  \n",
    "     legend.cex = 1,\n",
    "     col = c('#abdda4', '#2b83ba','#fdae61', '#d7191c'),\n",
    "     legend.x=\"topleft\") \n",
    "title(main=\"\", cex=3)\n",
    "\n",
    "dev.off()\n",
    "\n",
    "display_pdf(file = './figures/cum_inc.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## P-value\n",
    "print(cuminc(YearstoAnalysis, EventAnalysis, CH))\n",
    "## P-value is <0.001 [corresponding test statistics is 174.7842, under 3 d.f.s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tally"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segs_filtered %>%\n",
    "filter(post_anyblood == 1) %>%\n",
    "mutate(heme_cat = ifelse(heme_cat %in% c('ET', 'PV'), 'MPN', heme_cat)) %>%\n",
    "reshape2::dcast(heme_cat ~ chrom + type, fun.aggregate = function(x){length(unique(x))}, value.var = 'dmp_patient_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "segs_filtered %>%\n",
    "filter(post_anyblood == 1) %>%\n",
    "mutate(heme_cat = ifelse(heme_cat %in% c('ET', 'PV'), 'MPN', heme_cat)) %>%\n",
    "reshape2::dcast(heme_cat ~ ., fun.aggregate = function(x){length(unique(x))}, value.var = 'dmp_patient_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>%\n",
    "mutate(heme_cat = ifelse(heme_cat %in% c('ET', 'PV'), 'MPN', heme_cat)) %>%\n",
    "filter(post_anyblood == 1) %>% \n",
    "group_by(heme_cat) %>%\n",
    "summarise(\n",
    "    total = n(),\n",
    "    n_ai = sum(ch_cnv_auto),\n",
    "    n_sm = sum(ch_nonsilent),\n",
    "    n_ai_only = sum(ch_cnv_auto & (!ch_nonsilent))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>%\n",
    "filter(!is.na(heme_cat2)) %>% \n",
    "summarise(\n",
    "    total = n(),\n",
    "    n_ai = sum(ch_cnv_auto),\n",
    "    prop_ai = n_ai/total,\n",
    "    n_sm = sum(ch_nonsilent),\n",
    "    n_sm_only = sum(ch_nonsilent & (!ch_cnv_auto)),\n",
    "    prop_sm_only = n_sm_only/total,\n",
    "    n_ai_only = sum(ch_cnv_auto & (!ch_nonsilent)),\n",
    "    prop_ai_only = n_ai_only/total\n",
    ")\n",
    "\n",
    "M_wide %>%\n",
    "filter(!is.na(heme_cat2)) %>% \n",
    "group_by(heme_cat2) %>%\n",
    "summarise(\n",
    "    total = n(),\n",
    "    n_ai = sum(ch_cnv_auto),\n",
    "    n_sm = sum(ch_nonsilent),\n",
    "    n_sm_only = sum(ch_nonsilent & (!ch_cnv_auto)),\n",
    "    n_ai_only = sum(ch_cnv_auto & (!ch_nonsilent))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% count(ch_cnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% filter((JAK2) & chr9_p_loh) %>% filter(post_heme == 1) %>% pull(VAF_JAK2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overall Survival"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_survival = M_wide %>% mutate (death_status = ifelse(is.na(date_death) , 0,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colnames(M_wide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5.4, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "fit = M_wide_survival %>%\n",
    "    survfit(formula = Surv(time_lastfu, death_status) ~ ch_cnv_auto, data = .)\n",
    "\n",
    "size_coef = 0.3\n",
    "\n",
    "# png( \"/work/isabl/home/sirenkom/teng_mCA/R1_overall_survival.png\",p, width = 6, height = 6)\n",
    "ggsurvplot(\n",
    "    fit,\n",
    "    # Defining plotting parameters\n",
    "    data = M_wide_survival,\n",
    "    size = 3 * size_coef,              # Change line size\n",
    "    fontsize = 10 * size_coef,           # Size of the table's labels\n",
    "    palette = 'Dark2',\n",
    "    ggtheme = theme_classic(),         # Change ggplot2 theme\n",
    "    tables.theme = clean_theme(),    # Change table's theme\n",
    "    # Adding useful info to the plot\n",
    "#     conf.int = TRUE,         # Add confidence interval\n",
    "    pval = TRUE,               # Add p-value\n",
    "    pval.size = 12 * size_coef,\n",
    "    risk.table = TRUE,         # Add risk table\n",
    "    # Adding labels and title\n",
    "    #  title = paste0(g,\" survival curves\"),\n",
    "    legend.labs = c(\"mCA-\",  \"mCA+\"),    # Change legend labels\n",
    "    legend.title = \"\",\n",
    "    legend = \"right\",\n",
    "    xlab = \"Time in days\",    # Customize X axis label\n",
    "    ylab = \"tMN free proportion\",    # Customize Y axis label\n",
    "    # Defining risk table plotting parameters\n",
    "    risk.table.col = \"strata\",        # Risk table color by groups\n",
    "    risk.table.height = 0.25,   # Useful to change when you have multiple groups\n",
    "    risk.table.y.text = FALSE  # Displaying colors instead of name in risk table\n",
    ")\n",
    "# dev.off()\n",
    "\n",
    "# do_plot(p, \"R1_overall_survival.pdf\", w = 5, h = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5.4, repr.plot.height = 5, repr.plot.res = 300)\n",
    "\n",
    "fit = M_wide_survival %>%\n",
    "    survfit(formula = Surv(time_lastfu, death_status) ~ ch_nonsilent_auto, data = .)\n",
    "\n",
    "size_coef = 0.3\n",
    "\n",
    "# png( \"/work/isabl/home/sirenkom/teng_mCA/R1_overall_survival.png\",p, width = 6, height = 6)\n",
    "ggsurvplot(\n",
    "    fit,\n",
    "    # Defining plotting parameters\n",
    "    data = M_wide_survival,\n",
    "    size = 3 * size_coef,              # Change line size\n",
    "    fontsize = 10 * size_coef,           # Size of the table's labels\n",
    "    palette = 'Dark2',\n",
    "    ggtheme = theme_classic(),         # Change ggplot2 theme\n",
    "    tables.theme = clean_theme(),    # Change table's theme\n",
    "    # Adding useful info to the plot\n",
    "#     conf.int = TRUE,         # Add confidence interval\n",
    "    pval = TRUE,               # Add p-value\n",
    "    pval.size = 12 * size_coef,\n",
    "    risk.table = TRUE,         # Add risk table\n",
    "    # Adding labels and title\n",
    "    #  title = paste0(g,\" survival curves\"),\n",
    "    legend.labs = c(\"mCA-\",  \"mCA+\"),    # Change legend labels\n",
    "    legend.title = \"\",\n",
    "    legend = \"right\",\n",
    "    xlab = \"Time in days\",    # Customize X axis label\n",
    "    ylab = \"tMN free proportion\",    # Customize Y axis label\n",
    "    # Defining risk table plotting parameters\n",
    "    risk.table.col = \"strata\",        # Risk table color by groups\n",
    "    risk.table.height = 0.25,   # Useful to change when you have multiple groups\n",
    "    risk.table.y.text = FALSE  # Displaying colors instead of name in risk table\n",
    ")\n",
    "# dev.off()\n",
    "\n",
    "# do_plot(p, \"R1_overall_survival.pdf\", w = 5, h = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "getwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide_survival %>% count( death_status)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## co-occuring cnvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## MS EDIT \n",
    "\n",
    "segs_filtered %>% group_by(dmp_patient_id) %>% filter(length(unique(cnv_label))>1) %>% summarise(all_cnvs = paste(unique(sort(cnv_label)), collapse = \",\")) %>% count(all_cnvs) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "segs_filtered %>% filter(chrom == 20 & type == 'del') %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "segs_filtered %>% filter(chrom == 23 & type == 'del') %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "segs_filtered %>% filter(chrom == 15 & type == 'amp') %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "segs_filtered %>% filter(chrom == 12 & type == 'amp') %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "segs_filtered %>% filter(chrom == 11 & type == 'loh') %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "segs_filtered %>% filter(chrom == 14 & type == 'loh') %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "segs_filtered %>% filter(chrom == 22 & type == 'loh') %>% count(ch_nonsilent)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_wide %>% filter(!ind_cytotoxic_therapy & !XRT & therapy_detailed) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 3.6.1 gaot",
   "language": "R",
   "name": "ir361_gaot"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  },
  "toc-autonumbering": true,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
